---
title: "Estimating the contribution of large-scale environmental and genetic influences on depression symptoms in a large US longitudinal cohort" 
author: "Katherine N Thompson"
date: "2024-09-17"
output:  
  html_document:
    df_print: paged
    toc: yes
    toc_depth: 2
    toc_float:
      collapsed: false
    number_sections: false
    highlight: monochrome
    theme: flatly
    code_folding: show
    includes:
      after_body: footer.html
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                       comment = NA,
                       prompt = FALSE,
                       cache = FALSE,
                       message = FALSE,
                       warning = FALSE,
                       results = 'markup')
 
 options(bitmapType = 'quartz') # to render fonts better
```

```{r clear global environment, include=FALSE}
remove(list = ls())
```

To download packages on rossmann cluster use the following in the terminal - 
Module load r/4.4.1
Rscript -e "install.packages('tidyverse', repos='https://cloud.r-project.org')"

```{r load packages, include=FALSE}
library(knitr)
library(haven)
library(psych) 
library(tidyr)
library(lavaan)
library(ggplot2)
library(data.table)
library(reshape2)
library(questionr) 
library(viridis) # For a color-blind-friendly palette
library(RColorBrewer) # For high-contrast color palettes
library(egg)
library(purrr)
library(tibble)
library(Cairo)
library(tidyverse)
library(dplyr) #conflicts with other packages
```

```{r set seed, include=FALSE}
# Set seed for reproducibility
set.seed(1293734)
```

# Data prep

## Read in clean Add Health data

### Wave I

```{r read in R data from cleaning stage1}
load(file = "/project/rche/data/datasets/addhealth/clean_progress/WAVEI/STAGE2/dat_clean_WAVEI_STAGE2.RData")

# inspect
head(dat) 
colnames(dat)      # column names - hash to avoid printing in HTML

# checks 
table(dat$BIO_SEX_numeric_clean) # summary for sex variables
table(dat$BIO_SEX_factor_clean) # summary for sex variables
length(dat[duplicated(dat$AID)]) # check for duplicated IDs
length(colnames(dat)) # number of variables included
dat[, duplicated(colnames(dat))] # check for duplicated variables

# number of character variables
char_var <- colnames(dat)[
  !grepl("_numeric|_factor", colnames(dat)) & 
    sapply(dat, is.character)
]
char_var
length(char_var)
char_var_Ids <- c("AID", "SCID", "SSCID", "COMMID", "INTID", "VERSION", "SQID", "SSCHLCDE", "MACNO") # these are all descriptive ID variables
char_var_noIds <- setdiff(char_var, char_var_Ids) # only one non-ID character variable

# number of numeric variables 
num_var <- colnames(dat)[
  !grepl("_numeric|_factor", colnames(dat)) & 
    sapply(dat, is.numeric) &
    !grepl("^[a-z]", colnames(dat)) # no variables that begin with a lower case letter 
]
num_var
length(num_var)

num_var_true <- colnames(dat)[
  !grepl("_numeric|_factor|raw|clean", colnames(dat)) & 
    sapply(dat, is.numeric) &
    !grepl("^[a-z]", colnames(dat)) # no variables that begin with a lower case letter 
]
num_var_true
length(num_var_true) # this includes all the raw numeric variables 

# number of categorical variables 
cat_var <- colnames(dat)[
  grepl("_numeric_clean", colnames(dat))
]
cat_var
length(cat_var)

# number of categorical variables - all formats
cat_var_all <- colnames(dat)[
  grepl("_numeric_clean|_numeric_raw|_factor_clean|_factor_raw", colnames(dat))
]
cat_var_all
length(cat_var_all)

all_var_lists <- c(cat_var_all, num_var, char_var)

# Find variables in the dataset that are not in the combined lists - these are recoded variables 
recoded_vars <- colnames(dat)[
    grepl("^[a-z]", colnames(dat)) # recoded variables that begin with a lower case letter 
]
recoded_vars
length(recoded_vars)

length(cat_var_all) + length(num_var) + length(char_var) + length(recoded_vars) # total number of variables = 10428
length(cat_var) + length(num_var) + length(char_var) # raw number of variables = 2844
length(cat_var) + length(num_var_true) + 1 # true number of variables to be included in analyses = 2833 (one character var that isn't an ID variable)
```

Keep individuals that have all the key information variables.  

```{r filter on key variables}
# Create a new dataset with only individuals included in the regression
dat <- dat[!is.na(dat$H1GI1Y_numeric_clean) &
           !is.na(dat$BIO_SEX_numeric_clean), ]
```

### Wave 4

We will read in depression scores and diagnostic information from wave IV. 

```{r read in wave 4 raw data}
dat.W4.raw <- read_xpt("/project/rche/data/datasets/addhealth/Core Files - Wave IV/Wave IV In Home Interview Data/wave4/wave4.xpt")
```

```{r read in wave 4 depression items}
CESD.W4.items <- c("H4MH18", 
                   "H4MH19", 
                   "H4MH20", 
                   "H4MH21", 
                   "H4MH22", 
                   "H4MH23", 
                   "H4MH24", 
                   "H4MH25", 
                   "H4MH26", 
                   "H4MH27") 

CESD.W4.items.clean <- c("H4MH18_clean", 
                   "H4MH19_clean", 
                   "H4MH20_clean", 
                   "H4MH21_clean", 
                   "H4MH22_clean", 
                   "H4MH23_clean", 
                   "H4MH24_clean", 
                   "H4MH25_clean", 
                   "H4MH26_clean", 
                   "H4MH27_clean") 
```

```{r cleaning missingness coding}
dat.W4 <- dat.W4.raw %>%
  mutate(across(c(CESD.W4.items, H4ID5H, H4ID6H), 
                ~ case_when(. %in% c(6, 7, 8, 9) ~ NA_real_, 
                            TRUE ~ .),
                .names = "{.col}_clean")) 
```

### Ancestry FLAG

Here - you need to input which ancestry group you would like to subset the code to run for:

* All ancestries together
* European ancestry
* African ancestry
* Admixed american/hispanic ancestry 
* East Asian ancestry

ALL - All ancestries together
EUR – European (Mostly Northern, Southern, and Western European ancestry)
AFR – African (Sub-Saharan African ancestry)
AMR – Admixed American (Latin American ancestry, often a mix of Indigenous, European, and African)
EAS – East Asian (Chinese, Japanese, Korean, etc.)
SAS – South Asian (India, Pakistan, Bangladesh, Sri Lanka, etc.) - *not included here*

This will run all following code within the ancestry flag that you select - and will organise the output into the file structure depending on that ancestry. Please unhash the ancestry that you would like to run the code for. 

The above N is printed for the sample you will run the script for. This is only applicable for the subsample models - as these are the only models with genetic data. For the full sample models - this flag will not apply. 

```{r ancestry code flag}
# ancestry_flag = "ALL"
 ancestry_flag = "EUR"
# ancestry_flag = "AFR"
# ancestry_flag = "AMR"
# ancestry_flag = "EAS"
```

### Genetic data

All genetic data has been through genotype imputation and quality control separately to this script. 

*The following conditions have been applied to each ancestry group separately:* 

Stage 1: Post-imputation check
* SNP QC: Filter SNPs with an INFO score (indicate the quality of genotype imputation) <0.8 

Stage 2: SNP QC (mostly)
* SNP QC: call rate ≥ 0.98 (--geno 0.02)
* SNP QC: with MAF ≥ 0.01 (--maf 0.01)
* SNP QC: Hardy-Weinberg equilibrium (HWE) in controls p value ≥ 1e-06 (--hwe 1e-6) - ancestry specific
* Sample QC: call rate in cases or controls ≥ 0.98 (--mind 0.02)

Stage 3: Sample QC
* Sample QC: FHET within +/- 0.20 in cases or controls (--het) - ancestry specific 
* Sample QC: Sex violations - genetic sex does not match pedigree sex (--check-sex) - ancestry specific

Stage 4: Additional QC: 
* Sample QC: Excluding one of any pair of individuals with a genomic relationship >0.125

Look at the N for post-QC genetic data. 

```{r read in the fam file from the raw genetic data}
if(ancestry_flag == "ALL"){
postQC_gen_dat <- read.table("/project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/archive/imputed/1000g/addhealth_imputed_1000g.QC.rel.ALL.fam", header = FALSE)
}

if(ancestry_flag == "AFR"){
postQC_gen_dat <- read.table("/project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/archive/imputed/1000g/addhealth_imputed_1000g.QC.AFR.rel.fam", header = FALSE)
}

if(ancestry_flag == "AMR"){
postQC_gen_dat <- read.table("/project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/archive/imputed/1000g/addhealth_imputed_1000g.QC.AMR.rel.fam", header = FALSE)
}

if(ancestry_flag == "EAS"){
postQC_gen_dat <- read.table("/project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/archive/imputed/1000g/addhealth_imputed_1000g.QC.EAS.rel.fam", header = FALSE)
}

if(ancestry_flag == "EUR"){
postQC_gen_dat <- read.table("/project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/archive/imputed/1000g/addhealth_imputed_1000g.QC.EUR.rel.fam", header = FALSE)
}

grm_ids_nosibs <- postQC_gen_dat %>% dplyr::rename(FID = V1, AID = V2) %>% mutate(across(everything(), as.numeric))
  
# individuals with QC genetic data
nrow(grm_ids_nosibs)
grm_ids_nosibs
```

The structure of all ID (fam) files read in is FID, AID,  0, 0, 0, 0
The structure of mtg2 will need to be          AID, AID2, 0, 0, 0, 0

### Ancestry PCs

Addhelath reports genetic ancestry in it's second release of data. 
In the file /project/rche/data/datasets/addhealth/Genetic Files/Wave IV Polygenic Scores - Release 2/pgs2.xpt 

PCs have been created from our addhealth data. 

```{r 20 ancestry PCs}
ancestry_PCs <- read.table("/project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/archive/ahgwas.QC.rel.ALL.pca.eigenvec", header = FALSE)

ancestry_PCs <- ancestry_PCs %>% dplyr::rename(FID = V1, 
                                        AID = V2, 
                                        an_PC1 = V3, 
                                        an_PC2 = V4, 
                                        an_PC3 = V5, 
                                        an_PC4 = V6, 
                                        an_PC5 = V7, 
                                        an_PC6 = V8, 
                                        an_PC7 = V9, 
                                        an_PC8 = V10, 
                                        an_PC9 = V11, 
                                        an_PC10 = V12, 
                                        an_PC11 = V13, 
                                        an_PC12 = V14,
                                        an_PC13 = V15,
                                        an_PC14 = V16,
                                        an_PC15 = V17,
                                        an_PC16 = V18,
                                        an_PC17 = V19,
                                        an_PC18 = V20,
                                        an_PC19 = V21,
                                        an_PC20 = V22) %>% 
  mutate(across(everything(), as.numeric))

ancestry_PCs
```

```{r read in release2 ancestry}
rel2 <- read_xpt("/project/rche/data/datasets/addhealth/Genetic Files/Wave IV Polygenic Scores - Release 2/pgs2.xpt")

colnames(rel2)

rel2_ancestry <- rel2 %>%
  mutate(AID = as.character(AID), FID = as.character(FID)) %>%
  select(AID, PSANCEST, PSRACE)

# join genetic ancestry and self-reported race to dat file
dat <- left_join(dat, rel2_ancestry, by = "AID")

# have a look at frequencies for *genetic* ancestry - this will only be for people that have genetic data
freq(dat$PSANCEST)
```
```{r subset based on ancestry pheno data}
if(ancestry_flag == "ALL"){
  # data
  dat <- dat
  # paths
  ancestry_sub_path <- "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/"
  full_sample_path <- "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/full_sample/"
  # N
  print(nrow(dat))
}

if(ancestry_flag == "EUR"){
  # data
  dat <- dat %>% filter(PSANCEST == "European ancestry")
  # paths
  ancestry_sub_path <- "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/"
  # N
  print(nrow(dat))
}

if(ancestry_flag == "AFR"){
  # data
  dat <- dat %>% filter(PSANCEST == "African ancestry")
  # paths
  ancestry_sub_path <- "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/"
  # N
  print(nrow(dat))
}

if(ancestry_flag == "AMR"){
  # data
  dat <- dat %>% filter(PSANCEST == "Hispanic ancestry")
  # paths
  ancestry_sub_path <- "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/"
  # N
  print(nrow(dat))
}

if(ancestry_flag == "EAS"){
  # data
  dat <- dat %>% filter(PSANCEST == "East Asian ancest")
  # paths
  ancestry_sub_path <- "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/"
  # N
  print(nrow(dat))
}
```

## Functions

```{r functions}
# replace ggsave - cluster issues with ggsave()
save_gg_cairo <- function(plot, filename, width = 900, height = 800) {
  library(Cairo)
  CairoPNG(filename, width = width, height = height)
  print(plot)
  if (!is.null(dev.list())) dev.off()
}

# function to identify variables with more than a specified number of NAs
find_vars_with_na <- function(data, threshold) {
  vars_with_na <- colSums(is.na(data))
  vars_with_na_more_than_threshold <- names(vars_with_na[vars_with_na > threshold])
  num_na <- vars_with_na[vars_with_na > threshold]
  return(list(num_na))
}

# Random imputation function
random_sample_imputation <- function(x) {
  missing <- is.na(x) # Identify missing values
  x[missing] <- sample(x[!missing], size = sum(missing), replace = TRUE) # Draw the exact number of replacements
  return(x)
}

# # function to count the number of individuals with NA for all variables beginning with "P"
count_na_all_P <- function(data) {
  na_rows <- apply(data, 1, function(x) all(is.na(x[grep("^P", names(x))])))
  num_na_all_P <- sum(na_rows)
  return(num_na_all_P)
}

# function to drop individuals with NA for over a specified percentage of variables
drop_individuals_na_over_percentage <- function(data, threshold_percentage) {
  num_vars <- ncol(data)
  threshold_count <- threshold_percentage / 100 * num_vars
  na_counts <- rowSums(is.na(data))
  data_filtered <- data[na_counts <= threshold_count, ]
  return(data_filtered)
}

# Function to create dummy variables for each categorical variable, excluding a reference category
create_dummy_variables <- function(data, reference_category = NULL) {
  numeric_vars <- names(data)[grep("_numeric", names(data))] # Find variables containing "_numeric"
  
  for (var in numeric_vars) {
    categories <- unique(data[[var]]) # Get unique categories
    categories <- categories[!is.na(categories)] # Remove NA values
    
    # Only proceed if there are more than 2 categories
    if (length(categories) > 2) {
      # Define the reference category
      ref_cat <- if (!is.null(reference_category[[var]])) {
        reference_category[[var]]
      } else {
        categories[1]  # Default to the first category if no specific reference is provided
      }
      
      # Exclude the reference category from dummy variable creation
      for (cat in categories[categories != ref_cat]) {
        data <- data %>%
          mutate(!!paste0(var, "_dummy_", cat) := as.numeric(data[[var]] == cat))
      }
      
      # Remove the original variable, only keep dummy variables
      data <- data %>%
        select(-all_of(var))
    }
  }
  
  return(data)
}

# likelihood ratio test
LRT <- function(path_full, path_constrained, Outcome = NA, Comparison = NA, df) {
  # Construct full paths
  lines_full <- readLines(paste0(ancestry_sub_path, path_full))
  lines_constrained <- readLines(paste0(ancestry_sub_path, path_constrained))
  
  # Find lines with "LKH"
  line_full <- grep("LKH", lines_full, value = TRUE)
  line_constrained <- grep("LKH", lines_constrained, value = TRUE)
  
  # Extract log-likelihoods
  LKH_full <- as.numeric(strsplit(line_full, "\\s+")[[1]][3])
  LKH_constrained <- as.numeric(strsplit(line_constrained, "\\s+")[[1]][3])
  
  # Likelihood Ratio Test
  LRT <- -2 * (LKH_constrained - LKH_full)
  pval <- pchisq(LRT, df = df, lower.tail = FALSE)
  
  interpretation <- ifelse(pval < 0.05,
                           "Full model significantly improves fit",
                           "No significant improvement from full model")
  
  # Create and return result tibble
  tibble(
    Outcome = Outcome,
    Comparison = Comparison,
    Type = "LRT",
    Constrained_LKH = LKH_constrained,
    Full_LKH = LKH_full,
    Value = LRT,
    P_value = pval,
    Interpretation = interpretation
  )
}
```

```{r extract_delta_output function}
extract_delta_output <- function(
  outcome,
  model_prefix,
  ancestry_sub_path,
  domain_labels = NULL,
  return_split = FALSE,
  domain_model = FALSE,
  domain_filename = NULL,
  sensitivity = FALSE,
  manual_correlation_labels = NULL  
) {
  # determine the file path
  file_path <- if (sensitivity) {
    paste0(ancestry_sub_path, "sensitivity/", model_prefix, "_reml_", outcome, ".out3")
  } else if (domain_model) {
    if (!is.null(domain_filename)) {
      paste0(ancestry_sub_path, "domains/", domain_filename)
    } else {
      paste0(ancestry_sub_path, "domains/", model_prefix, "_", outcome, ".out3")
    }
  } else {
    paste0(ancestry_sub_path, "oneE/", model_prefix, "_reml_", outcome, ".out3")
  }
  
  if (!file.exists(file_path)) stop("File not found: ", file_path)
  output_raw <- readLines(file_path)

  # variance components
  est_raw <- grep("Ratio:", output_raw, value = TRUE)
  est_split <- strsplit(est_raw, "\\s+")
  est_df <- do.call(rbind, lapply(est_split, function(x) {
    data.frame(
      parameter = "Variance",
      domain = NA_character_,
      variance_std = as.numeric(x[3]),
      SE = as.numeric(x[5]),
      P_value = as.numeric(x[7])
    )
  }))

  # assign domain labels 
  if (!is.null(domain_labels)) {
    if (length(domain_labels) == nrow(est_df)) {
      est_df$domain <- domain_labels
    } else {
      warning("Length of domain_labels does not match number of components. Assigning generic names.")
      est_df$domain <- paste0("Component_", seq_len(nrow(est_df)))
    }
  } else {
    est_df$domain <- paste0("Component_", seq_len(nrow(est_df)))
  }

  # covariance contribution (for e_g_cov models) 
  if (grepl("e_g_cov", model_prefix)) {
    domain_only_sum <- sum(est_df$variance_std, na.rm = TRUE)
    cov_contribution <- 1 - domain_only_sum
    cov_row <- data.frame(
      parameter = "Covariance",
      domain = "Covariance",
      variance_std = cov_contribution,
      SE = NA,
      P_value = NA
    )
    est_df <- rbind(est_df, cov_row)
  }

  # correlation estimates 
  cor_df <- NULL
  if (grepl("e_g_cov", model_prefix) || domain_model) {
    cor_raw <- grep("Cor  :", output_raw, value = TRUE)
    if (length(cor_raw) > 0) {
      cor_matrix <- do.call(rbind, lapply(cor_raw, function(line) {
        parts <- strsplit(trimws(line), "\\s+")[[1]]
        data.frame(
          correlation_r = as.numeric(parts[3]),
          SE = as.numeric(parts[5]),
          P_value = as.numeric(parts[7])
        )
      }))

      exclude_domains <- c("Covariance", "Error")
      exclude_patterns <- c("interaction", "gxe")
      correlatable_domains <- est_df$domain[
        !est_df$domain %in% exclude_domains &
          !grepl(paste0(exclude_patterns, collapse = "|"), est_df$domain, ignore.case = TRUE)
      ]

      if (!is.null(manual_correlation_labels)) {
        if (length(manual_correlation_labels) != nrow(cor_matrix)) {
          stop("Length of manual_correlation_labels does not match number of correlations in output.")
        }

        cor_df_raw <- purrr::map2_dfr(manual_correlation_labels, seq_along(manual_correlation_labels), function(label, i) {
          rho <- cor_matrix$correlation_r[i]
          se <- cor_matrix$SE[i]
          pval <- cor_matrix$P_value[i]

          vars <- unlist(strsplit(label, " ~ "))
          var1 <- vars[1]
          var2 <- vars[2]

          var1_val <- est_df$variance_std[est_df$domain == var1]
          var2_val <- est_df$variance_std[est_df$domain == var2]
          cov_val <- rho * sqrt(var1_val * var2_val)
          var_expl <- 2 * cov_val

          data.frame(
            parameter = label,
            domain = "Correlation",
            unscaled_variance = var_expl,
            SE = round(se, 4),
            P_value = round(pval, 4),
            correlation_r = round(rho, 4)
          )
        })
      } else {
        domain_pairs <- combn(correlatable_domains, 2, simplify = FALSE)

        if (length(domain_pairs) != nrow(cor_matrix)) {
          stop("Number of correlations does not match number of domain pairs. ",
               "Detected pairs: ", length(domain_pairs),
               ", correlations in file: ", nrow(cor_matrix),
               ". Likely due to uncorrelatable domain (e.g., GxE) being included.")
        }

        cor_df_raw <- purrr::map2_dfr(domain_pairs, seq_along(domain_pairs), function(pair, i) {
          var1 <- pair[1]
          var2 <- pair[2]
          rho <- cor_matrix$correlation_r[i]
          se <- cor_matrix$SE[i]
          pval <- cor_matrix$P_value[i]

          var1_val <- est_df$variance_std[est_df$domain == var1]
          var2_val <- est_df$variance_std[est_df$domain == var2]
          cov_val <- rho * sqrt(var1_val * var2_val)
          var_expl <- 2 * cov_val

          data.frame(
            parameter = paste0(var1, " ~ ", var2),
            domain = "Correlation",
            unscaled_variance = var_expl,
            SE = round(se, 4),
            P_value = round(pval, 4),
            correlation_r = round(rho, 4)
          )
        })
      }

      total_covariance_est <- est_df$variance_std[est_df$domain == "Covariance"]
      scale_factor <- total_covariance_est / sum(cor_df_raw$unscaled_variance, na.rm = TRUE)

      cor_df <- cor_df_raw %>%
        mutate(
          variance_std = round(unscaled_variance * scale_factor, 4)
        ) %>%
        select(parameter, domain, variance_std, SE, P_value, correlation_r)
    }
  }

  # combine all rows 
  if (!is.null(cor_df)) {
    cor_df$correlation_r <- cor_df$correlation_r
    full_df <- bind_rows(est_df, cor_df)
  } else {
    est_df$correlation_r <- NA_real_
    full_df <- est_df
  }

  # post-processing 
  full_df <- full_df %>%
    group_by(model = model_prefix, outcome) %>%
    mutate(
      CI_lower = if_else(!is.na(SE), variance_std - 1.96 * SE, NA_real_),
      CI_upper = if_else(!is.na(SE), variance_std + 1.96 * SE, NA_real_),
      raw_total = sum(variance_std[parameter %in% c("Variance", "Covariance") & domain != "Error"], na.rm = TRUE),
      variance_std_tot = if_else(parameter %in% c("Variance", "Covariance"), raw_total, NA_real_),
      across(c(variance_std, SE, CI_lower, CI_upper, variance_std_tot, P_value), ~ round(.x, 4)),
      input = case_when(
        domain_model & grepl("e_g_cov_domexe", model_prefix)                  ~ "Total environment and genetic, DxD and rD",
        domain_model & grepl("e_g_domexe", model_prefix)                      ~ "Total environment and genetic, DxD",
        domain_model & grepl("e_g_cov_health_all_reml_dropped", model_prefix) ~ "Environment and genetic reduced, rD",
        domain_model & grepl("e_g_cov_all_reml", model_prefix)                ~ "Environment and genetic, rD",
        domain_model & grepl("domain_reml_ge_health_dropped", model_prefix)   ~ "Environment domains reduced",
        domain_model & grepl("domain_reml_ge_health", model_prefix)           ~ "Environment domains and genetic",
        model_prefix == "g"                                                    ~ "Genetic",
        model_prefix == "e"                                                    ~ "Total environment",
        model_prefix == "e_g"                                                  ~ "Total environment and genetic",
        model_prefix == "e_g_cov"                                              ~ "Total environment and genetic, rGE",
        model_prefix == "e_g_gxe"                                              ~ "Total environment and genetic, GxE",
        model_prefix == "e_g_cov_gxe"                                          ~ "Total environment and genetic, rGE and GxE",
        TRUE                                                                   ~ "Other"
      ),
      flag_neg_variance = if_else(parameter %in% c("Variance", "Covariance") & variance_std < 0, TRUE, FALSE),
      flag_corr_out_of_bounds = if_else(domain == "Correlation" & 
                                          (!is.na(CI_lower) & (CI_lower < -1 | CI_upper > 1)), TRUE, FALSE)
    ) %>%
    select(-raw_total) %>%
    ungroup() %>%
    select(outcome, domain, input, variance_std, SE, CI_lower, CI_upper,
           variance_std_tot, model, parameter, P_value,
           flag_neg_variance, flag_corr_out_of_bounds, correlation_r)

  # optional split return 
  if (return_split) {
    list(
      variance_table = full_df %>% filter(domain != "Correlation"),
      correlation_table = full_df %>% filter(domain == "Correlation")
    )
  } else {
    return(full_df)
  }
}
```

## Theory-based environmental data

**Linear mixed models will be used to estimate the variance explained by known predictors of depression in Wave I of Add Health data collection.**

We will use linear mixed models to analyse the combined contribution of many environmental factors as demonstrated by Zhou and Lee (2021). As an initial step, we will compute linear mixed models (LMM) using only Wave I data. We will assess the variance in depression symptoms explained by (a restricted number of) theoretically-selected predictors of depression at Wave I. Predictors will be selected based on previous literature showing evidence of an association between the predictor and depression symptoms. Depression symptoms will be assessed using the CES-D (described above). Two depression sum scores will be calculated: The first with the full scale of depression symptoms included and the second a reduced set that is consistent with the items collected at Wave IV (omitting somatic symptoms).

* I have removed any variables that are considered depression symptoms (moodiness and frequent crying) as well as suicide attempts/treatment for suicide. 
* I have included positive and negative determinants here - usually the same variable on opposite ends of the scale e.g. H1PF35_numeric (feeling socially accepted). 

### Data cleaning

#### Recode

This recoding is only used when looking at the entire variable as one - not when looking at dummy variables. This is mainly done to use for factor analysis in stage G. 

I have recoded some items to get some descriptives for these variables going across mutliple items. 

Variables that need recoding:  H1GH16_numeric_recode - DONE, H1GH27I_numeric_recode - DONE, H1CO10F_numeric_recode - DONE, H1SU2_numeric_recode - DONE, H1SU3_numeric_recode - DONE, H1SU5_numeric_recode - DONE, H1SU7_numeric_recode - DONE, H1FP21_numeric_recode - DONE.  

```{r recoding for descriptives}
dat <- dat %>%
  mutate(PA55_clean = ifelse(PA55 == 9996, NA_real_, PA55))

# separate rape victim (female) from rape perpretrator (male) as the question is asked this way
dat <- dat %>%
  mutate(
    H1CO10F =
      ifelse(BIO_SEX_factor_clean == "Female" & H1CO10_factor_clean == "Yes",
             yes = 1,
             no = 0),
    H1CO10F_numeric = as.numeric(H1CO10F),
    H1CO10F_factor = recode_factor(as_factor(H1CO10F),
        "1" = "Yes",
        "0" = "No"),
    H1CO10M =
      ifelse(BIO_SEX_factor_clean == "Male" & H1CO10_factor_clean == "Yes",
             yes = 1,
             no = 0),
    H1CO10M_numeric = as.numeric(H1CO10M),
    H1CO10M_factor = recode_factor(as_factor(H1CO10M),
        "1" = "Yes",
        "0" = "No")
  )

dat <- dat %>%
  mutate(H1GH16_numeric_recode =  # frequency of period pains
           case_when(
             BIO_SEX_factor_clean == "Male" ~ 5, # does not apply to males - NOT missing due to nonresponse
             H1GH16_numeric_clean == 0 ~ 0,
             H1GH16_numeric_clean == 1 ~ 1,
             H1GH16_numeric_clean == 2 ~ 2,
             H1GH16_numeric_clean == 3 ~ 3,
             H1GH16_numeric_clean == 4 ~ 4
            )
  ) %>%
  mutate(
    H1GH16_factor_recode =
      recode_factor(as_factor(H1GH16_numeric_recode), # frequency of period pains
        "0" = "Never",
        "1" = "Just a few times",
        "2" = "About once a week",
        "3" = "Almost every day",
        "4" = "Every day",
        "5" = "Male"
        )
    ) %>%
  mutate(H1GH27I_numeric_recode =  # couldn't pay for healthcare
           case_when(
             H1GH26_numeric_clean == 0 ~ 2,  # they did not not seek healthcare in the first instance
             H1GH27I_numeric_clean == 0 ~ 0, # reason for not seeking healthcare - couldn't pay - No
             H1GH27I_numeric_clean == 1 ~ 1  # reason for not seeking healthcare - couldn't pay - Yes
           )
  ) %>%
  mutate(
    H1GH27I_factor_recode =        # couldn't pay for healthcare
      recode_factor(as_factor(H1GH27I_numeric_recode),
        "0" = "Other reason did not seeek healthcare",
        "1" = "Couldn't pay for healthcare",
        "2" = "Did not seek healthcare"
        )
    ) %>%
  mutate(H1CO10F_numeric_recode =  # experience of rape
           case_when(
             BIO_SEX_factor_clean == "Male" ~ 2, # all males
             H1CO10F_numeric == 1 ~ 1,     # have reported rape
             H1CO10F_numeric == 0 ~ 0,     # no rape
             H1CO1_numeric_clean == 1 ~ 3,       # sex but not answered rape Q
             H1CO1_numeric_clean == 0 ~ 4        # no sex and not answered rape Q
           )
  ) %>%
  mutate(
    H1CO10F_factor_recode =  # experience of rape
      recode_factor(as_factor(H1CO10F_numeric_recode),
        "0" = "Female: Not reported rape",
        "1" = "Female: reported rape",
        "2" = "Male",
        "3" = "Had sex but NA for rape Q",
        "4" = "Not had sex"
        )
  ) %>%
  mutate(H1SU2_numeric_recode =  # attempted suicide
           case_when(
             H1SU2_numeric_clean == 0 ~ 0,  # considered but not attempted
             H1SU2_numeric_clean == 1 ~ 1,  # attempted 1 time
             H1SU2_numeric_clean == 2 ~ 2,  # attempted 2 or 3 times
             H1SU2_numeric_clean == 3 ~ 3,  # attempted 4 or 5 times
             H1SU2_numeric_clean == 4 ~ 4,  # attempted 6 or more times
             H1SU1_numeric_clean == 1 ~ 5   # considered attempt but NA for attempt
           )
  ) %>%
  mutate(
    H1SU2_factor_recode =    # attempted suicide
      recode_factor(as_factor(H1SU2_numeric_recode),
        "0" = "Considered but not attempted",
        "1" = "Attempted 1 time",
        "2" = "Attempted 2 or 3 times",
        "3" = "Attempted 4 or 5 times",
        "4" = "Attempted 6 or more times",
        "5" = "Considered attempt but NA for attempt"
        )
  ) %>%
  mutate(H1SU3_numeric_recode =  # did any suicide attempt result in treatment from a doctor/nurse
           case_when(
             H1SU3_numeric_clean == 0 ~ 0,
             H1SU3_numeric_clean == 1 ~ 1,
             H1SU1_numeric_clean == 1 ~ 2,  # considered attempt but NA for treatment from doctor
             H1SU1_numeric_clean == 0 ~ 3   # did not report attempt
           )
  ) %>%
  mutate(
    H1SU3_factor_recode =    # did any suicide attempt result in treatment from a doctor/nurse
      recode_factor(as_factor(H1SU3_numeric_recode),
        "0" = "No",
        "1" = "Yes",
        "2" = "Considered attempt but NA for treatment for doctor",
        "3" = "Not considered"
        )
  ) %>%
  mutate(H1SU5_numeric_recode =  # friend succeeded suicide attempt
           case_when(
             H1SU5_numeric_clean == 0 ~ 0,
             H1SU5_numeric_clean == 1 ~ 1,
             H1SU4_numeric_clean == 1 ~ 2,  # reported friend attempt but NA for if they succeeded
             H1SU4_numeric_clean == 0 ~ 3   # did not report attempt by friend
           )
  ) %>%
  mutate(
    H1SU5_factor_recode =        # friend succeeded suicide attempt
      recode_factor(as_factor(H1SU5_numeric_recode),
        "0" = "No",
        "1" = "Yes",
        "2" = "Reported friend tried to kill themselves but NA for success",
        "3" = "No report of attempt by friend"
        )
  ) %>%
 mutate(H1SU7_numeric_recode =     # family succeeded suicide attempt
           case_when(
             H1SU7_numeric_clean == 0 ~ 0,
             H1SU7_numeric_clean == 1 ~ 1,
             H1SU6_numeric_clean == 1 ~ 2,  # reported family attempt but NA for if they succeeded
             H1SU6_numeric_clean == 0 ~ 3   # did not report attempt by friend
           )
  ) %>%
  mutate(
    H1SU7_factor_recode =        # family succeeded suicide attempt
      recode_factor(as_factor(H1SU7_numeric_recode),
        "0" = "No",
        "1" = "Yes",
        "2" = "Reported friend tried to kill themselves but NA for success",
        "3" = "No report of attempt by friend"
        )
  ) %>%
  mutate(H1FP21_numeric_recode = # experience of still birth/miscarriage/abortion/live birth - coded for first three pregnancies
           case_when(
            # H1FP21_1_numeric == 1 | H1FP21_2_numeric == 1 | H1FP21_3_numeric == 1 ~ 1, # any pregnancy pregnancy hasn't ended
            # H1FP21_1_numeric == 2 | H1FP21_2_numeric == 2 | H1FP21_3_numeric == 2 ~ 2, # any pregnancy live birth
             H1FP21_1_numeric_clean == 4 | H1FP21_2_numeric_clean == 4 | H1FP21_3_numeric_clean == 4 ~ 4, # any pregnancy still birth/ miscarriage
             H1FP21_1_numeric_clean == 5 | H1FP21_2_numeric_clean == 5 | H1FP21_3_numeric_clean == 5 ~ 5, # any pregnancy abortion
             H1FP7_numeric_clean == 1 ~ 6,           # reported being pregnant but not reported still birth/miscarriage/abortion
             H1FP7_numeric_clean == 0 ~ 7,           # had sex but never been pregnant
             BIO_SEX_factor_clean == "Male" ~ 8,     # male - not possible to be pregnant
             H1CO1_numeric_clean == 0 ~ 0            # female but never had sex (not asked this Q)
           )
  ) %>%
  mutate(
    H1FP21_factor_recode =    # experience of still birth/miscarriage/abortion/live birth - coded for first three pregnancies
      recode_factor(as_factor(H1FP21_numeric_recode),
        "4" = "Any pregnancy: still birth/miscarriage",
        "5" = "Any pregnancy: abortion",
        "6" = "Reported pregnancy but not still birth/miscarriage/abortion",
        "7" = "Had sex but never been pregnant",
        "8" = "Male: not possible to be pregnant",
        "0" = "Female: never had sex"
        )
  )
```

We will include the numeric versions of the categorical variables. This includes numeric coding for missing responses which includes more information. We will recode all variables to be binary dummy variables. 

#### Select

Select the variables for the ERM. 

```{r select theory driven ERM data}
dat.erm.theory1 <- dat %>%
  select(AID,
      # Demographics
      #   BIO_SEX_numeric_raw,      # biological sex (binary: male=1/female=2)
      #   race_recoded_numeric,     # race (categorical: native american/black/white/asian/other)
         PA55_clean,               # total family income before taxes - all incomes and sources (numeric: distribution spikes on tens) 
         PA56_numeric_raw,         # do you have enough money to pay your bills (binary: Yes=1, no=0)
         PA59_numeric_raw,         # how easy/hard is it to get medical care for your family (categorical ordered: very easy,somewhat easy, somewhat hard, very hard)
         H1NB5_numeric_raw,        # do you usually feel safe in your neighbourhood (binary: Yes=1, no=0)
         H1NB6_numeric_raw,        # how happy are you with living in your neighbourhood (categorical ordered: not at all, very little, somewhat, quite a bit, very much)
         H1GH27I_numeric_raw,      # # reason for not seeking healthcare - couldn't pay (categorical: No to H1GH26, Yes to couldn't pay, No to couldn't pay)
     #   H1GH27I_numeric_recode,   # reason for not seeking healthcare - couldn't pay (categorical: No to H1GH26, Yes to couldn't pay, No to couldn't pay)
        
      # Health 
         H1GH1_numeric_raw,        # general health (categorical ordered: excellent, very good, good, fair, poor)
         H1PL1_numeric_raw,        # permanent physical condition (binary: Yes=1, no=0)
         H1GH2_numeric_raw,        # frequency of headache - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH3_numeric_raw,        # frequency of feeling hot - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH4_numeric_raw,        # frequency of stomach ache - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH5_numeric_raw,        # frequency of cold sweats - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH6_numeric_raw,        # frequency of feeling weak - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH7_numeric_raw,        # frequency of sore throat - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH8_numeric_raw,        # frequency of fatigue - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH9_numeric_raw,        # frequency of painful urination - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH10_numeric_raw,       # frequency of feeling really sick - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH11_numeric_raw,       # frequency of waking up tired - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH12_numeric_raw,       # frequency of skin problems - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH13_numeric_raw,       # frequency of dizziness - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH14_numeric_raw,       # frequency of chest pains - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH15_numeric_raw,       # frequency of aches and pains - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH16_numeric_raw,       # frequency of period pains - past 12 months - (categorical ordered: Never, just a few times, once a week, almost every day, every day)
     #   H1GH16_numeric_recode,    # frequency of period pains - past 12 months - (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH17_numeric_raw,       # frequency of poor appetite - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH18_numeric_raw,       # frequency of sleep problems - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH19_numeric_raw,       # frequency of trouble relaxing - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
     #   H1GH20_numeric_raw,       # frequency of moodiness - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
     #   H1GH21_numeric_raw,       # frequency of frequent crying - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH22_numeric_raw,       # frequency of fearfulness - past 12 months (categorical ordered: Never, just a few times, once a week, almost every day, every day)
         H1GH26_numeric_raw,       # thought needed medical care but didn't seek it - past year (binary: Yes=1, No=0)
         H1GH28_numeric_raw,       # how do you think of yourself in weight (categorical ordered: 1=very underweight, slightly underweight, about right, slightly overweight, 5=very overweight)
         H1GH29_numeric_raw,       # trying to change weight (categorical: lose weight, gain weight, stay the same weight, not trying to do anything about weight)
         H1HS3_numeric_raw,        # psychological counselling in past year (binary: yes=1, no=0)
         H1GH48_numeric_raw,       # health or emotional problem caused day off school (categorical frequency: Never, a few times, once a week, almost every day, every day)
         H1GH49_numeric_raw,       # health or emotional problem caused miss recreational activity (categorical frequency: Never, a few times, once a week, almost every day, every day)
         
     # Difficulties at school 
         H1ED15_numeric_raw,       # trouble getting along with teachers - past year (categorical frequency: Never, a few times, once a week, almost every day, every day)
         H1ED16_numeric_raw,       # trouble paying attention in school - past year (categorical frequency: Never, a few times, once a week, almost every day, every day)
         H1ED17_numeric_raw,       # trouble getting homework done - past year (categorical frequency: Never, a few times, once a week, almost every day, every day)
         H1ED18_numeric_raw,       # trouble getting along with other students - past year (categorical frequency: Never, a few times, once a week, almost every day, every day)
         H1ED19_numeric_raw,       # feel close to people at school - past year (categorical frequency: Never, a few times, once a week, almost every day, every day)
         H1ED20_numeric_raw,       # feel like you are part of your school - past year  (categorical frequency: Never, a few times, once a week, almost every day, every day)
         H1ED21_numeric_raw,       # students at school are prejudiced - past year (categorical frequency: Never, a few times, once a week, almost every day, every day)
         H1ED22_numeric_raw,       # happy to be at your school - past year (categorical frequency: Never, a few times, once a week, almost every day, every day)
         H1ED23_numeric_raw,       # teachers treat students fairly - past year (categorical frequency: Never, a few times, once a week, almost every day, every day)
         H1ED24_numeric_raw,       # feel safe in your school - past year (categorical frequency: Never, a few times, once a week, almost every day, every day)
         
     # Relationships with friends and family
         H1DA7_numeric_raw,        # frequency hanging out with friends (categorical frequency: not at all, 1 or 2 times, 3 or 4 times, 5 or more times)
         H1WP9_numeric_raw,        # how close do you feel to your mother (categorical frequency: not at all, very little, somewhat, quite a bit, very much)
         H1WP10_numeric_raw,       # how much do you think she cares about you (categorical frequency: not at all, very little, somewhat, quite a bit, very much)
         H1WP13_numeric_raw,       # how close to you feel to your father - lots missing. (categorical frequency: not at all, very little, somewhat, quite a bit, very much)
         H1WP14_numeric_raw,       # how much to you think he cares about you - lots missing. (categorical frequency: not at all, very little, somewhat, quite a bit, very much)
         H1PF1_numeric_raw,        # your mother is warm and loving to you (categorical frequency: strongly agree, agree, neither agree or disagree, disagree, strongly disagree)
         H1PF4_numeric_raw,        # satisfied with the way you and your mother communicate (categorical frequency: strongly agree, agree, neither agree or disagree, disagree, strongly disagree)
         H1PF5_numeric_raw,        # satisfied with your relationships with your mother (categorical frequency: strongly agree, agree, neither agree or disagree, disagree, strongly disagree)
         H1PF24_numeric_raw,       # satisfied with the way you and your father communicate - lots missing. (categorical frequency: strongly agree, agree, neither agree or disagree, disagree, strongly disagree)
         H1PF25_numeric_raw,       # satisfied with your relationship with your father - lots missing. (categorical frequency: strongly agree, agree, neither agree or disagree, disagree, strongly disagree)
         H1PR1_numeric_raw,        # feel that adults care about you (categorical ordered: not at all, very little, somewhat, quite a bit, very much, does not apply)
         H1PR2_numeric_raw,        # feel your teachers care about you (categorical ordered: not at all, very little, somewhat, quite a bit, very much, does not apply)
         H1PR3_numeric_raw,        # feel your parents care about you (categorical ordered: not at all, very little, somewhat, quite a bit, very much, does not apply)
         H1PR4_numeric_raw,        # feel your friends care about you (categorical ordered: not at all, very little, somewhat, quite a bit, very much, does not apply)
         H1PR5_numeric_raw,        # feel that people in your family home understand you (categorical ordered: not at all, very little, somewhat, quite a bit, very much, does not apply)
         H1PR6_numeric_raw,        # feel you want to leave home (categorical ordered: not at all, very little, somewhat, quite a bit, very much, does not apply)
         H1PR7_numeric_raw,        # feel that you and your family have fun together (categorical ordered: not at all, very little, somewhat, quite a bit, very much, does not apply)
         H1PR8_numeric_raw,        # feel that your family pays attention to you (categorical ordered: not at all, very little, somewhat, quite a bit, very much, does not apply)
         
     # Positive cognition and social acceptance     
         H1PF33_numeric_raw,       # you like yourself (categorical frequency: strongly agree, agree, neither agree or disagree, disagree, strongly disagree) - strange ordering 
         H1PF34_numeric_raw,       # you feel like you are doing everything just about right (categorical frequency: strongly agree, agree, neither agree or disagree, disagree, strongly disagree)
         H1PF35_numeric_raw,       # you feel socially accepted (categorical frequency: strongly agree, agree, neither agree or disagree, disagree, strongly disagree)
         H1PF36_numeric_raw,       # you feel loved and wanted (categorical frequency: strongly agree, agree, neither agree or disagree, disagree, strongly disagree)

         # Stressful life events / traumatic experiences
         H1CO10_numeric_raw,       # been raped - asked about being raped and raped someone 
      #  H1CO10F_numeric_recode,   # been raped - female only. (categorical: 0=not raped and female, 1= raped and female, 2=Male and not possible to be raped). 
         H1NR3_numeric_raw,        # given someone sex in exchange for drugs or money (binary: Yes=1, No=0)
         H1FV1_numeric_raw,        # saw someone shoot or stab another person - past year (categorical frequency: never, once, more than once)
         H1FV2_numeric_raw,        # someone pulled a knife or gun on you - past year (categorical frequency: never, once, more than once)
         H1FV3_numeric_raw,        # someone shot you - past year (categorical frequency: never, once, more than once)
         H1FV4_numeric_raw,        # someone cut or stabbed you - past year (categorical frequency: never, once, more than once)
         H1FV5_numeric_raw,        # you got into a physical fight - past year (categorical frequency: never, once, more than once)
         H1FV6_numeric_raw,        # you were jumped - past year (categorical frequency: never, once, more than once)
      #  H1FP21_numeric_recode,    # ever experienced abortion/miscarriage (categorical, still birth, abortion, pregnant but not mis/abort, had sex but not pregnant, male, never had sex (female))
         H1FP21_1_numeric_raw,     # how did pregnancy 1 end (categorical: abortion, live birth, hasn't ended, still birth/miscarriage)
         H1FP21_2_numeric_raw,     # how did pregnancy 2 end (categorical: abortion, live birth, hasn't ended, still birth/miscarriage)
         H1FP21_3_numeric_raw,     # how did pregnancy 3 end (categorical: abortion, live birth, hasn't ended, still birth/miscarriage)
      #   H1SU1_numeric_raw,        # did you ever seriously think about suicide - past 12 months (binary: Yes=1, No=0)
      #  H1SU2_numeric_recode,     # how many times did you attempt suicide - past 12 months (categorical ordered: 0 times, 1 time, 2 or 3 times, 4 or 5 times, 6 or more times, attempt but NA to this Q)
      #  H1SU3_numeric_recode,     # did any attempt result in treatment from a doctor/nurse (categorical: yes, no, reported attempt but NA for treatment from doctor)
         H1SU4_numeric_raw,        # have any of your friends tried to kill themselves - past 12 months (binary: Yes=1, No=0)
         H1SU5_numeric_raw,        # have any of them succeeded 
      #   H1SU5_numeric_recode,    # have any of them succeeded - combined with H1SU4_numeric - (categorical: yes, no, reported friend attempt but NA outcome, no attempt)
         H1SU6_numeric_raw,        # have any of your family members tried to kill themselves - past 12 months (binary: Yes=1, No=0)
         H1SU7_numeric_raw         # have any suceeded
       #  H1SU7_numeric_recode     # have any of them succeeded - combined with H1SU4_numeric - (categorical: yes, no, reported family attempt but NA outcome, no attempt)
         )

length(colnames(dat.erm.theory1))
nrow(dat.erm.theory1)
head(dat.erm.theory1)
```

#### Missingness

We cannot compute the matrix unless we have full data. We will need to include as much of the data as possible. Therefore, we ahve kept in the codes that indicate differential missing data. 

For example: 
* "Legitimate skip (7)", "Refused (7)", "Don't know (8)", _numeric variables to actively code in the matrices - these will then be made into dummy variables. 

We will identify variables that have high missingness to consider imputation options (random sample imputation)

Find variables with substantial missingness (>100 missing responses):

```{r identify variables with high missingness}
threshold <- 1 # set threshold

# dat.erm.theory1
find_vars_with_na(dat.erm.theory1, threshold)
```

Variables that are parent report (SES related variables) have highest missingness as some parents did not complete the questionnaire. 
PA55 - annual income (continuous) - measured in thousands - 9996 is coded as refusal of question
PA56_numeric - have enough money to pay bills 
PA59_numeric - easy to get medical care 

##### Random sample imputation for PA55, PA56 and PA59

We will randomly impute the PA55_clean variable so that we don't lose so many people in the whole matrix. 

```{r random imputation for PA55_clean}
# Create new var
dat.erm.theory1 <- dat.erm.theory1 %>%
  mutate(PA55_clean_impute = random_sample_imputation(PA55_clean)) %>%
  mutate(PA56_numeric_raw_impute = random_sample_imputation(PA56_numeric_raw)) %>%
  mutate(PA59_numeric_raw_impute = random_sample_imputation(PA59_numeric_raw))
  
# Check 
freq(dat.erm.theory1$PA55_clean)
freq(dat.erm.theory1$PA55_clean_impute)

freq(dat.erm.theory1$PA56_numeric_raw)
freq(dat.erm.theory1$PA56_numeric_raw_impute)

freq(dat.erm.theory1$PA59_numeric_raw)
freq(dat.erm.theory1$PA59_numeric_raw_impute)

# remove non-imputed variable 
dat.erm.theory1 <- dat.erm.theory1 %>% 
  select(-PA55_clean, -PA56_numeric_raw, -PA59_numeric_raw)
```

```{r parent questionnaire data missing}
# # in the original data - how many people did not have any parent level data
# count_na_all_P(dat)
```

2750 have missing parent level data - hashed out as takes a few mins to run. 

```{r filter out individuals who are misisng over 90% of the variables}
threshold_percentage <- 80 # Adjust the threshold percentage as needed
dat.erm.theory1_filtered <- drop_individuals_na_over_percentage(dat.erm.theory1, threshold_percentage)

# number of individuals dropped 
nrow(dat.erm.theory1) - nrow(dat.erm.theory1_filtered)
```

For now we will drop the people that have missing data on the selected variables - but will need to use imputation/include missingness as an option to save more of the data points. 
The more data we include, the more people we lose as we need full data. 

```{r missingness of dat.erm.theory1}
nrow(dat.erm.theory1) # 20728 people in original data 
dat.erm.theory1_clean <- dat.erm.theory1 %>% drop_na() # drop all missing values to see difference in N
nrow(dat.erm.theory1_clean) # 20727 people
nrow(dat.erm.theory1) - nrow(dat.erm.theory1_clean) # dropped 1 person
ncol(dat.erm.theory1_clean) # 82 variables + ID variable

# set everything to be numeric - comes in handy later for ordering ID files
dat.erm.theory1_clean <- dat.erm.theory1_clean %>% mutate(across(everything(), as.numeric))
head(dat.erm.theory1_clean)
```

#### Dummy recoding

Create dummy variables for the categorical variables with more than two categories. 

The reference categories below are considered what is "healthy" or "typical". 

```{r reference categories for dummy variables}
# set the reference categories for each dummy coded variable - default is first category if not included
reference_category <- list(# demographics
                          # "race_recoded_numeric" = 1,     # 1 = white, also default first category, largest 
                           "PA56_numeric_raw_impute" = 1,  # 1 = yes money for bills, largest 
                           "PA59_numeric_raw_impute" = 1,  # 1 = easy to get medical care, largest 
                           "H1NB5_numeric_raw" = 1,        # 1 = yes safe in neighborhood, largest 
                           "H1NB6_numeric_raw" = 5,        # 5 = very happy in neighborhood, second largest 
                           # health
                           "H1GH1_numeric_raw" = 1,        # 1 = excellent general health, second largest 
                           "H1PL1_numeric_raw" = 0,        # 0 = no physical condition, largest 
                           "H1GH2_numeric_raw" = 0,        # 0 = never headache, third largest 
                           "H1GH3_numeric_raw" = 0,        # 0 = never feeling hot, largest 
                           "H1GH4_numeric_raw" = 0,        # 0 = never stomach ache, third largest 
                           "H1GH5_numeric_raw" = 0,        # 0 = never cold sweats, largest 
                           "H1GH6_numeric_raw" = 0,        # 0 = never feeling weak, largest 
                           "H1GH7_numeric_raw" = 0,        # 0 = never sore throat, second largest 
                           "H1GH8_numeric_raw" = 0,        # 0 = never fatigue, third largest 
                           "H1GH9_numeric_raw" = 0,        # 0 = never painful urination, largest 
                           "H1GH10_numeric_raw" = 0,       # 0 = never feeling really sick, largest 
                           "H1GH11_numeric_raw" = 0,       # 0 = never waking up tired, third largest 
                           "H1GH12_numeric_raw" = 0,       # 0 = never skin problems, second largest 
                           "H1GH13_numeric_raw" = 0,       # 0 = never dizziness, largest 
                           "H1GH14_numeric_raw" = 0,       # 0 = never chest pains, largest 
                           "H1GH15_numeric_raw" = 0,       # 0 = never aches and pains, second largest 
                           "H1GH16_numeric_raw" = 0,       # 0 = never period pains, second largest (after being male)
                           "H1GH17_numeric_raw" = 0,       # 0 = never poor appetite, largest
                           "H1GH18_numeric_raw" = 0,       # 0 = never sleep problems, largest
                           "H1GH19_numeric_raw" = 0,       # 0 = never trouble relaxing, largest
                           # "H1GH20_numeric_raw" = 0,     # 0 = never moodiness, third largest
                           # "H1GH21_numeric_raw" = 0,     # 0 = never frequent crying, largest
                           "H1GH22_numeric_raw" = 0,       # 0 = never fearfulness, largest
                           "H1GH26_numeric_raw" = 0,       # 0 = no didn't seek med care, largest
                           "H1GH27I_numeric_raw" = 0,      # 0 = no to couldn't pay, largest after leg skip
                           "H1GH28_numeric_raw" = 3,       # 0 = about right weight, largest
                           "H1GH29_numeric_raw" = 3,       # 0 = stay the same weight, largest
                           "H1HS3_numeric_raw" = 0,        # 0 = not had counselling, largest - *usure on this reference cat*
                           "H1GH48_numeric_raw" = 0,       # 0 = never health/emotion problem day off, largest
                           "H1GH49_numeric_raw" = 0,       # 0 = never health/emotion problem miss activity, largest
                           # difficulties at school
                           "H1ED15_numeric_raw" = 0,       # 0 = never trouble getting along w teachers, second largest
                           "H1ED16_numeric_raw" = 0,       # 0 = never trouble paying attention, second largest 
                           "H1ED17_numeric_raw" = 0,       # 0 = never trouble doing homework, second largest
                           "H1ED18_numeric_raw" = 0,       # 0 = never trouble getting along w students, second largest
                           "H1ED19_numeric_raw" = 5,       # 5 = every day close to people, smallest
                           "H1ED20_numeric_raw" = 5,       # 5 = every day part of school, smallest
                           "H1ED21_numeric_raw" = 0,       # 0 = never prejudice, fourth largest 
                           "H1ED22_numeric_raw" = 5,       # 5 = every day happy to be at school, second largest
                           "H1ED23_numeric_raw" = 5,       # 5 = every day treat students fairly, third largest
                           "H1ED24_numeric_raw" = 5,       # 5 = every day feel safe in school, second largest
                           # friends and family 
                           "H1DA7_numeric_raw" = 4,        # 4 = 5 or more times see friends, smallest
                           "H1WP9_numeric_raw" = 5,        # 5 = very close to mother, largest
                           "H1WP10_numeric_raw" = 5,       # 5 = very much mother cares, largest
                           "H1WP13_numeric_raw" = 5,       # 5 = very close to father, largest
                           "H1WP14_numeric_raw" = 5,       # 5 = very much father cares, largest
                           "H1PF1_numeric_raw" = 1,        # 1 = strong agree mother is warm, largest 
                           "H1PF4_numeric_raw" = 1,        # 1 = strong agree satisfied w mother communication, second largest
                           "H1PF5_numeric_raw" = 1,        # 1 = strong agree satisfied w mother relationship, largest 
                           "H1PF24_numeric_raw" = 1,       # 1 = strong agree satisfied w father communication, third largest 
                           "H1PF25_numeric_raw" = 1,       # 1 = strong agree satisfied w father relationship, largest 
                           "H1PR1_numeric_raw" = 5,        # 5 = very much feel adults care, largest 
                           "H1PR2_numeric_raw" = 5,        # 5 = very much feel teachers care, third largest 
                           "H1PR3_numeric_raw" = 5,        # 5 = very much feel parents care, largest 
                           "H1PR4_numeric_raw" = 5,        # 5 = very much feel friends care, largest 
                           "H1PR5_numeric_raw" = 5,        # 5 = very much think family understand you, third largest
                           "H1PR6_numeric_raw" = 1,        # 1 = not at all want to leave home, largest
                           "H1PR7_numeric_raw" = 5,        # 5 = very much family have fun, second largest
                           "H1PR8_numeric_raw" = 5,        # 5 = very much family pays attention, second largest
                           # positive cognition and social acceptance
                           "H1PF33_numeric_raw" = 1,       # 1 = strong agree you like yourself, second largest
                           "H1PF34_numeric_raw" = 1,       # 1 = strong agree doing everything right, third largest 
                           "H1PF35_numeric_raw" = 1,       # 1 = strong agree feel socially accepted, second largest 
                           "H1PF36_numeric_raw" = 1,       # 1 = strong agree feel loved, second largest
                           # sle
                           "H1CO10_numeric_raw" = 0,       # 0 = not raped, second largest after male
                           "H1NR3_numeric_raw" = 0,        # 0 = not exchanged sex, largest 
                           "H1FV1_numeric_raw" = 0,        # 0 = never seen stabbing, largest
                           "H1FV2_numeric_raw" = 0,        # 0 = never had knife pulled, largest
                           "H1FV3_numeric_raw" = 0,        # 0 = never been shot, largest
                           "H1FV4_numeric_raw" = 0,        # 0 = never been stabbed, largest
                           "H1FV5_numeric_raw" = 0,        # 0 = never in physical fight, largest
                           "H1FV6_numeric_raw" = 0,        # 0 = never been jumped, largest
                         # "H1FP21_numeric_recode" = 6,    # 6 = pregnant but not abortion, fourth largest after male
                           "H1FP21_1_numeric_raw" = 7,     # 7 = legitimate skip - no pregnancy 
                           "H1FP21_2_numeric_raw" = 7,     # 7 = legitimate skip - no pregnancy 
                           "H1FP21_3_numeric_raw" = 7,     # 7 = legitimate skip - no pregnancy 
                           "H1SU4_numeric_raw" = 0,        # 0 = friend tried to kill themselves
                           "H1SU5_numeric_raw" = 0,        # 0 = friend didn't succeed suicide
                           "H1SU6_numeric_raw" = 0,        # 0 = family not tried to kill themselves
                           "H1SU7_numeric_raw" = 0         # 0 = family didn't succeed
                           )

# Create new dataset with dummy variables
dat.erm.theory1.dummy <- create_dummy_variables(dat.erm.theory1_clean, reference_category)

# Check the names of the variables
length(colnames(dat.erm.theory1.dummy))
```

#### PCA transformation

To reduce collinearity in the data - eigenvectors are extracted from the covariance matrix and multiplied to tbe original E data. 

The paper for justification on this is here: https://www.nature.com/articles/s41598-021-00427-y 

Eigenvectors represent the directions of the new feature space (principal components), while eigenvalues indicate the amount of variance in the data along these directions.

```{r PCA transformation}
# create data as matrix without ID column 
head(dat.erm.theory1.dummy)
dat.erm.theory1.dummy_noID <- as.matrix(dat.erm.theory1.dummy[,-1]) 
head(as.data.frame(dat.erm.theory1.dummy_noID))

# covariance matrix of all the variables 
dat.erm.theory1.dummy_covmat <- var(dat.erm.theory1.dummy_noID)  

# PCA transformation
dat.erm.theory1.dummy_eig <- eigen(dat.erm.theory1.dummy_covmat) # extract eigen values and vectors 
is.matrix(dat.erm.theory1.dummy_eig$vectors) # check this 

dat.erm.theory1.dummy_pca_vector <- dat.erm.theory1.dummy_noID%*%dat.erm.theory1.dummy_eig$vectors # multiply all exposures by the eigenvectors
dat.erm.theory1.dummy_pca_scaled <- scale(dat.erm.theory1.dummy_pca_vector) # normalise this matrix - minus the mean and divide by SD 

m <- dim(dat.erm.theory1.dummy_pca_scaled)[2] # m is the number of variables 

# divide each element by sqrt(m)
dat.erm.theory1.dummy_pca <- dat.erm.theory1.dummy_pca_scaled/sqrt(m)
head(as.data.frame(dat.erm.theory1.dummy_pca))

# attach AID variable
dat.erm.theory1.dummy_pca <- as.data.frame(cbind(dat.erm.theory1.dummy$AID, dat.erm.theory1.dummy_pca)) # add ID variable

# rename ID variable and make all vars numeric 
dat.erm.theory1.dummy_pca <- dat.erm.theory1.dummy_pca %>% dplyr::rename(AID = V1) %>% mutate(across(everything(), as.numeric))

# have a look
head(dat.erm.theory1.dummy_pca)
dim(dat.erm.theory1.dummy_pca)
nrow(dat.erm.theory1.dummy_pca)
ncol(dat.erm.theory1.dummy_pca)

# drop any empty variables
dat.erm.theory1.dummy_pca <- dat.erm.theory1.dummy_pca %>% select(where(~ any(!is.na(.))))
```

## Depression outcome 

### Wave 1

#### Sum score W1A 

All questions asked about in the last week - from Center for Epidemiologic Studies Depression Scale (CES-D): 

* H1FS1 - You were bothered by things that usually don’t bother you
* H1FS2 - You didn’t feel like eating, your appetite was poor
* H1FS3 - You felt that you could not shake off the blues, even with help from your family and your friends
* H1FS4 - You felt that you were just as good as other people (*needs reverse coding*)
* H1FS5 - You had trouble keeping your mind on what you were doing
* H1FS6 - You felt depressed
* H1FS7 - You felt you were too tired to do things
* H1FS8 - You felt hopeful about the future (*needs reverse coding*)
* H1FS9 - You thought your life had been a failure
* H1FS10 - You felt fearful 
* **Missing Q from revised CES-D here "Q11: My sleep was restless"**
* H1FS11 - You were happy (*needs reverse coding*)
* H1FS12 - You talked less than usual 
* H1FS13 - You felt lonely 
* H1FS14 - People were unfriendly to you
* H1FS15 - You enjoyed life (*needs reverse coding*)
* **Missing Q from revised CES-D here "Q17: I had crying spells"**
* H1FS16 - You felt sad
* H1FS17 - You felt that people disliked you
* H1FS18 - It was hard to get started doing things
* H1FS19 - You felt life was not worth living **Additional Q from revised CES-D**

Sum score A for depression at wave one includes the entire feelings scale included in the wave one questionnaire. Question on sleep problems and crying spells is missing compared to the online copy of the CES-D, and includes an item on suicidal thoughts. 

```{r depression items W1A}
CESD.W1A.items <- c("H1FS1_numeric_clean",          # bothered 
                   "H1FS2_numeric_clean",           # poor appetite
                   "H1FS3_numeric_clean",           # couldn't shake the blues
                   "H1FS4_numeric_clean",           # just as good as other people  - reversed
                   "H1FS5_numeric_clean",           # trouble paying attention
                   "H1FS6_numeric_clean",           # felt depressed
                   "H1FS7_numeric_clean",           # too tired to do things
                   "H1FS8_numeric_clean",           # hopeful about future          - reversed
                   "H1FS9_numeric_clean",           # life had been a failure
                   "H1FS10_numeric_clean",          # felt fearful 
                   "H1FS11_numeric_clean",          # happy                         - reversed
                   "H1FS12_numeric_clean",          # talked less than usual
                   "H1FS13_numeric_clean",          # felt lonely
                   "H1FS14_numeric_clean",          # people were unfriendly
                   "H1FS15_numeric_clean",          # enjoyed life                  - reversed 
                   "H1FS16_numeric_clean",          # felt sad 
                   "H1FS17_numeric_clean",          # felt people disliked you
                   "H1FS18_numeric_clean",          # hard to get started
                   "H1FS19_numeric_clean"           # life was not worth living
         ) 

CESD.items.key.W1A <- c(  # scoring key: -1 is reverse coded
                      1, # 1
                      1, # 2
                      1,  # 3
                     -1, # 4
                      1, # 5 
                      1, # 6
                      1, # 7 
                     -1, # 8
                      1, # 9
                      1, # 10
                     -1, # 11
                      1, # 12
                      1, # 13
                      1, # 14
                     -1, # 15
                      1, # 16
                      1, # 17
                      1, # 18
                      1  # 19
                     ) 
```

```{r depression sum score W1A}
# sum score for CES-D depression W1A
CESD.scored.items.W1A <- scoreItems(keys = CESD.items.key.W1A,
                                items = dat[CESD.W1A.items],
                                totals = TRUE,    # calculate totals
                                missing = TRUE,   # keep missing items
                                impute = 'none',  # don't impute missing values
                                min = 0, 
                                max = 3)

# add score to data frame
dat <- dat %>%
  mutate(CESD_W1A = as.numeric(CESD.scored.items.W1A$scores))

# check 
freq(dat$CESD_W1A)
```

##### Missingness

```{r loop for missingness of depression items}
items.missing <- c()

for (i in CESD.W1A.items) {
  
  new_varname <- paste0(i, "_missing")
  
  dat <- dat %>%
     mutate(!!new_varname :=
           if_else(
               !is.na(!!sym(i)),     # !! unquotes the variable and tells it to use as object not string 
               true = "Not missing",
               false = "Missing"
           ))
  
  items.missing <- c(items.missing, new_varname) # creates a list of these variables 
}

# rename the general list to W1A specific
CESD.W1A.items.missing <- items.missing
```

```{r output for missingness CESD.W1A}
CESD.W1A.missing.freq <- map_dfr(CESD.W1A.items.missing, ~ freq(dat[[.]]) %>% head(1), .id = "Item")
row.names(CESD.W1A.missing.freq) <- NULL

# number of missing for each item 
CESD.W1A.missing.freq
```


```{r number of items that are missing CESD.W1A}
dat <- dat %>%
  mutate(total_missing_CESD.W1A = 
           rowSums(select(., CESD.W1A.items.missing) == "Missing")) # sums the number of "missing"s for each person 

freq(dat$total_missing_CESD.W1A) # number of items individuals have missing
```

##### Create outcome pheno dataset W1A

* Scores preadjusted, i.e. taking residuals of a regression model, for the following variables: year of birth, sex, and top 10 ancestry-specific genetic PCs
* Standardise scores to mean 0 and SD 1
* Select ID and sum score variables
* Remove any individuals that have more than one item (<2) missing

*ANCESTRY PCS HAVE BEEN INCLUDED IN THE OUTCOME WHEN SUBSETTING THE DATA INTO THE SMALLER SAMPLE*

```{r regress out covars W1A}
dat$CESD_W1A_resid <- resid(
  lm(
    CESD_W1A ~ H1GI1Y_factor_clean + BIO_SEX_factor_clean,
    data = dat
    )
  )

# check 
freq(dat$CESD_W1A_resid)
```

```{r standardise W1A}
# standardise
dat <- dat %>% 
  mutate(CESD_W1A_stan = as.numeric(scale(CESD_W1A_resid)))
  
# check 
freq(dat$CESD_W1A_stan)
```

```{r remove individuals with missing depression items into dat.clean.W1A}
dat.clean.W1A <- dat %>%
  filter(total_missing_CESD.W1A < 2) %>%
  select(AID,     # ID
         CESD_W1A_stan # depression sum score W1A - residualised and standardised
         )

# make ID numeric 
pheno_W1A <- dat.clean.W1A %>% mutate(AID = as.numeric(AID))

psych::alpha(dat[,CESD.W1A.items])

freq(dat$total_missing_CESD.W1A) 
nrow(dat.clean.W1A) # this gives us 20656 people 
head(dat.clean.W1A)
length(dat.clean.W1A$AID) == length(unique(dat.clean.W1A$AID))

# hist(dat$CESD_W1A) 
bar_CESD_W1A <- ggplot(dat.clean.W1A, aes(x = CESD_W1A_stan)) +
                  geom_histogram(binwidth = 0.2, fill = "blue", color = "black", alpha = 0.7) +
                  labs(title = "Histogram of CESD_W1A Values", x = "Value", y = "Frequency") +
                  theme_minimal()

save_gg_cairo(plot = bar_CESD_W1A, filename = "/home/thom1336/findmedep/figures/bar_CESD_W1A.png")
```

#### Sum score W1B 

The items that map onto the wave IV items. 

* H1FS1 - You were bothered by things that usually don't bother you (Q1)
* H1FS3 - You could not shake off the blues, even with help from your family and your friends (Q3)
* H1FS4 - You felt you were just as good as other people (Q4) (*needs reverse coding*)
* H1FS5 - You had trouble keeping your mind on what you were doing (Q5)
* H1FS6 - You felt depressed (Q6)
* H1FS7 - You felt that you were too tired to do things (Q7? "I felt that everything I did was an effort")
* H1FS11 - You felt happy (*needs reverse coding*) (Q11)
* H1FS15 - You enjoyed life (*needs reverse coding*) (Q15)
* H1FS16 - You felt sad (Q16)
* H1FS17 - You felt that people disliked you, during the past seven days (Q17)

```{r depression items W1B}
CESD.W1B.items <- c("H1FS1_numeric_clean",          # bothered 
                   "H1FS3_numeric_clean",           # couldn't shake the blues
                   "H1FS4_numeric_clean",           # just as good as other people  - reversed
                   "H1FS5_numeric_clean",           # trouble paying attention
                   "H1FS6_numeric_clean",           # felt depressed
                   "H1FS7_numeric_clean",           # too tired to do things
                   "H1FS11_numeric_clean",          # happy                       - reversed
                   "H1FS15_numeric_clean",          # enjoyed life               - reversed 
                   "H1FS16_numeric_clean",          # felt sad 
                   "H1FS17_numeric_clean"           # felt people disliked you
         ) 

CESD.items.key.W1B <- c(1, # scoring key: -1 is reverse coded
                      1, 
                     -1, 
                      1, 
                      1, 
                      1, 
                     -1, 
                     -1, 
                      1, 
                      1) 
```

```{r depression sum score W1B}
# sum score for CES-D depression W1B
CESD.scored.items.W1B <- scoreItems(keys = CESD.items.key.W1B,
                                items = dat[CESD.W1B.items],
                                totals = TRUE,    # calculate totals
                                missing = TRUE,   # keep missing items
                                impute = 'none',  # don't impute missing values
                                min = 0, 
                                max = 3)

# add score to data frame
dat <- dat %>%
  mutate(CESD_W1B = as.numeric(CESD.scored.items.W1B$scores))
```

###### Missingness

```{r loop for missingness of depression items}
items.missing <- c()

for (i in CESD.W1B.items) {
  
  new_varname <- paste0(i, "_missing")
  
  dat <- dat %>%
     mutate(!!new_varname :=
           if_else(
               !is.na(!!sym(i)),     # !! unquotes the variable and tells it to use as object not string 
               true = "Not missing",
               false = "Missing"
           ))
  
  items.missing <- c(items.missing, new_varname) # creates a list of these variables 
}

# rename the general list to W1B specific
CESD.W1B.items.missing <- items.missing
```

```{r output for missingness CESD.W1B}
CESD.W1B.missing.freq <- map_dfr(CESD.W1B.items.missing, ~ freq(dat[[.]]) %>% head(1), .id = "Item")
row.names(CESD.W1B.missing.freq) <- NULL

# number of missing for each item 
CESD.W1B.missing.freq
```

Each item has around 40-60 missing responses. 

```{r number of items that are missing CESD.W1B}
dat <- dat %>%
  mutate(total_missing_CESD.W1B = 
           rowSums(select(., CESD.W1B.items.missing) == "Missing")) # sums the number of "missing"s for each person 

freq(dat$total_missing_CESD.W1B) # number of items individuals have missing
```

20634 have no missing items 
42 have 1 missing item
7 have 2 missing items
30 have 10 missing items 

##### Create outcome pheno dataset W1B

* Scores preadjusted, i.e. taking residuals of a regression model, for the following variables: year of birth, sex, and top 10 ancestry-specific genetic PCs
* Standardise scores to mean 0 and SD 1
* Select ID and sum score variables
* Remove any individuals that have more than one item (<2) missing

*ANCESTRY PCS HAVE BEEN INCLUDED IN THE OUTCOME WHEN SUBSETTING THE DATA INTO THE SMALLER SAMPLE*

```{r regress out covars W1B}
dat$CESD_W1B_resid <- resid(
  lm(
    CESD_W1B ~ H1GI1Y_factor_clean + BIO_SEX_factor_clean,
    data = dat
    )
  )

# check 
freq(dat$CESD_W1B_resid)
```

```{r standardise W1B}
# standardise
dat <- dat %>% 
  mutate(CESD_W1B_stan = as.numeric(scale(CESD_W1B_resid)))
  
# check 
freq(dat$CESD_W1B_stan)
```

```{r remove individuals with missing depression items into dat.clean.W1B}
dat.clean.W1B <- dat %>%
  filter(total_missing_CESD.W1B < 2) %>%
  select(AID,     # ID
         CESD_W1B_stan # depression sum score W1A
         )

pheno_W1B <- dat.clean.W1B %>% mutate(AID = as.numeric(AID))

freq(dat$total_missing_CESD.W1B) 
nrow(dat.clean.W1B) # this gives us 20676 people 
head(dat.clean.W1B)

bar_CESD_W1B <- ggplot(dat.clean.W1B, aes(x = CESD_W1B_stan)) +
                  geom_histogram(binwidth = 0.2, fill = "blue", color = "black", alpha = 0.7) +
                  labs(title = "Histogram of CESD_W1B Values", x = "Value", y = "Frequency") +
                  theme_minimal()

save_gg_cairo(plot = bar_CESD_W1B, filename = "/home/thom1336/findmedep/figures/bar_CESD_W1B.png")
```

### Wave 4

#### Sum score W4

```{r depression items W4}
CESD.items.key.W4 <- c(1, # scoring key: -1 is reverse coded
                      1, 
                     -1, 
                      1, 
                      1, 
                      1, 
                     -1, 
                     -1, 
                      1, 
                      1) 
```

```{r create sum score for depression at wave IV}
# sum score for CES-D depression wave IV 
CESD.scored.items.W4 <- scoreItems(keys = CESD.items.key.W4,
                                items = dat.W4[CESD.W4.items.clean],
                                totals = TRUE,    # calculate totals
                                missing = TRUE,   # keep misisng items
                                impute = 'none',  # don't impute missing values
                                min = 0, 
                                max = 3)

# add score to dat.W4a frame
dat.W4 <- dat.W4 %>%
  mutate(CESD_W4 = as.numeric(CESD.scored.items.W4$scores))

# check 
freq(dat.W4$CESD_W4)
```

##### Missingness

```{r function for missingness of depression items wave IV}
items.missing <- c()

for (i in CESD.W4.items.clean) {
  
  new_varname <- paste0(i, "_missing")
  
  dat.W4 <- dat.W4 %>%
     mutate(!!new_varname := if_else(
               !is.na(!!sym(i)), 
               "Not missing", 
               "Missing"
           ))
  
  items.missing <- c(items.missing, new_varname) # creates a list of these variables 
}

# rename the general list to W4 specific
CESD.W4.items.missing <- items.missing
```

```{r number of items that are missing CESD.W4}
dat.W4 <- dat.W4 %>%
  mutate(total_missing_CESD.W4 = 
           rowSums(select(., CESD.W4.items.missing) == "Missing")) # sums the number of "missing"s for each person 

dat.W4_raw <- dat.W4

freq(dat.W4_raw$total_missing_CESD.W4) # number of items individuals have missing
```

15672 people have no missing items.
25 have 1 missing item

##### Create outcome pheno dataset W4

* Scores preadjusted, i.e. taking residuals of a regression model, for the following variables: year of birth, sex, and top 10 ancestry-specific genetic PCs
* Standardise scores to mean 0 and SD 1
* Select ID and sum score variables
* Remove any individuals that have more than one item (<2) missing

*ANCESTRY PCS HAVE BEEN INCLUDED IN THE OUTCOME WHEN SUBSETTING THE DATA INTO THE SMALLER SAMPLE*

```{r attach W1 vars to W4}
dat_W1_keyvars <- dat %>% select(AID, H1GI1Y_factor_clean, BIO_SEX_factor_clean) 

dat.W4 <- left_join(dat.W4, dat_W1_keyvars, by = "AID") # 15701 people
nrow(dat.W4)
```

```{r remove those with more than two missing}
dat.W4 <- dat.W4 %>%
  filter(total_missing_CESD.W4 < 2) %>%
  select(AID, CESD_W4, H1GI1Y_factor_clean, BIO_SEX_factor_clean)

# Create a new dataset with only individuals included in the regression
dat.W4 <- dat.W4[!is.na(dat.W4$H1GI1Y_factor_clean) &
                 !is.na(dat.W4$BIO_SEX_factor_clean), ]  # 15686 people

nrow(dat.W4)
```

```{r regress out covars W4}
dat.W4$CESD_W4_resid <- resid(
  lm(
    CESD_W4 ~ H1GI1Y_factor_clean + BIO_SEX_factor_clean,
    data = dat.W4
    )
  )

# check 
freq(dat.W4$CESD_W4_resid)
```

```{r standardise W4}
# standardise
dat.W4 <- dat.W4 %>% 
  mutate(CESD_W4_stan = as.numeric(scale(CESD_W4_resid)))
  
# check 
freq(dat.W4$CESD_W4_stan)
```

```{r create phenoW4 dataset}
dat.clean.W4 <- dat.W4 %>%
  select(AID,     # ID
         CESD_W4_stan # depression sum score W1A
         )

pheno_W4 <- dat.clean.W4 %>% mutate(AID = as.numeric(AID))

freq(dat$total_missing_CESD.W4) 
nrow(dat.clean.W4) # this gives us 15686 people
head(dat.clean.W4)

bar_CESD_W4 <- ggplot(dat.clean.W4, aes(x = CESD_W4_stan)) +
                  geom_histogram(binwidth = 0.2, fill = "blue", color = "black", alpha = 0.7) +
                  labs(title = "Histogram of CESD_W4 Values", x = "Value", y = "Frequency") +
                  theme_minimal()

save_gg_cairo(plot = bar_CESD_W4, filename = "/home/thom1336/findmedep/figures/bar_CESD_W4.png")
```

## Subsample files

### Matching Ids

```{r individuals that have all data}
# check ancestry flag
ancestry_flag

# Find the matching IDs between genetic, enviro, and pheno 
ALL_matching_ids_nosibs <- Reduce(intersect, 
                                  list(dat.erm.theory1.dummy_pca$AID, grm_ids_nosibs$AID, pheno_W1A$AID, pheno_W1B$AID, pheno_W4$AID, ancestry_PCs$AID))
length(ALL_matching_ids_nosibs) 
```

Frequencies based on ancestry flag:
* ALL: 7695 individuals with all clean data 
* EUR: 5030
* AFR: 1573
* AMR: 831
* EAS: 305 

### ID file

```{r create ID file for individuals have have all data}
# FID and AID format
matching_id_grm_erm_FID <- grm_ids_nosibs %>% 
  filter(AID %in% ALL_matching_ids_nosibs) %>%
  mutate(
    FID = as.numeric(FID),
    AID = as.numeric(AID),
    PID = as.numeric(0),
    MID = as.numeric(0),
    Sex = as.numeric(0),
    Phenotype = as.numeric(0)
    ) %>%
  select(FID, AID, PID, MID, Sex, Phenotype)

# AID and AID2 format 
matching_id_grm_erm_AID <- grm_ids_nosibs %>% 
  filter(AID %in% ALL_matching_ids_nosibs) %>%
  mutate(
    AID = as.numeric(AID),
    AID2 = as.numeric(AID),
    PID = as.numeric(0),
    MID = as.numeric(0),
    Sex = as.numeric(0),
    Phenotype = as.numeric(0)
    ) %>%
  select(AID, AID2, PID, MID, Sex, Phenotype)

nrow(matching_id_grm_erm_AID)
```

```{r write the ID file for matching IDs}
write.table(matching_id_grm_erm_FID, paste0(ancestry_sub_path, "matching_id_grm_erm.txt"), row.names = FALSE, col.names = FALSE)
```

Need to re-create the bim bed fam files with the 7695 with complete data - **THIS ONLY NEEDS TO BE DONE ONCE OR IF CHANGING PARTICIPANTS INCLUDED**

Previously - I was using clean_genetic_dat_nosib bfiles which are saved in /project/rche/data/datasets/addhealth/scratch/kthompson/plink_output/. 
Now we have moved to addhealth_imputed_1000g.QC.rel, so these will be read in from the /project/rche/data/datasets/addhealth/scratch/kthompson/qc/ directory. 
Copy over files from qc folder 

```{bash copy over genetic files}
# cd /project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/
# 
# # List of population-specific valid sample files
# populations=("AFR" "AMR" "EAS" "EUR")
# 
# # Copy over ALL files
# cp /project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/ahgwas.QC.rel.ALL.bed /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/clean_genetic_dat_raw.bed
# cp /project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/ahgwas.QC.rel.ALL.bim /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/clean_genetic_dat_raw.bim
# cp /project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/ahgwas.QC.rel.ALL.fam /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/clean_genetic_dat_raw.fam
# 
# # Loop through each population and run PLINK
# for pop in "${populations[@]}"; do
# cp /project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/"ahgwas.QC.${pop}.rel.bed" /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/${pop}/clean_genetic_dat_raw.bed
# cp /project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/"ahgwas.QC.${pop}.rel.bim" /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/${pop}/clean_genetic_dat_raw.bim
# cp /project/rche/data/datasets/addhealth/clean_progress/genetic_files/qc/"ahgwas.QC.${pop}.rel.fam" /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/${pop}/clean_genetic_dat_raw.fam
# done
```

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/plink_sub_files.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following commands as a job on Rossmann: 

```{bash get list of ids that have genetic data}
#  directories=(
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/"
#  )
# 
#  for dir in "${directories[@]}"; do
#      echo "Processing: $dir"
#      cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
# 
#      # Run PLINK command
#      /project/rche/data/datasets/addhealth/scratch/kthompson/findme/./plink2 \
#      --bfile clean_genetic_dat_raw \
#      --keep matching_id_grm_erm.txt \
#      --make-bed \
#      --out clean_genetic_dat_full
# 
#      echo "Finished processing $dir"
#      
#      rm clean_genetic_dat_raw.*
#  done
```

```{r read in new ID file and order base ID file }
grm_ids_clean <- read.table(paste0(ancestry_sub_path, "clean_genetic_dat_full.fam"), header = FALSE)
grm_ids_clean <- grm_ids_clean %>% dplyr::rename(FID = V1, AID = V2) %>% mutate(across(everything(), as.numeric))
nrow(grm_ids_clean)

# reorder the base ID files to match genetic file 
matching_id_grm_erm_FID <- matching_id_grm_erm_FID[order(match(matching_id_grm_erm_FID$AID, grm_ids_clean$AID)), ]
matching_id_grm_erm_AID <- matching_id_grm_erm_AID[order(match(matching_id_grm_erm_AID$AID, grm_ids_clean$AID)), ]

# check 
identical(as.numeric(matching_id_grm_erm_FID$AID), as.numeric(grm_ids_clean$AID)) # these ID files are not ordered the same
identical(as.numeric(matching_id_grm_erm_AID$AID), as.numeric(grm_ids_clean$AID)) # these ID files are not ordered the same
```

Now the *matching_id_grm_erm_FID* and *matching_id_grm_erm_AID* ID files match the order of the genetic file *clean_genetic_dat_full.fam*. 

### Enviro file 

Create dataset for E-REML that only contains individuals that ALSO have genetic data (7695). Will also be used for e_g models. 

7695 people and 479 variables (including ID)

Need to create additional AID variable in enviro data and reorder to match genetic data file. 

```{r create erm data with only matching ids}
# filter to only those that are matching for erm data 
dat.erm.theory1.dummy_pca_grm_nosibs <- as.data.frame(dat.erm.theory1.dummy_pca) %>%
  filter(AID %in% ALL_matching_ids_nosibs) 

dim(dat.erm.theory1.dummy_pca_grm_nosibs)

# attach additional AID variable - mtg 
dat.erm.theory1.dummy_pca_grm_nosibs_mtg <- cbind(dat.erm.theory1.dummy_pca_grm_nosibs$AID, dat.erm.theory1.dummy_pca_grm_nosibs)
dat.erm.theory1.dummy_pca_grm_nosibs_mtg <- dat.erm.theory1.dummy_pca_grm_nosibs_mtg %>% dplyr::rename(AID2 = AID, AID = `dat.erm.theory1.dummy_pca_grm_nosibs$AID`) 

head(dat.erm.theory1.dummy_pca_grm_nosibs_mtg)

# reorder to match genetic file 
theory1_clean_dummy_pca_reordered <- dat.erm.theory1.dummy_pca_grm_nosibs_mtg[order(match(dat.erm.theory1.dummy_pca_grm_nosibs_mtg$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

head(theory1_clean_dummy_pca_reordered)

# check they are the same order
identical(as.numeric(matching_id_grm_erm_FID$AID), theory1_clean_dummy_pca_reordered$AID) # these should be the same - so now need to run the ERM to match and remember to use this ID file as the input file when running the model 

# make all numeric
theory1_clean_dummy_pca_reordered_numeric <- apply(theory1_clean_dummy_pca_reordered, 2, as.numeric)
is.matrix(theory1_clean_dummy_pca_reordered_numeric) && all(sapply(theory1_clean_dummy_pca_reordered_numeric, is.numeric)) # check matrix is numeric

# have a look at data file
as.tibble(theory1_clean_dummy_pca_reordered_numeric)
head(matching_id_grm_erm_AID)
```

In order for mtg2 to recognise all the IDs in the files, we will need to use data *theory1_clean_dummy_pca_reordered_numeric* and ID file *matching_id_grm_erm_FID or matching_id_grm_erm_AID* to match the ID order created by the genetic files. 

### Pheno file 

We re-create the pheno scores now that we have the PC information for this subset sample. So we first filter for all those that have PC information. 

```{r join to ancestry PCs}
# filter for only those that have ancestry PCs
dat_sub <- dat %>% mutate(AID = as.numeric(AID)) %>% filter(AID %in% ALL_matching_ids_nosibs)  
dat_PCs <- left_join(dat_sub, ancestry_PCs, by = "AID") 

dat.W4_sub <- dat.W4 %>% mutate(AID = as.numeric(AID)) %>% filter(AID %in% ALL_matching_ids_nosibs)  
dat.W4_PCs <- left_join(dat.W4_sub, ancestry_PCs, by = "AID") 
```

```{r regress out covars W1A}
dat_PCs$CESD_W1A_resid <- resid(
  lm(
    CESD_W1A ~ H1GI1Y_factor_clean + BIO_SEX_factor_clean 
    + an_PC1 + an_PC2 + an_PC3 + an_PC4 + an_PC5 + an_PC6 + an_PC7 + an_PC8 + an_PC9 + an_PC10 + an_PC11 + an_PC12 + an_PC13 + an_PC14 + an_PC15 + an_PC16 + an_PC17 + an_PC18 + an_PC19 + an_PC20
 , data = dat_PCs
    )
  )

# standardise
dat_PCs <- dat_PCs %>% 
  mutate(CESD_W1A_stan = as.numeric(scale(CESD_W1A_resid)))
  
# 
dat.clean.W1A <- dat_PCs %>%
  filter(total_missing_CESD.W1A < 2) %>%
  select(AID,     # ID
         CESD_W1A_stan # depression sum score W1A - residualised and standardised
         )

# make ID numeric 
pheno_W1A <- dat.clean.W1A %>% mutate(AID = as.numeric(AID))
nrow(dat.clean.W1A)
```

```{r regress out covars W1B}
dat_PCs$CESD_W1B_resid <- resid(
  lm(
    CESD_W1B ~ H1GI1Y_factor_clean + BIO_SEX_factor_clean 
    + an_PC1 + an_PC2 + an_PC3 + an_PC4 + an_PC5 + an_PC6 + an_PC7 + an_PC8 + an_PC9 + an_PC10 + an_PC11 + an_PC12 + an_PC13 + an_PC14 + an_PC15 + an_PC16 + an_PC17 + an_PC18 + an_PC19 + an_PC20
 , data = dat_PCs
    )
  )

# standardise
dat_PCs <- dat_PCs %>% 
  mutate(CESD_W1B_stan = as.numeric(scale(CESD_W1B_resid)))
  
dat.clean.W1B <- dat_PCs %>%
  filter(total_missing_CESD.W1B < 2) %>%
  select(AID,     # ID
         CESD_W1B_stan # depression sum score W1B - residualised and standardised
         )

# make ID numeric 
pheno_W1B <- dat.clean.W1B %>% mutate(AID = as.numeric(AID))
nrow(dat.clean.W1B)
```

```{r regress out covars W4}
dat.W4_PCs$CESD_W4_resid <- resid(
  lm(
    CESD_W4 ~ H1GI1Y_factor_clean + BIO_SEX_factor_clean 
    + an_PC1 + an_PC2 + an_PC3 + an_PC4 + an_PC5 + an_PC6 + an_PC7 + an_PC8 + an_PC9 + an_PC10 + an_PC11 + an_PC12 + an_PC13 + an_PC14 + an_PC15 + an_PC16 + an_PC17 + an_PC18 + an_PC19 + an_PC20
 , data = dat.W4_PCs
    )
  )

# standardise
dat.W4_PCs <- dat.W4_PCs %>% 
  mutate(CESD_W4_stan = as.numeric(scale(CESD_W4_resid)))

dat.clean.W4 <- dat.W4_PCs %>%
  select(AID,     # ID
         CESD_W4_stan # depression sum score W4  - residualised and standardised
         )

pheno_W4 <- dat.clean.W4 %>% mutate(AID = as.numeric(AID))
nrow(dat.clean.W4) # this gives us 8768 people
```

```{r match up genetic and phenotypic data for W1A}
# filter for only those that are matching 
pheno_W1A_nosibs <- pheno_W1A %>%
  filter(AID %in% ALL_matching_ids_nosibs) 

# join family ID to pheno file 
pheno_W1A_grm_nosibs <- left_join(pheno_W1A_nosibs, matching_id_grm_erm_FID, by = "AID") 

# FID and IID structure
# order the variables to be FID, ID, pheno
pheno_W1A_grm_nosibs_FID <- pheno_W1A_grm_nosibs %>%
  select(FID, AID, CESD_W1A_stan) 

pheno_W1A_grm_nosibs_FID <- pheno_W1A_grm_nosibs_FID[order(match(pheno_W1A_grm_nosibs_FID$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

# IID and IID structure 
pheno_W1A_grm_nosibs_IID <- pheno_W1A_grm_nosibs %>%
  mutate(AID2 = AID) %>%
  select(AID, AID2, CESD_W1A_stan) 

pheno_W1A_grm_nosibs_IID <- pheno_W1A_grm_nosibs_IID[order(match(pheno_W1A_grm_nosibs_IID$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]
```

```{r match up genetic and phenotypic data for W1B}
# filter for only those that are matching 
pheno_W1B_nosibs <- pheno_W1B %>%
  filter(AID %in% ALL_matching_ids_nosibs) 

# join family ID to pheno file 
pheno_W1B_grm_nosibs <- left_join(pheno_W1B_nosibs, matching_id_grm_erm_FID, by = "AID") 

# FID and IID structure
# order the variables to be FID, ID, pheno
pheno_W1B_grm_nosibs_FID <- pheno_W1B_grm_nosibs %>%
  select(FID, AID, CESD_W1B_stan) 

pheno_W1B_grm_nosibs_FID <- pheno_W1B_grm_nosibs_FID[order(match(pheno_W1B_grm_nosibs_FID$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

# IID and IID structure 
pheno_W1B_grm_nosibs_IID <- pheno_W1B_grm_nosibs %>%
  mutate(AID2 = AID) %>%
  select(AID, AID2, CESD_W1B_stan) 

pheno_W1B_grm_nosibs_IID <- pheno_W1B_grm_nosibs_IID[order(match(pheno_W1B_grm_nosibs_IID$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]
```

```{r IID and IID structure}
# IID and IID structure 
pheno_W4_grm_nosibs_IID <- dat.clean.W4 %>%
  mutate(AID2 = AID) %>%
  select(AID, AID2, CESD_W4_stan) 

pheno_W4_grm_nosibs_IID <- pheno_W4_grm_nosibs_IID %>%
  mutate_all(as.numeric)

pheno_W4_grm_nosibs_IID <- pheno_W4_grm_nosibs_IID[order(match(pheno_W4_grm_nosibs_IID$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]
```

### Covariate file

The FID and IID order for phenotype file and covariate files (cc, qc) should be the same.

```{r create categorical covariate file sub sample}
# IID and IID structure
dat.cov_cat_sub_IID <- dat %>%
  filter(AID %in% ALL_matching_ids_nosibs) %>%
  mutate(AID = as.numeric(AID)) %>%
  mutate(AID2 = AID) %>%
  select(AID,     # ID1
         AID2,    # ID2
         BIO_SEX_factor_clean,
         H1GI1Y_factor_clean
         )

# match order across files
dat.cov_cat_sub_IID <- dat.cov_cat_sub_IID[order(match(dat.cov_cat_sub_IID$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

# check 
identical(as.numeric(matching_id_grm_erm_FID$AID), as.numeric(dat.cov_cat_sub_IID$AID)) 

# FID and IID structure
dat.cov_cat_sub_FID <- dat %>%
  filter(AID %in% ALL_matching_ids_nosibs) %>%
  mutate(AID = as.numeric(AID)) 

dat.cov_cat_sub_FID <- left_join(dat.cov_cat_sub_FID, matching_id_grm_erm_FID, by = "AID")

dat.cov_cat_sub_FID <- dat.cov_cat_sub_FID %>%
  select(FID,     # FID
         AID,     # AID
         BIO_SEX_factor_clean,
         H1GI1Y_factor_clean
         )

dat.cov_cat_sub_FID <- dat.cov_cat_sub_FID[order(match(dat.cov_cat_sub_FID$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

# check 
identical(as.numeric(matching_id_grm_erm_FID$AID), as.numeric(dat.cov_cat_sub_FID$AID)) 
```

```{r create categorical covariate file sub sample cont}
# IID and IID structure
dat.cov_cont_sub_IID <- dat_PCs %>%
  filter(AID %in% ALL_matching_ids_nosibs) %>%
  mutate(AID = as.numeric(AID)) %>%
  mutate(AID2 = AID) %>%
  select(AID,     # ID1
         AID2,    # ID2
         an_PC1, an_PC2, an_PC3, an_PC4, an_PC5, an_PC6, an_PC7, an_PC8, an_PC9, an_PC10,
         an_PC11, an_PC12, an_PC13, an_PC14, an_PC15, an_PC16, an_PC17, an_PC18, an_PC19, an_PC20
         )

# match order across files
dat.cov_cont_sub_IID <- dat.cov_cont_sub_IID[order(match(dat.cov_cont_sub_IID$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

# check 
identical(as.numeric(matching_id_grm_erm_FID$AID), as.numeric(dat.cov_cont_sub_IID$AID)) 

# FID and IID structure
dat.cov_cont_sub_FID <- dat_PCs %>%
  filter(AID %in% ALL_matching_ids_nosibs) %>%
  mutate(AID = as.numeric(AID)) 

dat.cov_cont_sub_FID <- dat.cov_cont_sub_FID %>%
  select(FID,     # FID
         AID,     # AID
         an_PC1, an_PC2, an_PC3, an_PC4, an_PC5, an_PC6, an_PC7, an_PC8, an_PC9, an_PC10,
         an_PC11, an_PC12, an_PC13, an_PC14, an_PC15, an_PC16, an_PC17, an_PC18, an_PC19, an_PC20
         )

dat.cov_cont_sub_FID <- dat.cov_cont_sub_FID[order(match(dat.cov_cont_sub_FID$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

# check 
identical(as.numeric(matching_id_grm_erm_FID$AID), as.numeric(dat.cov_cont_sub_FID$AID)) 
```

## Write files

### Sub samples 

Subsample that match in order according to ID file (N = 8768)

* ID file:  *matching_id_grm_erm_FID* or *matching_id_grm_erm_AID* to match the ID order created by the genetic files. 
* enviro file: *theory1_clean_dummy_pca_reordered_numeric* contains complete data pca tarnsformed, ordered, and numeric 
* pheno file: *pheno_W1A_grm_nosibs_FID* or *pheno_W1B_grm_nosibs_IID* for W1A and W1B
* genetic file: *clean_genetic_dat_full* for bed, bim, fam files

```{r final check the order of each file}
# ID file and enviro file
identical(as.numeric(matching_id_grm_erm_AID$AID), as.tibble(theory1_clean_dummy_pca_reordered_numeric)$AID)

# ID file and genetic ID file
identical(as.numeric(matching_id_grm_erm_AID$AID), as.tibble(grm_ids_clean)$AID)

# ID file and pheno file
identical(as.numeric(matching_id_grm_erm_AID$AID), as.tibble(pheno_W1A_grm_nosibs_IID)$AID)
identical(as.numeric(matching_id_grm_erm_AID$AID), as.tibble(pheno_W1B_grm_nosibs_IID)$AID)

# ID file and covariate file
identical(as.numeric(matching_id_grm_erm_AID$AID), as.tibble(dat.cov_cat_sub_IID)$AID)
identical(as.numeric(matching_id_grm_erm_AID$AID), as.tibble(dat.cov_cont_sub_IID)$AID)
```

* dat.erm.theory1_clean contains non-missing data and includes all categorical variables as numeric categorical. 
* dat.erm.theory1.dummy contains non-missing data and includes all categorical variables as dummy variables (binary).
* dat.erm.theory1.dummy_pca contains non-misisng data and includes categorical variables as dummy variables, AND is PCA transformed.
* theory1_clean_dummy_pca_reordered_numeric contains non-misisng data and includes categorical variables as dummy variables, is PCA transformed, ordered to match ID file and set to be numeric. 

```{r sub sample files}
# ID files
## oneE
write.table(matching_id_grm_erm_FID, paste0(ancestry_sub_path, "oneE/IDs_FID.fam"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
write.table(matching_id_grm_erm_AID, paste0(ancestry_sub_path, "oneE/IDs_IID.fam"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
## iterations - wholevar
write.table(matching_id_grm_erm_FID, paste0(ancestry_sub_path, "iterations/wholevar/IDs_FID.fam"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
write.table(matching_id_grm_erm_AID, paste0(ancestry_sub_path, "iterations/wholevar/IDs_IID.fam"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
## iterations - incremental add 
write.table(matching_id_grm_erm_FID, paste0(ancestry_sub_path, "iterations/incremental_add/IDs_FID.fam"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
write.table(matching_id_grm_erm_AID, paste0(ancestry_sub_path, "iterations/incremental_add/IDs_IID.fam"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
## domains
write.table(matching_id_grm_erm_FID, paste0(ancestry_sub_path, "domains/IDs_FID.fam"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
write.table(matching_id_grm_erm_AID, paste0(ancestry_sub_path, "domains/IDs_IID.fam"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# enviro file
write.table(theory1_clean_dummy_pca_reordered_numeric, paste0(ancestry_sub_path, "oneE/theory1_clean_dummy_pca"), col.names = FALSE, row.names = FALSE, quote = FALSE) 

# pheno files
## oneE
write.table(pheno_W1A_grm_nosibs_IID, paste0(ancestry_sub_path, "oneE/pheno_W1A_IID.txt"), row.names = FALSE, col.names = FALSE) 
write.table(pheno_W1A_grm_nosibs_FID, paste0(ancestry_sub_path, "oneE/pheno_W1A_FID.txt"), row.names = FALSE, col.names = FALSE)
write.table(pheno_W1B_grm_nosibs_IID, paste0(ancestry_sub_path, "oneE/pheno_W1B_IID.txt"), row.names = FALSE, col.names = FALSE) 
write.table(pheno_W1B_grm_nosibs_FID, paste0(ancestry_sub_path, "oneE/pheno_W1B_FID.txt"), row.names = FALSE, col.names = FALSE)
## iterations - wholevar - W1A only 
write.table(pheno_W1A_grm_nosibs_IID, paste0(ancestry_sub_path, "iterations/wholevar/pheno_W1A_IID.txt"), row.names = FALSE, col.names = FALSE)
write.table(pheno_W1A_grm_nosibs_FID, paste0(ancestry_sub_path, "iterations/wholevar/pheno_W1A_FID.txt"), row.names = FALSE, col.names = FALSE)
## domains
write.table(pheno_W1A_grm_nosibs_IID, paste0(ancestry_sub_path, "domains/pheno_W1A_IID.txt"), row.names = FALSE, col.names = FALSE)
write.table(pheno_W1A_grm_nosibs_FID, paste0(ancestry_sub_path, "domains/pheno_W1A_FID.txt"), row.names = FALSE, col.names = FALSE)
write.table(pheno_W1B_grm_nosibs_IID, paste0(ancestry_sub_path, "domains/pheno_W1B_IID.txt"), row.names = FALSE, col.names = FALSE)
write.table(pheno_W1B_grm_nosibs_FID, paste0(ancestry_sub_path, "domains/pheno_W1B_FID.txt"), row.names = FALSE, col.names = FALSE)
## gcta - NOT NEEDED ANYMORE BUT WILL NEED TO EDIT FURTHER IN CODE - DELETE ONCE MADE EDITS
# write.table(pheno_W1A_grm_nosibs_IID, paste0(ancestry_sub_path, "/project/rche/data/datasets/addhealth/scratch/kthompson/gcta_output/complete_dat_nosibs/pheno_W1A_IID.txt", row.names = FALSE, col.names = FALSE) 

# covariate files
## oneE
write.table(dat.cov_cat_sub_IID,  paste0(ancestry_sub_path, "oneE/covariate_cat_IID.txt"), row.names = FALSE, col.names = FALSE) 
write.table(dat.cov_cat_sub_FID,  paste0(ancestry_sub_path, "oneE/covariate_cat_FID.txt"), row.names = FALSE, col.names = FALSE)
write.table(dat.cov_cont_sub_IID,  paste0(ancestry_sub_path, "oneE/covariate_cont_IID.txt"), row.names = FALSE, col.names = FALSE) 
write.table(dat.cov_cont_sub_FID,  paste0(ancestry_sub_path, "oneE/covariate_cont_FID.txt"), row.names = FALSE, col.names = FALSE)
## iterations
write.table(dat.cov_cat_sub_IID,  paste0(ancestry_sub_path, "iterations/wholevar/covariate_cat_IID.txt"), row.names = FALSE, col.names = FALSE) 
write.table(dat.cov_cat_sub_FID,  paste0(ancestry_sub_path, "iterations/wholevar/covariate_cat_FID.txt"), row.names = FALSE, col.names = FALSE)
write.table(dat.cov_cont_sub_IID,  paste0(ancestry_sub_path, "iterations/wholevar/covariate_cont_IID.txt"), row.names = FALSE, col.names = FALSE) 
write.table(dat.cov_cont_sub_FID,  paste0(ancestry_sub_path, "iterations/wholevar/covariate_cont_FID.txt"), row.names = FALSE, col.names = FALSE)
## domains
write.table(dat.cov_cat_sub_IID,  paste0(ancestry_sub_path, "domains/covariate_cat_IID.txt"), row.names = FALSE, col.names = FALSE) 
write.table(dat.cov_cat_sub_FID,  paste0(ancestry_sub_path, "domains/covariate_cat_FID.txt"), row.names = FALSE, col.names = FALSE)
write.table(dat.cov_cont_sub_IID,  paste0(ancestry_sub_path, "domains/covariate_cont_IID.txt"), row.names = FALSE, col.names = FALSE) 
write.table(dat.cov_cont_sub_FID,  paste0(ancestry_sub_path, "domains/covariate_cont_FID.txt"), row.names = FALSE, col.names = FALSE)
```

```{r write W4 pheno files}
# pheno files
## oneE
write.table(pheno_W4_grm_nosibs_IID,  paste0(ancestry_sub_path, "oneE/pheno_W4_IID.txt"), row.names = FALSE, col.names = FALSE)
## domains
write.table(pheno_W4_grm_nosibs_IID,  paste0(ancestry_sub_path, "domains/pheno_W4_IID.txt"), row.names = FALSE, col.names = FALSE)
```

# Descriptives

```{r missingness for W4 sum score}
dat.W4_full <- dat.W4  %>%
  filter(AID %in% ALL_matching_ids_nosibs) 

freq(dat.W4_full$total_missing_CESD.W4)
```

```{r calculate age at W4}
dat1_raw <- read_xpt("/project/rche/data/datasets/addhealth/Core Files - Wave I/Wave I In Home Interview Data/allwave1.xpt")

# clean covariate missing scores
dat.age <- dat1_raw %>% 
  filter(AID %in% ALL_matching_ids_nosibs) %>%
  dplyr::mutate(
  H1GI1Y_clean =  case_when(H1GI1Y %in% c(96, 97, 98, 99) ~ NA_real_, TRUE ~ H1GI1Y),
  H1GI1M_clean =  case_when(H1GI1M %in% c(96, 97, 98, 99) ~ NA_real_, TRUE ~ H1GI1M),
  interview_date = as.Date(paste(paste0("19", IYEAR), IMONTH, IDAY, sep = "-"), format = "%Y-%m-%d"),
  birthday_date = # birth date - combine to date format - have set to be the 1st of each month so not accurate
           as.Date(paste(paste0("19", H1GI1Y_clean), H1GI1M_clean, "1", sep = "-"), format = "%Y-%m-%d"),
  age_days = # age at interview (approx) - days
           as.numeric(difftime(interview_date, birthday_date, units = "days")),
  age_years =  # age at interview (approx) - years
           as.numeric(floor(age_days/365))
  ) %>%
  select(AID, interview_date, birthday_date, age_days, age_years) 

summary(dat.age$age_years)


dat.W4.age <- dat.W4.raw %>% 
  filter(AID %in% ALL_matching_ids_nosibs) %>%
  dplyr::mutate(
  BIO_SEX4_clean = case_when(BIO_SEX4 %in% c(6, 7, 8, 9) ~ NA_real_, TRUE ~ BIO_SEX4),
  H4OD1Y_clean =  case_when(H4OD1Y %in% c(96, 97, 98, 99) ~ NA_real_, TRUE ~ H4OD1Y),
  H4OD1M_clean =  case_when(H4OD1M %in% c(96, 97, 98, 99) ~ NA_real_, TRUE ~ H4OD1M),
  interview_date4 = as.Date(paste(IYEAR4, IMONTH4, IDAY4, sep = "-"), format = "%Y-%m-%d"),
  birthday_date4 = # birth date - combine to date format - have set to be the 1st of each month so not accurate
           as.Date(paste(H4OD1Y_clean, H4OD1M_clean, "1", sep = "-"), format = "%Y-%m-%d"),
  age_days4 = # age at interview (approx) - days
           as.numeric(difftime(interview_date4, birthday_date4, units = "days")),
  age_years4 =  # age at interview (approx) - years
           as.numeric(floor(age_days4/365))
  ) %>%
  select(AID, BIO_SEX4_clean, interview_date4, birthday_date4, age_years4)  

summary(dat.W4.age$age_years4)
freq(dat.W4.age$BIO_SEX4_clean)
```

```{r crombach alphas}
cron.alpha.dat <- dat %>%
  filter(AID %in% ALL_matching_ids_nosibs) 

psych::alpha(cron.alpha.dat[,CESD.W1A.items])
psych::alpha(cron.alpha.dat[,CESD.W1B.items])
```

# A. Linear Regression

We will use a linear regression model as a benchmark against the LMM. LMM can accommodate more variables than linear regression and we will use the LMM in the following stages of analysis. Whilst including many predictors in a single regression model violates assumptions of independence, the linear regression model will provide a baseline comparative R2 estimate to compare to that of the linear mixed model.

```{r create list of predictors for regression}
# get column names and convert them into a list
predictors_list <- as.list(names(dat.erm.theory1)[-which(names(dat.erm.theory1) == "AID")])  # exclude the outcome variable

# dummy vars 
predictors_list_dummy <- as.list(names(dat.erm.theory1.dummy)[-which(names(dat.erm.theory1.dummy) == "AID")])  # exclude the outcome variable

# Replace "_numeric" with "_factor" in the list so the lm model recognizes factor levels 
predictors_list_factor <- lapply(predictors_list, function(x) {
  if (grepl("_numeric", x)) {
    gsub("_numeric", "_factor", x)
  } else {
    x
  }
})
```

```{r regression datasets W1A}
# create dataset with depression outcome for dummy data - W1A 
dat.erm.theory1_dummy_W1Areg <- dat.erm.theory1.dummy %>%
  inner_join(pheno_W1A, by = "AID")

# create dataset with depression outcome for dummy data - W1B 
dat.erm.theory1_dummy_W1Breg <- dat.erm.theory1.dummy %>%
  inner_join(pheno_W1B, by = "AID")

# create dataset with depression outcome for dummy data - W1B 
dat.erm.theory1_dummy_W4reg <- dat.erm.theory1.dummy %>%
  inner_join(pheno_W4, by = "AID")
```

```{r formula for regression models}
# formula for depression scores for dummy vars 
formula_W1A_dummy <- as.formula(paste("CESD_W1A_stan ~", paste(predictors_list_dummy, collapse = " + ")))
formula_W1B_dummy <- as.formula(paste("CESD_W1B_stan ~", paste(predictors_list_dummy, collapse = " + ")))
formula_W4_dummy <- as.formula(paste("CESD_W4_stan ~", paste(predictors_list_dummy, collapse = " + ")))
```

```{r regression W1A dummy}
model_W1A_dummy_full <- lm(formula_W1A_dummy, data = dat.erm.theory1_dummy_W1Areg)
model_W1A_dummy_output <- summary(model_W1A_dummy_full)

model_W1A_dummy_output$r.squared
model_W1A_dummy_output$adj.r.squared
```

```{r regression W1B dummy}
model_W1B_dummy_full <- lm(formula_W1B_dummy, data = dat.erm.theory1_dummy_W1Breg)
model_W1B_dummy_output <- summary(model_W1B_dummy_full)

model_W1B_dummy_output$r.squared
model_W1B_dummy_output$adj.r.squared
```

```{r regression W4 dummy}
model_W4_dummy_full <- lm(formula_W4_dummy, data = dat.erm.theory1_dummy_W4reg)
model_W4_dummy_output <- summary(model_W4_dummy_full)

model_W4_dummy_output$r.squared
model_W4_dummy_output$adj.r.squared
```

In both scores, environmental data explained around 50% of the variance in depression symptoms. 

# B. OneE

*Mtg2 general command stucture:* 

./mtg2 -p {plink fam file name} -d {phenotype file name} –g {grm file name} -cc {class covariate file name} –qc {continuous covariate file name} -out {output file name} -sv {starting value file name} –mod {number of traits}

NOTE: -p, -d and -g can be replaced with -fam, -pheno and -grm, respectively.

Generate N x N product matrix using -pdmx

In the context of PLINK, a widely used tool for genome-wide association studies and population genetics, the .fam file is a crucial component of the standard PLINK binary file format. The file stores sample information and is associated with the .bed file (which contains the genotype data) and the .bim file (which contains the SNP information). The PLINK fam file is your *.fam file that used in estimating the grm. Please see https://www.cog-genomics.org/plink2/formats#fam. Briefly, the columns have Family ID (FID), individual ID (IID), father ID, mother ID, gender and phenotypes. MTG2 only use the first two columns (FID and IID).

Based on the demo files given by Zhou and Lee: 
.fam is the IDs
.mat is the matrix to be included (also can include multiple matrices but dont understand how)
.dat file is the phenotype file 

When individuals have multiple phenotypes across different environments (repeated measures), one needs to model residual covariances (permanent environmental effects). This can be done a multiple random effects model including one of GRMs as identity matrix. This is done using -mg - so when we have e, g, exe, gxe then we will use the -mg command. 

I have used -g here until we include multiple matrices. 

**All mtg2 comments hashed out as an example of what is included in the job files** 

Need ID file that is AID, AID not FID, AID. 

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_e_reml_sub.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following commands as a job on Rossmann:

```{bash mtg2 ereml subset pca}
# directories=(
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
# )
# 
#  for dir in "${directories[@]}"; do
#      echo "Processing: $dir"
#      cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
# 
#      # Create relationship matrix 
#      ./../../../mtg2 -p IDs_IID.fam -pdmx theory1_clean_dummy_pca -thread 100 -out theory1_clean_dummy_pca_output.e
#      
#      # Compute E-REML
#      ./../../../mtg2 -p IDs_IID.fam -g theory1_clean_dummy_pca_output.e -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_reml_W1A.out > e_reml_W1A.log
#     ./../../../mtg2 -p IDs_IID.fam -g theory1_clean_dummy_pca_output.e -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_reml_W1B.out > e_reml_W1B.log
#     ./../../../mtg2 -p IDs_IID.fam -g theory1_clean_dummy_pca_output.e -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_reml_W4.out > e_reml_W4.log
#
#     echo "Finished processing $dir"
# done
```

### Delta output prep

```{bash}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_reml_W1A.out > e_reml_W1A.out2
       grep -vwE '(LKH|h2)' e_reml_W1B.out > e_reml_W1B.out2
       grep -vwE '(LKH|h2)' e_reml_W4.out > e_reml_W4.out2
done
```

Prepare a do file for MTG2 on to construct functions of model parameters in R: 
The numbers represent the rows in the out file - corresponding estimates of V. 

* ‘R’ is to get the ratio of variance
* R 2 1 3 4 (i.e. 2 / (1 + 2 + 3 + 4))
* R 3 1 2 4 (i.e. 3 / (1 + 2 + 3 + 4))
* R 4 1 2 3 (i.e. 4 / (1 + 2 + 3 + 4))
* The format here is to put the number of the estimate you want first and then all other numbers after

So R 2 1 3 4 4 will do 2 / 1 + 2 + 3 + 4 + 4
* Cannot compute the proportion for cov(ge) because it does not let you do 2*4 in the do file. But we can work this out ourselves doing 1-all variance components (without a SE estimate).

```{r do file for e_reml_W1A}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_reml_W1A.do")          # create file with this name
cat("e_reml_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(2, "\n")                   # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1", "\n")             # line 3: compute prop. of variance due to environments
cat("R 1 2", "\n")             # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_reml_W1B}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_reml_W1B.do")          # create file with this name
cat("e_reml_W1B.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(2, "\n")                   # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1", "\n")             # line 3: compute prop. of variance due to environments
cat("R 1 2", "\n")             # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_reml_W4}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_reml_W4.do")          # create file with this name
cat("e_reml_W4.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(2, "\n")                   # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1", "\n")             # line 3: compute prop. of variance due to environments
cat("R 1 2", "\n")             # line 4: compute prop. of variance due to error
sink()
```

Output using Delta

```{bash output for e_g_cov_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_reml_W1A.do > e_reml_W1A.out3
    ./../../../mtg2 -delta2 e_reml_W1B.do > e_reml_W1B.out3
    ./../../../mtg2 -delta2 e_reml_W4.do > e_reml_W4.out3
done
```

### Output table 

```{r output for e_reml_W1A}
# For "e" model - e_reml_W1A.out3 - note order of domains 
e_reml_W1A.output <- extract_delta_output(outcome = "W1A", 
                                         model_prefix = "e",
                                         ancestry_sub_path,
                                         domain_labels = c("Environment", "Error"))

# For "e" model - e_reml_W1B.out3 - note order of domains 
e_reml_W1B.output <- extract_delta_output(outcome = "W1B", 
                                         model_prefix = "e",
                                         ancestry_sub_path,
                                         domain_labels = c("Environment", "Error"))

# For "e" model - e_reml_W4.out3 - note order of domains 
e_reml_W4.output <- extract_delta_output(outcome = "W4", 
                                         model_prefix = "e",
                                         ancestry_sub_path,
                                         domain_labels = c("Environment", "Error"))

e_reml_W1A.output
e_reml_W1B.output
e_reml_W4.output
```

**Output provided in part D**

# C. G - SUPPLEMENT

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/gcta_mtg2_g_reml_sub.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following steps as a job on Rossmann: 

```{bash compute greml}
# directories=(
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
# )
# 
# for dir in "${directories[@]}"; do
#      echo "Processing: $dir"
#      cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
# 
#      # GRM in gcta
#      ./../../../gcta64 --bfile ../clean_genetic_dat_full --make-grm-gz --autosome --out grm_full_out # excludes sex chromosomes
#      
#      # REML in mtg2
#       ./../../../mtg2 -p IDs_IID.fam -zg grm_full_out.grm.gz -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out g_reml_W1A.out > g_reml_W1A.log
#       ./../../../mtg2 -p IDs_IID.fam -zg grm_full_out.grm.gz -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out g_reml_W1B.out > g_reml_W1B.log
#       ./../../../mtg2 -p IDs_IID.fam -zg grm_full_out.grm.gz -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out g_reml_W4.out > g_reml_W4.log
#
#     # remove previous file 
#      rm grm_full_out.grm
      # unzip the genetic file
#      gunzip grm_full_out.grm.gz
#      
#      echo "Finished processing $dir"
# done
```

### Delta output prep

```{bash}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' g_reml_W1A.out > g_reml_W1A.out2
       grep -vwE '(LKH|h2)' g_reml_W1B.out > g_reml_W1B.out2
       grep -vwE '(LKH|h2)' g_reml_W4.out > g_reml_W4.out2
done
```

Prepare a do file for MTG2 on to construct functions of model parameters in R: 
The numbers represent the rows in the out file - corresponding estimates of V. 

* ‘R’ is to get the ratio of variance
* R 2 1 3 4 (i.e. 2 / (1 + 2 + 3 + 4))
* R 3 1 2 4 (i.e. 3 / (1 + 2 + 3 + 4))
* R 4 1 2 3 (i.e. 4 / (1 + 2 + 3 + 4))
* The format here is to put the number of the estimate you want first and then all other numbers after

So R 2 1 3 4 4 will do 2 / 1 + 2 + 3 + 4 + 4
* Cannot compute the proportion for cov(ge) because it does not let you do 2*4 in the do file. But we can work this out ourselves doing 1-all variance components (without a SE estimate).

```{r do file for g_reml_W1A}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("g_reml_W1A.do")          # create file with this name
cat("g_reml_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(2, "\n")                   # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1", "\n")             # line 3: compute prop. of variance due to genetics
cat("R 1 2", "\n")             # line 4: compute prop. of variance due to error
sink()
```

```{r do file for g_reml_W1B}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("g_reml_W1B.do")          # create file with this name
cat("g_reml_W1B.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(2, "\n")                   # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1", "\n")             # line 3: compute prop. of variance due to genetics
cat("R 1 2", "\n")             # line 4: compute prop. of variance due to error
sink()
```

```{r do file for g_reml_W4}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("g_reml_W4.do")          # create file with this name
cat("g_reml_W4.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(2, "\n")                   # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1", "\n")             # line 3: compute prop. of variance due to genetics
cat("R 1 2", "\n")             # line 4: compute prop. of variance due to error
sink()
```

Output using Delta

```{bash output for g_g_cov_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 g_reml_W1A.do > g_reml_W1A.out3
    ./../../../mtg2 -delta2 g_reml_W1B.do > g_reml_W1B.out3
    ./../../../mtg2 -delta2 g_reml_W4.do > g_reml_W4.out3
done
```

### Output table 

```{r output for e_reml_W1A}
# For "e" model - g_reml_W1A.out3 - note order of domains 
g_reml_W1A.output <- extract_delta_output(outcome = "W1A", 
                                         model_prefix = "g",
                                         ancestry_sub_path,
                                         domain_labels = c("Genetic", "Error"))

# For "e" model - g_reml_W1B.out3 - note order of domains 
g_reml_W1B.output <- extract_delta_output(outcome = "W1B", 
                                         model_prefix = "g",
                                         ancestry_sub_path,
                                         domain_labels = c("Genetic", "Error"))

# For "e" model - g_reml_W4.out3 - note order of domains 
g_reml_W4.output <- extract_delta_output(outcome = "W4", 
                                         model_prefix = "g",
                                         ancestry_sub_path,
                                         domain_labels = c("Genetic", "Error"))

g_reml_W1A.output
g_reml_W1B.output
g_reml_W4.output
```

**Output provided in part D**

# D. OneE and G 

This file has four columns, we need to reformat this to be three columns: position1 position2 relatedness. So we remove the third column. 

```{r read in and reformat grm}
grm_full_out.grm <- fread(paste0(ancestry_sub_path, "oneE/grm_full_out.grm"), header = FALSE)

head(grm_full_out.grm) # the third column is the numer of SNPs included so we remove this to just get the off-diagonal of the relatedness matrices

grm_full_out.grm_reformat <- grm_full_out.grm %>% select(V1, V2, V4)

head(grm_full_out.grm_reformat)

write.table(grm_full_out.grm_reformat, paste0(ancestry_sub_path, "oneE/grm_full_out.grm_reformat"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

Create a txt file with the erm and grm in a list together

```{r create e_g txt file}
# create txt file with erm and grm
erm.grm.txt <- c("grm_full_out.grm_reformat", "theory1_clean_dummy_pca_output.e") 
write.table(erm.grm.txt, paste0(ancestry_sub_path, "oneE/erm.grm.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_e_g_reml_sub.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann: 

Compute REML for W1A and W1B with both g and e included in the model. 

```{bash compute ereml and greml model}
# directories=(
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
# )
# 
# for dir in "${directories[@]}"; do
#      echo "Processing: $dir"
#      cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
# 
#      # E-REML and G-REML
#      ./../../mtg2 -p IDs_IID.fam -mg erm.grm.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_reml_W1A.out > e_g_reml_W1A.log
#      ./../../mtg2 -p IDs_IID.fam -mg erm.grm.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_reml_W1B.out > e_g_reml_W1B.log
#      ./../../mtg2 -p IDs_IID.fam -mg erm.grm.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_reml_W4.out > e_g_reml_W4.log
#      
#      echo "Finished processing $dir"
# done
```

The model presented by KT at WCPG was on a slightly different dataset (siblings included and people dropped based on SES missingness). This model is saved as: WCPG_e_g_reml_W1A.log and WCPG_e_g_reml_W1A.out in the same directory (but code not here as has now been rewritten).

### Delta output prep

```{bash}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_reml_W1A.out > e_g_reml_W1A.out2
       grep -vwE '(LKH|h2)' e_g_reml_W1B.out > e_g_reml_W1B.out2
       grep -vwE '(LKH|h2)' e_g_reml_W4.out > e_g_reml_W4.out2
done
```

Prepare a do file for MTG2 on to construct functions of model parameters in R: 
The numbers represent the rows in the out file - corresponding estimates of V. 

* ‘R’ is to get the ratio of variance
* R 2 1 3 4 (i.e. 2 / (1 + 2 + 3 + 4))
* R 3 1 2 4 (i.e. 3 / (1 + 2 + 3 + 4))
* R 4 1 2 3 (i.e. 4 / (1 + 2 + 3 + 4))
* The format here is to put the number of the estimate you want first and then all other numbers after

So R 2 1 3 4 4 will do 2 / 1 + 2 + 3 + 4 + 4
* Cannot compute the proportion for cov(ge) because it does not let you do 2*4 in the do file. But we can work this out ourselves doing 1-all variance components (without a SE estimate).

```{r do file for e_g_reml_W1A}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_reml_W1A.do")          # create file with this name
cat("e_g_reml_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(3, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3", "\n")             # line 3: compute prop. of variance due to genetics
cat("R 3 1 2", "\n")             # line 4: compute prop. of variance due to environments
cat("R 1 2 3", "\n")             # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_reml_W1B}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_reml_W1B.do")
cat("e_g_reml_W1B.out2", "\n")     # line 1: specify the file with parameter estimates
cat(3, "\n")                           # line 2: tot. # of variance & covariance components in the file
cat("R 2 1 3", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2", "\n")               # line 4: compute prop. of variance due to environments
cat("R 1 2 3", "\n")               # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_reml_W4}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_reml_W4.do")
cat("e_g_reml_W4.out2", "\n")     # line 1: specify the file with parameter estimates
cat(3, "\n")                           # line 2: tot. # of variance & covariance components in the file
cat("R 2 1 3", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2", "\n")               # line 4: compute prop. of variance due to environments
cat("R 1 2 3", "\n")               # line 4: compute prop. of variance due to error
sink()
```

Output using Delta

```{bash output for e_g_cov_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_reml_W1A.do > e_g_reml_W1A.out3
    ./../../../mtg2 -delta2 e_g_reml_W1B.do > e_g_reml_W1B.out3
    ./../../../mtg2 -delta2 e_g_reml_W4.do > e_g_reml_W4.out3
done
```

### Output table 

```{r output for e_reml_W1A}
# For "e_g" model - e_g_reml_W1A.out3 - note order of domains 
e_g_reml_W1A.output <- extract_delta_output(outcome = "W1A", 
                                         model_prefix = "e_g",
                                         ancestry_sub_path,
                                         domain_labels = c("Genetic", "Environment", "Error"))

# For "e_g" model - e_g_reml_W1B.out3 - note order of domains 
e_g_reml_W1B.output <- extract_delta_output(outcome = "W1B", 
                                         model_prefix = "e_g",
                                         ancestry_sub_path,
                                         domain_labels = c("Genetic", "Environment", "Error"))

# For "e_g" model - e_g_reml_W4.out3 - note order of domains 
e_g_reml_W4.output <- extract_delta_output(outcome = "W4", 
                                         model_prefix = "e_g",
                                         ancestry_sub_path,
                                         domain_labels = c("Genetic", "Environment", "Error"))

e_g_reml_W1A.output
e_g_reml_W1B.output
e_g_reml_W4.output
```

```{r plot data from delta}
e_g_reml.output <- rbind(e_reml_W1A.output, e_reml_W1B.output, e_reml_W4.output,
                         g_reml_W1A.output, g_reml_W1B.output, g_reml_W4.output,
                         e_g_reml_W1A.output, e_g_reml_W1B.output, e_g_reml_W4.output)
```

### LRT 

The significance of the variance components can be determined through a series of model comparisons using likelihood ratio tests. 

Here we compare the variance components across the y ~ e + error and y ~ g + e + error models. 

LRT = −2 * (log-likelihood of constrained model − log-likelihood of full model)

Low p-value (< 0.05): Full model fits significantly better → the additional parameters improve model fit
High p-value (≥ 0.05): No significant improvement → stick with the simpler model

Constrained model with have a higher absolute LKH value, but as it's negative it's always the lowest number. 

Nested model is y = e + error (simpler, fewer predictors)
If significant (<0.05) = the full model provides a significantly better fit than the nested model
If non-significant (>0.05) = the nested model (simpler model) fits the data as well as the full model (more complex model)

```{r LRT test W1A ge}
LRT_e_ge_W1A <- LRT(path_full = "oneE/e_g_reml_W1A.out", 
                    path_constrained = "oneE/e_reml_W1A.out",
                    Outcome = "W1A",
                    Comparison = "OneE compared to G&oneE", 
                    df = 1)

LRT_e_ge_W1B <- LRT(path_full = "oneE/e_g_reml_W1B.out", 
                    path_constrained = "oneE/e_reml_W1B.out",
                    Outcome = "W1B",
                    Comparison = "OneE compared to G&oneE",
                    df = 1)

LRT_e_ge_W4 <- LRT(path_full = "oneE/e_g_reml_W4.out", 
                    path_constrained = "oneE/e_reml_W4.out",
                    Outcome = "W4",
                    Comparison = "OneE compared to G&oneE",
                    df = 1)

LRT_e_ge <- rbind(LRT_e_ge_W1A, LRT_e_ge_W1B, LRT_e_ge_W4)
LRT_e_ge
```

***

# E. OneE and G, rGE

To estimate the rGE - we need to specify in the model that we want to calculate the correlation between the two random effects. 

## Compute rGE matrix (Q)
To do this - we first estimate Q which is caluclated as (from mtg2 manual) 
[transposed(sqrt(GRM)*sqrt(ERM)) + transposed(sqrt(GRM)*sqrt(ERMtransposed)]

We can do this in mtg2 through the following steps: 
* i. ensure positive definite kernel matrices
* ii. Cholesky decomposition
* iii. compute Q

```{r create cholesky rge txt file}
# create txt file with erm and grm cholesky
erm.grm.chol <- c("grm_full_out.grm_reformat.bend.chol", "theory1_clean_dummy_pca_output.e.bend.chol")
write.table(erm.grm.chol, paste0(ancestry_sub_path, "oneE/erm.grm.chol"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

```{r create matrix rge txt file}
# create txt file with erm, grm, covmatrix (erm.grm.chol.matmat2)
erm.grm.cov.txt <- c("grm_full_out.grm_reformat", "theory1_clean_dummy_pca_output.e", "erm.grm.chol.matmat2")
write.table(erm.grm.cov.txt, paste0(ancestry_sub_path, "oneE/erm.grm.cov.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_rge_reml_sub.txt

### Steps i, ii, iii

Note: this is a prerequisite for cholesky decomposition. See section 2.10 of the MTG2 manual for more details on NPD matrix bending.

```{bash rge positive definite}
# cd /project/rche/data/datasets/addhealth/scratch/kthompson/mtg2_output/complete_dat_nosibs/oneE/
# # GRM
# ./../../mtg2 -p IDs_IID.fam -g grm_full_out.grm_reformat -thread 80 -bend 2 # gives grm_full_out.grm_reformat.bend
# 
# # ERM
# ./../../mtg2 -p IDs_IID.fam -g theory1_clean_dummy_pca_output.e -thread 80 -bend 2 # theory1_clean_dummy_pca_output.e.bend
```

### ii. Cholesky decomposition

```{bash rge cholesky}
# cd /project/rche/data/datasets/addhealth/scratch/kthompson/mtg2_output/complete_dat_nosibs/oneE/
# # GRM
# ./../../mtg2 -p IDs_IID.fam -g grm_full_out.grm_reformat.bend -thread 80 -chol 1 # output grm_full_out.grm_reformat.bend.chol
# 
# # ERM
# ./../../mtg2 -p IDs_IID.fam -g theory1_clean_dummy_pca_output.e.bend -thread 80 -chol 1 # output theory1_clean_dummy_pca_output.e.bend.chol
```

### iii. compute Q

```{bash rge Q}
# cd /project/rche/data/datasets/addhealth/scratch/kthompson/mtg2_output/complete_dat_nosibs/oneE/
# ./../../mtg2 -p IDs_IID.fam -mg erm.grm.chol -thread 80 -matmat 2 # output erm.grm.chol.matmat2
```

## REML

Compute REML for W1A and W1B with both g and e included in the model. 

```{bash compute ereml and greml model}
# cd /project/rche/data/datasets/addhealth/scratch/kthompson/mtg2_output/complete_dat_nosibs/oneE/
# ./../../mtg2 -p IDs_IID.fam -mg erm.grm.cov.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_reml_W1A.out > e_g_cov_reml_W1A.log
# ./../../mtg2 -p IDs_IID.fam -mg erm.grm.cov.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_reml_W1B.out > e_g_cov_reml_W1B.log
# ./../../mtg2 -p IDs_IID.fam -mg erm.grm.cov.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_reml_W4.out > e_g_cov_reml_W4.log
```

To detect the covariance between random effects, we compare the one with the covariance term and the one without (step D above) using the likelihood ratio test with one degree of freedom (since the two models differ by the covariance term).

## Output (rge)

**We cannot use the normal mtg2 output files!!!** This is because including the covariance component changes the proportion of variance explained.

When including a covariance term, the total var(y) = var(s) + var(g) + 2cov(gs) + var(error), so we need to include two of the covariance estimates in the Ratio calculation and we can do this manually in R ourselves, but also using delta2 in mtg2. This is not computationally intensive and can be run in R - it is just creating files and tranforming variance components. 

We use MTG2 to estimate proporitions of phenotypic variance explained by genetics and environments, correlation between genetic and enviornmental effects on phenotypes, and standard errors of these estimates (derived using the Delta Method).

Create a file that contains variance & covariance estimates & the information matrix (i.e., var-cov matrix of model parameters). These pieces of information are inside the CORE GREML outfile, but to be readily used by MTG2, lines with ‘LKH’ & ‘h2’ need to be removed from the out file:

```{bash pick out the LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_cov_reml_W1A.out > e_g_cov_reml_W1A.out2
       grep -vwE '(LKH|h2)' e_g_cov_reml_W1B.out > e_g_cov_reml_W1B.out2
       grep -vwE '(LKH|h2)' e_g_cov_reml_W4.out > e_g_cov_reml_W4.out2
done
```

Prepare a do file for MTG2 on to construct functions of model parameters in R: 
The numbers represent the rows in the out file - corresponding estimates of V. 

* ‘R’ is to get the ratio of variance
* R 2 1 3 4 (i.e. 2 / (1 + 2 + 3 + 4))
* R 3 1 2 4 (i.e. 3 / (1 + 2 + 3 + 4))
* R 4 1 2 3 (i.e. 4 / (1 + 2 + 3 + 4))
* The format here is to put the number of the estimate you want first and then all other numbers after

So R 2 1 3 4 4 will do 2 / 1 + 2 + 3 + 4 + 4
* Cannot compute the proportion for cov(ge) because it does not let you do 2*4 in the do file. But we can work this out ourselves doing 1-all variance components (without a SE estimate).

* ‘C’ is to get the correlation of covariance
* C 5 2 3 (i.e. 5 / sqrt(2 * 3))
* Here we have the covariance estimate first, followed by the two variances used for the covariance

```{r do file for e_g_cov_reml_W1A}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_cov_reml_W1A.do")          # create file with this name
cat("e_g_cov_reml_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(4, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 4", "\n")             # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4 4", "\n")             # line 4: compute prop. of variance due to environments
cat("R 1 2 3 4 4", "\n")             # line 4: compute prop. of variance due to error
cat("C 4 2 3", "\n")                 # line 5: compute correlation between g & e
sink()
```

```{r do file for e_g_cov_reml_W1B}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_cov_reml_W1B.do")
cat("e_g_cov_reml_W1B.out2", "\n")     # line 1: specify the file with parameter estimates
cat(4, "\n")                           # line 2: tot. # of variance & covariance components in the file
cat("R 2 1 3 4 4", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4 4", "\n")               # line 4: compute prop. of variance due to environments
cat("R 1 2 3 4 4", "\n")               # line 4: compute prop. of variance due to error
cat("C 4 2 3", "\n")                   # line 5: compute correlation between g & e
sink()
```

```{r do file for e_g_cov_reml_W4}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_cov_reml_W4.do")
cat("e_g_cov_reml_W4.out2", "\n")     # line 1: specify the file with parameter estimates
cat(4, "\n")                           # line 2: tot. # of variance & covariance components in the file
cat("R 2 1 3 4 4", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4 4", "\n")               # line 4: compute prop. of variance due to environments
cat("R 1 2 3 4 4", "\n")               # line 4: compute prop. of variance due to error
cat("C 4 2 3", "\n")                   # line 5: compute correlation between g & e
sink()
```

Output using Delta

```{bash output for e_g_cov_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_cov_reml_W1A.do > e_g_cov_reml_W1A.out3
    ./../../../mtg2 -delta2 e_g_cov_reml_W1B.do > e_g_cov_reml_W1B.out3
    ./../../../mtg2 -delta2 e_g_cov_reml_W4.do > e_g_cov_reml_W4.out3
done
```

```{r extract output}
e_g_cov_reml_W1A.output <- extract_delta_output(outcome = "W1A", 
                                        model_prefix = "e_g_cov", 
                                        ancestry_sub_path,
                                        domain_labels = c("Genetic", "Environment", "Error"),
                                        return_split = FALSE) # TRUE will split out correlation estimate as table

e_g_cov_reml_W1B.output <- extract_delta_output(outcome = "W1B", 
                                        model_prefix = "e_g_cov", 
                                        ancestry_sub_path,
                                        domain_labels = c("Genetic", "Environment", "Error"),
                                        return_split = FALSE) # TRUE will split out correlation estimate as table

e_g_cov_reml_W4.output <- extract_delta_output(outcome = "W4", 
                                        model_prefix = "e_g_cov", 
                                        ancestry_sub_path,
                                        domain_labels = c("Genetic", "Environment", "Error"),
                                        return_split = FALSE) # TRUE will split out correlation estimate as table

e_g_cov_reml.output <- rbind(e_g_cov_reml_W1A.output, e_g_cov_reml_W1B.output, e_g_cov_reml_W4.output)
```

### LRT

```{r LRT test rge}
LRT_ge_rge_W1A <- LRT(path_full = "oneE/e_g_cov_reml_W1A.out", 
                      path_constrained = "oneE/e_g_reml_W1A.out",
                      Outcome = "W1A",
                      Comparison = "G&oneE compared to rGE",
                      df = 1)

LRT_ge_rge_W1B <- LRT(path_full = "oneE/e_g_cov_reml_W1B.out", 
                      path_constrained = "oneE/e_g_reml_W1B.out",
                      Outcome = "W1B",
                      Comparison = "G&oneE compared to rGE",
                      df = 1)

LRT_ge_rge_W4 <- LRT(path_full = "oneE/e_g_cov_reml_W4.out", 
                      path_constrained = "oneE/e_g_reml_W4.out",
                      Outcome = "W4",
                      Comparison = "G&oneE compared to rGE",
                     df = 1)

LRT_ge_rge <- rbind(LRT_ge_rge_W1A, LRT_ge_rge_W1B, LRT_ge_rge_W4)
LRT_ge_rge
```

Nested model is y = g + e + error (simpler, fewer predictors)
If significant (<0.05) = the full model (y = g + e + rge + error) provides a significantly better fit than the nested model
If non-significant (>0.05) = the nested model (simpler model) fits the data as well as the full model (more complex model)

### Output table

```{r variance output}
e_g_reml.output
e_g_cov_reml.output
```

```{r LRT output}
LRT_e_ge
LRT_ge_rge
```

# F. OneE and G and GxE 

```{r filter and match nopca data }
# filter to only those that are matching for erm data 
dat.erm.theory1.dummy_nopca_grm_nosibs <- as.data.frame(dat.erm.theory1.dummy) %>%
  filter(AID %in% ALL_matching_ids_nosibs) 

anyNA(dat.erm.theory1.dummy_nopca_grm_nosibs)

# take out ID 
dat.erm.theory1.dummy_nopca_grm_nosibs_noID <- as.matrix(dat.erm.theory1.dummy_nopca_grm_nosibs[,-1]) 

# have a look
head(as.data.frame(dat.erm.theory1.dummy_nopca_grm_nosibs_noID))

# scale the data
dat.erm.theory1.dummy_nopca_grm_nosibs_noID_scaled <- scale(dat.erm.theory1.dummy_nopca_grm_nosibs_noID) # normalise this matrix - minus the mean and divide by SD 

# Need to remove any NA values
dat.erm.theory1.dummy_nopca_grm_nosibs_noID_scaled <-  as.data.frame(dat.erm.theory1.dummy_nopca_grm_nosibs_noID_scaled) %>% select(where(~ !all(is.na(.))))
  
dat.erm.theory1.dummy_nopca_grm_nosibs_noID_scaled <- as.matrix(dat.erm.theory1.dummy_nopca_grm_nosibs_noID_scaled)

# get number of variables
m <- dim(dat.erm.theory1.dummy_nopca_grm_nosibs_noID_scaled)[2] # m is the number of variables - NOW 395

# divide by sqrt(m)
dat.erm.theory1.dummy_nopca_grm_nosibs_noID_scaled <- dat.erm.theory1.dummy_nopca_grm_nosibs_noID_scaled/sqrt(m)

head(as.data.frame(dat.erm.theory1.dummy_nopca_grm_nosibs_noID_scaled))

# attach additional AID variable - mtg 
dat.erm.theory1.dummy_nopca_grm_nosibs_mtg <- cbind(dat.erm.theory1.dummy_nopca_grm_nosibs$AID, dat.erm.theory1.dummy_nopca_grm_nosibs$AID, dat.erm.theory1.dummy_nopca_grm_nosibs_noID_scaled)

dat.erm.theory1.dummy_nopca_grm_nosibs_mtg <- as.data.frame(dat.erm.theory1.dummy_nopca_grm_nosibs_mtg) %>% dplyr::rename(AID2 = V2, AID = V1) 

head(as.data.frame(dat.erm.theory1.dummy_nopca_grm_nosibs_mtg))

# reorder to match genetic file 
theory1_clean_dummy_nopca_reordered <- dat.erm.theory1.dummy_nopca_grm_nosibs_mtg[order(match(dat.erm.theory1.dummy_nopca_grm_nosibs_mtg$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

head(theory1_clean_dummy_nopca_reordered)

# check they are the same order
identical(as.numeric(matching_id_grm_erm_FID$AID), theory1_clean_dummy_nopca_reordered$AID) # these should be the same - so now need to run the ERM to match and remember to use this ID file as the input file when running the model 

# make all numeric
theory1_clean_dummy_nopca_reordered_numeric <- apply(theory1_clean_dummy_nopca_reordered, 2, as.numeric)
is.matrix(theory1_clean_dummy_nopca_reordered_numeric) && all(sapply(theory1_clean_dummy_nopca_reordered_numeric, is.numeric)) # check matrix is numeric

# have a look at data file
as.tibble(theory1_clean_dummy_nopca_reordered_numeric)
head(matching_id_grm_erm_AID)

# write table 
write.table(theory1_clean_dummy_nopca_reordered_numeric, paste0(ancestry_sub_path, "oneE/theory1_clean_dummy_nopca"), col.names = FALSE, row.names = FALSE, quote = FALSE)
```

```{r create e_g txt file}
# create txt file with erm and grm
erm.grm.gxe.txt <- c("grm_full_out.grm_reformat", "theory1_clean_dummy_pca_output.e", "erm_grm.gxe") 
write.table(erm.grm.gxe.txt, paste0(ancestry_sub_path, "oneE/erm.grm.gxe.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

Create relatedness matrix without pca transformation and caluclate interaction term between ERM and GRM 

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_e_g_gxe_reml_sub.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann:

```{bash e_g_gxe_reml}
# directories=(
#       "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
#       "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
#       "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
#       "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
#       "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
#  )
# 
# for dir in "${directories[@]}"; do
#       echo "Processing: $dir"
#       cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
# 
#       # Create relationship matrix for non-transformed ERM
#       ./../../../mtg2 -p IDs_IID.fam -pdmx theory1_clean_dummy_nopca -thread 100 -out theory1_clean_dummy_nopca_output.e
#       
#       # create interaction term from erm x grm 
#       paste theory1_clean_dummy_nopca_output.e grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > erm_grm.gxe
#       
#       # include in REML model 
#       ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.gxe.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_gxe_reml_W1A.out > e_g_gxe_reml_W1A.log
#       ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.gxe.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_gxe_reml_W1B.out > e_g_gxe_reml_W1B.log
#       ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.gxe.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_gxe_reml_W4.out > e_g_gxe_reml_W4.log
#       
# done
```

### Output 

```{bash pick out LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_gxe_reml_W1A.out > e_g_gxe_reml_W1A.out2
       grep -vwE '(LKH|h2)' e_g_gxe_reml_W1B.out > e_g_gxe_reml_W1B.out2
       grep -vwE '(LKH|h2)' e_g_gxe_reml_W4.out > e_g_gxe_reml_W4.out2
done
```

```{r do file for e_g_gxe_reml_W1A}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_gxe_reml_W1A.do")          # create file with this name
cat("e_g_gxe_reml_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(4, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4", "\n")               # line 4: compute prop. of variance due to environments
cat("R 4 1 2 3", "\n")               # line 4: compute prop. of variance due to gxe
cat("R 1 2 3 4", "\n")               # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_gxe_reml_W1B}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_gxe_reml_W1B.do")
cat("e_g_gxe_reml_W1B.out2", "\n")     # line 1: specify the file with parameter estimates
cat(4, "\n")                           # line 2: tot. # of variance & gxeariance components in the file
cat("R 2 1 3 4", "\n")                 # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4", "\n")                 # line 4: compute prop. of variance due to environments
cat("R 4 1 2 3", "\n")                 # line 4: compute prop. of variance due to gxe
cat("R 1 2 3 4", "\n")                 # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_gxe_reml_W4}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_gxe_reml_W4.do")
cat("e_g_gxe_reml_W4.out2", "\n")      # line 1: specify the file with parameter estimates
cat(4, "\n")                           # line 2: tot. # of variance & gxeariance components in the file
cat("R 2 1 3 4", "\n")                 # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4", "\n")                 # line 4: compute prop. of variance due to environments
cat("R 4 1 2 3", "\n")                 # line 4: compute prop. of variance due to gxe
cat("R 1 2 3 4", "\n")                 # line 4: compute prop. of variance due to error
sink()
```

Output using Delta

```{bash output for e_g_gxe_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_gxe_reml_W1A.do > e_g_gxe_reml_W1A.out3
    ./../../../mtg2 -delta2 e_g_gxe_reml_W1B.do > e_g_gxe_reml_W1B.out3
    ./../../../mtg2 -delta2 e_g_gxe_reml_W4.do > e_g_gxe_reml_W4.out3
done
```

```{r gxe output}
e_g_gxe_reml_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "e_g_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error")
)

e_g_gxe_reml_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "e_g_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error")
)

e_g_gxe_reml_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "e_g_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error")
)

e_g_gxe_reml.output <- rbind(e_g_gxe_reml_W1A.output, e_g_gxe_reml_W1B.output, e_g_gxe_reml_W4.output)
```

### LRT

```{r LRT test rge}
LRT_ge_gxe_W1A <- LRT(path_full = "oneE/e_g_gxe_reml_W1A.out", 
                      path_constrained = "oneE/e_g_reml_W1A.out",
                      Outcome = "W1A",
                      Comparison = "G&oneE compared to GxE",
                      df = 1)

LRT_ge_gxe_W1B <- LRT(path_full = "oneE/e_g_gxe_reml_W1B.out", 
                      path_constrained = "oneE/e_g_reml_W1B.out",
                      Outcome = "W1B",
                      Comparison = "G&oneE compared to GxE",
                      df = 1)

LRT_ge_gxe_W4 <- LRT(path_full = "oneE/e_g_gxe_reml_W4.out", 
                      path_constrained = "oneE/e_g_reml_W4.out",
                      Outcome = "W4",
                      Comparison = "G&oneE compared to GxE",
                     df = 1)

LRT_ge_gxe <- rbind(LRT_ge_gxe_W1A, LRT_ge_gxe_W1B, LRT_ge_gxe_W4)
LRT_ge_gxe
```

### Output table

```{r output table for coveg model}
e_g_gxe_reml.output
```

```{r contributions for GxE}
e_g_gxe_reml.output %>%
  mutate(contribution = variance_std/variance_std_tot*100)
```

# G. OneE and G, rGE AND GxE - SUPPLEMENT

```{r create e_g txt file}
# create txt file with erm and grm
erm.grm.rge.gxe.txt <- c("grm_full_out.grm_reformat", "theory1_clean_dummy_pca_output.e", "erm_grm.gxe", "erm.grm.chol.matmat2") 
write.table(erm.grm.rge.gxe.txt, paste0(ancestry_sub_path, "oneE/erm.grm.rge.gxe.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_e_g_cov_gxe_reml_sub.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann:

```{bash e_g_cov_gxe_reml}
#  directories=(
#        "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
#        "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
#        "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
#        "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
#        "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
#   )
#  
#  for dir in "${directories[@]}"; do
#        echo "Processing: $dir"
#        cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
#        
#        # REML model 
#        ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.rge.gxe.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_gxe_reml_W1A.out > e_g_cov_gxe_reml_W1A.log
#        ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.rge.gxe.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_gxe_reml_W1B.out > e_g_cov_gxe_reml_W1B.log
#        ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.rge.gxe.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_gxe_reml_W4.out > e_g_cov_gxe_reml_W4.log
#        
# done
```

### Output 

```{bash pick out LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_cov_gxe_reml_W1A.out > e_g_cov_gxe_reml_W1A.out2
       grep -vwE '(LKH|h2)' e_g_cov_gxe_reml_W1B.out > e_g_cov_gxe_reml_W1B.out2
       grep -vwE '(LKH|h2)' e_g_cov_gxe_reml_W4.out > e_g_cov_gxe_reml_W4.out2
done
```

```{r do file for e_g_cov_gxe_reml_W1A}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_cov_gxe_reml_W1A.do")          # create file with this name
cat("e_g_cov_gxe_reml_W1A.out2", "\n")   # specify the file with parameter estimates (above)
cat(5, "\n")                             # Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 5", "\n")               # compute prop. of variance due to genetics
cat("R 3 1 2 4 5 5", "\n")               # compute prop. of variance due to environments
cat("R 4 1 2 3 5 5", "\n")               # compute prop. of variance due to gxe
cat("R 1 2 3 4 5 5", "\n")               # compute prop. of variance due to error
cat("C 5 2 3", "\n")                     # compute correlation between g & e
sink()
```

```{r do file for e_g_cov_gxe_reml_W1B}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_cov_gxe_reml_W1B.do")          # create file with this name
cat("e_g_cov_gxe_reml_W1B.out2", "\n")   # specify the file with parameter estimates (above)
cat(5, "\n")                             # Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 5", "\n")               # compute prop. of variance due to genetics
cat("R 3 1 2 4 5 5", "\n")               # compute prop. of variance due to environments
cat("R 4 1 2 3 5 5", "\n")               # compute prop. of variance due to gxe
cat("R 1 2 3 4 5 5", "\n")               # compute prop. of variance due to error
cat("C 5 2 3", "\n")                     # compute correlation between g & e
sink()
```

```{r do file for e_g_cov_gxe_reml_W4}
setwd(paste0(ancestry_sub_path, "oneE/"))
sink("e_g_cov_gxe_reml_W4.do")          # create file with this name
cat("e_g_cov_gxe_reml_W4.out2", "\n")   # specify the file with parameter estimates (above)
cat(5, "\n")                             # Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 5", "\n")               # compute prop. of variance due to genetics
cat("R 3 1 2 4 5 5", "\n")               # compute prop. of variance due to environments
cat("R 4 1 2 3 5 5", "\n")               # compute prop. of variance due to gxe
cat("R 1 2 3 4 5 5", "\n")               # compute prop. of variance due to error
cat("C 5 2 3", "\n")                     # compute correlation between g & e
sink()
```

Output using Delta

```{bash output for e_g_gxe_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/oneE/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/oneE/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_cov_gxe_reml_W1A.do > e_g_cov_gxe_reml_W1A.out3
    ./../../../mtg2 -delta2 e_g_cov_gxe_reml_W1B.do > e_g_cov_gxe_reml_W1B.out3
    ./../../../mtg2 -delta2 e_g_cov_gxe_reml_W4.do > e_g_cov_gxe_reml_W4.out3
done
```

```{r gxe output}
e_g_cov_gxe_reml_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "e_g_cov_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error"),
  return_split = FALSE
)

e_g_cov_gxe_reml_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "e_g_cov_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error")
)

e_g_cov_gxe_reml_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "e_g_cov_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error")
)

e_g_cov_gxe_reml.output <- rbind(e_g_cov_gxe_reml_W1A.output, e_g_cov_gxe_reml_W1B.output, e_g_cov_gxe_reml_W4.output)
```

### LRT

```{r LRT test rge}
LRT_gxe_covgxe_W1A <- LRT(path_full = "oneE/e_g_cov_gxe_reml_W1A.out", 
                      path_constrained = "oneE/e_g_gxe_reml_W1A.out",
                      Outcome = "W1A",
                      Comparison = "GxE compared to GxE and rGE",
                      df = 1)

LRT_gxe_covgxe_W1B <- LRT(path_full = "oneE/e_g_cov_gxe_reml_W1B.out", 
                      path_constrained = "oneE/e_g_gxe_reml_W1B.out",
                      Outcome = "W1B",
                      Comparison = "GxE compared to GxE and rGE",
                      df = 1)

LRT_gxe_covgxe_W4 <- LRT(path_full = "oneE/e_g_cov_gxe_reml_W4.out", 
                      path_constrained = "oneE/e_g_gxe_reml_W4.out",
                      Outcome = "W4",
                      Comparison = "GxE compared to GxE and rGE",
                      df = 1)

LRT_gxe_covgxe <- rbind(LRT_gxe_covgxe_W1A, LRT_gxe_covgxe_W1B, LRT_gxe_covgxe_W4)
LRT_gxe_covgxe
```

### Output table

```{r output table for coveg model}
e_g_gxe_reml.output
e_g_cov_gxe_reml.output
```

***

# Output for all oneE models (B-G)

```{r oneE output variances}
oneE.output <- rbind(e_g_reml.output, e_g_cov_reml.output, e_g_gxe_reml.output, e_g_cov_gxe_reml.output)
oneE.output
```

```{r oneE output LRT}
LRT_oneE <- rbind(LRT_e_ge, LRT_ge_rge, LRT_ge_gxe, LRT_gxe_covgxe) 
LRT_oneE <- LRT_oneE %>% mutate(across(where(is.numeric), ~round(.x, 3)))

LRT_oneE
```

***

# H. Domains

## Define domains

First we will group the variables into their domains and save this as a new txt file. 

```{r select demographic items}
# demographics
dat.erm.theory1.dummy_dem <- dat.erm.theory1.dummy %>%
  select(
    AID,
    contains("PA55") |
    contains("PA56_numeric") |
    contains("PA59_numeric") |
    contains("H1NB5_numeric") |
    contains("H1NB6_numeric") |
    contains("H1GH27I_numeric") 
    )

write.table(dat.erm.theory1.dummy_dem, paste0(ancestry_sub_path, "domains/theory1_clean_dummy_dem.txt"), row.names = FALSE) 
```

```{r select health items}
# health
dat.erm.theory1.dummy_health <- dat.erm.theory1.dummy %>%
  select(
    AID,
    contains("H1GH1_numeric") |
    contains("H1PL1_numeric") |
    contains("H1GH2_numeric") |
    contains("H1GH3_numeric") |
    contains("H1GH4_numeric") |
    contains("H1GH5_numeric") |
    contains("H1GH6_numeric") |
    contains("H1GH7_numeric") |
    contains("H1GH8_numeric") |
    contains("H1GH9_numeric") |
    contains("H1GH10_numeric") |
    contains("H1GH11_numeric") |
    contains("H1GH12_numeric") |
    contains("H1GH13_numeric") |
    contains("H1GH14_numeric") |
    contains("H1GH15_numeric") |
    contains("H1GH16_numeric") |
    contains("H1GH17_numeric") |
    contains("H1GH18_numeric") |
    contains("H1GH19_numeric") |
    contains("H1GH22_numeric") |
    contains("H1GH26_numeric") |
    contains("H1GH28_numeric") |
    contains("H1GH29_numeric") |
    contains("H1HS3_numeric") |
    contains("H1GH48_numeric") |
    contains("H1GH49_numeric") 
    )

write.table(dat.erm.theory1.dummy_health, paste0(ancestry_sub_path, "domains/theory1_clean_dummy_health.txt"), row.names = FALSE) 

# general mental and physical health
dat.erm.theory1.dummy_health_general <- dat.erm.theory1.dummy %>%
  select(
    AID,
    contains("H1GH1_numeric") |
    contains("H1GH26_numeric") |
    contains("H1GH48_numeric") |
    contains("H1GH49_numeric") 
    )

write.table(dat.erm.theory1.dummy_health_general, paste0(ancestry_sub_path, "domains/theory1_clean_dummy_health_general.txt"), row.names = FALSE) 

# mental health - symptoms of depression or anxiety
dat.erm.theory1.dummy_health_mental <- dat.erm.theory1.dummy %>%
  select(
    AID,
    contains("H1GH8_numeric") |
    contains("H1GH18_numeric") |
    contains("H1GH19_numeric") |
    contains("H1GH22_numeric") |
    contains("H1HS3_numeric")
    )

write.table(dat.erm.theory1.dummy_health_mental, paste0(ancestry_sub_path, "domains/theory1_clean_dummy_health_mental.txt"), row.names = FALSE) 

# physical health 
dat.erm.theory1.dummy_health_physical <- dat.erm.theory1.dummy %>%
  select(
    AID,
    contains("H1PL1_numeric") |
    contains("H1GH2_numeric") |
    contains("H1GH3_numeric") |
    contains("H1GH4_numeric") |
    contains("H1GH5_numeric") |
    contains("H1GH6_numeric") |
    contains("H1GH7_numeric") |
    contains("H1GH9_numeric") |
    contains("H1GH10_numeric") |
    contains("H1GH11_numeric") |
    contains("H1GH12_numeric") |
    contains("H1GH13_numeric") |
    contains("H1GH14_numeric") |
    contains("H1GH15_numeric") |
    contains("H1GH16_numeric") |
    contains("H1GH17_numeric") |
    contains("H1GH28_numeric") |
    contains("H1GH29_numeric")
    )

write.table(dat.erm.theory1.dummy_health_physical, paste0(ancestry_sub_path, "domains/theory1_clean_dummy_health_physical.txt"), row.names = FALSE) 
```
         
```{r select difficulties at school items}
# Difficulties at school
dat.erm.theory1.dummy_school <- dat.erm.theory1.dummy %>%
  select(
    AID,
    contains("H1ED15_numeric") |
    contains("H1ED16_numeric") |
    contains("H1ED17_numeric") |
    contains("H1ED18_numeric") |
    contains("H1ED19_numeric") |
    contains("H1ED20_numeric") |
    contains("H1ED21_numeric") |
    contains("H1ED22_numeric") |
    contains("H1ED23_numeric") |
    contains("H1ED24_numeric") 
    )

write.table(dat.erm.theory1.dummy_school, paste0(ancestry_sub_path, "domains/theory1_clean_dummy_school.txt"), row.names = FALSE) 
```

```{r select friends and family items}
# relationships with friends and family
dat.erm.theory1.dummy_frfam <- dat.erm.theory1.dummy %>%
  select(
    AID,
    contains("H1DA7_numeric") |
    contains("H1WP9_numeric") |
    contains("H1WP10_numeric") |
    contains("H1WP13_numeric") |
    contains("H1WP14_numeric") |
    contains("H1PF1_numeric") |
    contains("H1PF4_numeric") |
    contains("H1PF5_numeric") |
    contains("H1PF24_numeric") |
    contains("H1PF25_numeric") |
    contains("H1PR1_numeric") |
    contains("H1PR2_numeric") |
    contains("H1PR3_numeric") |
    contains("H1PR4_numeric") |
    contains("H1PR5_numeric") |
    contains("H1PR6_numeric") |
    contains("H1PR7_numeric") |
    contains("H1PR8_numeric") 
    )

write.table(dat.erm.theory1.dummy_frfam, paste0(ancestry_sub_path, "domains/theory1_clean_dummy_frfam.txt"), row.names = FALSE) 
```

```{r select positive cognition items}
# positive cognition and social acceptance  
dat.erm.theory1.dummy_poscog <- dat.erm.theory1.dummy %>%
  select(
    AID,
    contains("H1PF33_numeric") |
    contains("H1PF34_numeric") |
    contains("H1PF35_numeric") |
    contains("H1PF36_numeric") 
    )

write.table(dat.erm.theory1.dummy_poscog, paste0(ancestry_sub_path, "domains/theory1_clean_dummy_poscog.txt"), row.names = FALSE)
```

```{r select stressful life events ERM data}
# stressful life events / traumatic experiences
dat.erm.theory1.dummy_sle <- dat.erm.theory1.dummy %>%
  select(
    AID,
    contains("H1CO10_numeric") |
    contains("H1NR3_numeric") |
    contains("H1FV1_numeric") |
    contains("H1FV2_numeric") |
    contains("H1FV3_numeric") |
    contains("H1FV4_numeric") |
    contains("H1FV5_numeric") |
    contains("H1FV6_numeric") |
    contains("H1FP21_1_numeric") |
    contains("H1FP21_2_numeric") |
    contains("H1FP21_3_numeric") |
    contains("H1SU4_numeric") |
    contains("H1SU5_numeric") |
    contains("H1SU6_numeric") |
    contains("H1SU7_numeric") 
    )

write.table(dat.erm.theory1.dummy_sle, paste0(ancestry_sub_path, "domains/theory1_clean_dummy_sle.txt"), row.names = FALSE) 
```

PCA transformation

Then we create another version of these files that includes the PCA transformation for each file, adds the additional ID variable, and subsets the sample to the 6157 that match the genetic files. 

```{r loop for PCA for domains}
input_dir <- paste0(ancestry_sub_path, "domains/")

# List all .txt files in the input directory
txt_files <- list.files(input_dir, pattern = "^theory.*\\.txt$", full.names = TRUE)

# Loop over each txt file
for (file in txt_files) {
  # Step 1: Read the current file into a dataframe
  current_df <- read.table(file, header = TRUE)

  # Step 2: Apply the dummy coding function
  dat_dummy <- current_df # dummy coding has already been done
  dummy_names <- setdiff(names(dat_dummy), "AID")

  # Step 3: PCA Transformation
  dat_dummy_noID <- as.matrix(dat_dummy[, -1])  # Exclude AID for PCA
  dummy_covmat <- var(dat_dummy_noID)           # Covariance matrix
  eig <- eigen(dummy_covmat)                    # Eigen decomposition
  pca_vector <- dat_dummy_noID %*% eig$vectors  # PCA transformation
  pca_scaled <- scale(pca_vector)               # Normalize the matrix
  m <- dim(pca_scaled)[2]
  pca <- pca_scaled / sqrt(m)                   # Normalize by sqrt(m)
  pca_df <- as.data.frame(pca)
  colnames(pca_df) <- dummy_names

  # Step 4: Attach the AID variable back
  pca_df <- cbind(AID = dat_dummy$AID, AID2 = dat_dummy$AID, pca_df)

  # Step 5: Apply corrections (filter and attach ID variable) - removed to include full sample
  filtered_df <- pca_df %>%
    filter(AID %in% ALL_matching_ids_nosibs)   # Filter based on matching IDs

  # Convert to numeric matrix (ensures all values are numeric)
  filtered_numeric <- filtered_df %>%
    mutate(AID = as.numeric(AID)) %>%
    mutate(AID2 = as.numeric(AID2))
  
  filtered_numeric <- filtered_numeric[, colSums(!is.na(filtered_numeric)) > 0]

  # Reorder the file to match the ID file
  filtered_reordered <- filtered_numeric[order(match(filtered_numeric$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

  # Save the current dataframe to a file
  temp_filename <- file.path(input_dir, paste0("dat_", tools::file_path_sans_ext(basename(file))))
  write.table(filtered_reordered, temp_filename, row.names = FALSE, col.names = FALSE, quote = FALSE, sep="\t")
}
```

## 1. Health separated (with genetics)

### REML

```{r create joint txt file}
# create txt file with all erms
erm.grm.domains.health.txt <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat")
  
write.table(erm.grm.domains.health.txt, paste0(ancestry_sub_path, "domains/erm.grm.domains.health.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_g_e_reml_dom_health_sub.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann:

```{bash g_e_reml_dom_health}
# directories=(
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
# )
# 
# 
# # Loop through each directory
# for dir in "${directories[@]}"; do
#     echo "Processing directory: $dir"
#     
#     # Change to the target directory
#     cd "$dir" || { echo "Failed to enter directory: $dir"; continue; }
#     
#     # cp ../oneE/grm_full_out.grm_reformat ../domains/grm_full_out.grm_reformat
# 
#     # Run REML only after all ERM outputs are generated
#     echo "Running REML for directory: $dir"
#     ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_W1A.out > domain_reml_ge_health_W1A.log
#     ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_W1B.out > domain_reml_ge_health_W1B.log
#     ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_W4.out > domain_reml_ge_health_W4.log
# done
```

### Delta output prep

```{r do file for e_g_cov_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_W1A.do")                      # create file with this name
cat("domain_reml_ge_health_W1A.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(10, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10", "\n")       # line 3: compute prop. of variance due to dem
cat("R 3 1 2 4 5 6 7 8 9 10", "\n")       # line 4: compute prop. of variance due to frfam
cat("R 4 1 2 3 5 6 7 8 9 10", "\n")       # line 5: compute prop. of variance due to poscog
cat("R 5 1 2 3 4 6 7 8 9 10", "\n")       # line 6: compute prop. of variance due to school
cat("R 6 1 2 3 4 5 7 8 9 10", "\n")       # line 7: compute prop. of variance due to sle
cat("R 7 1 2 3 4 5 6 8 9 10", "\n")       # line 8: compute prop. of variance due to hgeneral
cat("R 8 1 2 3 4 5 6 7 9 10", "\n")       # line 9: compute prop. of variance due to hmental
cat("R 9 1 2 3 4 5 6 7 8 10", "\n")       # line 10: compute prop. of variance due to hphysical
cat("R 10 1 2 3 4 5 6 7 8 9", "\n")       # line 11: compute prop. of variance due to genetics
cat("R 1 2 3 4 5 6 7 8 9 10", "\n")        # line 11: compute prop. of variance due to error
sink()
```

```{r do file for e_g_cov_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_W1B.do")                      # create file with this name
cat("domain_reml_ge_health_W1B.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(10, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10", "\n")       # line 3: compute prop. of variance due to dem
cat("R 3 1 2 4 5 6 7 8 9 10", "\n")       # line 4: compute prop. of variance due to frfam
cat("R 4 1 2 3 5 6 7 8 9 10", "\n")       # line 5: compute prop. of variance due to poscog
cat("R 5 1 2 3 4 6 7 8 9 10", "\n")       # line 6: compute prop. of variance due to school
cat("R 6 1 2 3 4 5 7 8 9 10", "\n")       # line 7: compute prop. of variance due to sle
cat("R 7 1 2 3 4 5 6 8 9 10", "\n")       # line 8: compute prop. of variance due to hgeneral
cat("R 8 1 2 3 4 5 6 7 9 10", "\n")       # line 9: compute prop. of variance due to hmental
cat("R 9 1 2 3 4 5 6 7 8 10", "\n")       # line 10: compute prop. of variance due to hphysical
cat("R 10 1 2 3 4 5 6 7 8 9", "\n")       # line 11: compute prop. of variance due to genetics
cat("R 1 2 3 4 5 6 7 8 9 10", "\n")        # line 11: compute prop. of variance due to error
sink()
```

```{r do file for e_g_cov_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_W4.do")                      # create file with this name
cat("domain_reml_ge_health_W4.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(10, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10", "\n")       # line 3: compute prop. of variance due to dem
cat("R 3 1 2 4 5 6 7 8 9 10", "\n")       # line 4: compute prop. of variance due to frfam
cat("R 4 1 2 3 5 6 7 8 9 10", "\n")       # line 5: compute prop. of variance due to poscog
cat("R 5 1 2 3 4 6 7 8 9 10", "\n")       # line 6: compute prop. of variance due to school
cat("R 6 1 2 3 4 5 7 8 9 10", "\n")       # line 7: compute prop. of variance due to sle
cat("R 7 1 2 3 4 5 6 8 9 10", "\n")       # line 8: compute prop. of variance due to hgeneral
cat("R 8 1 2 3 4 5 6 7 9 10", "\n")       # line 9: compute prop. of variance due to hmental
cat("R 9 1 2 3 4 5 6 7 8 10", "\n")       # line 10: compute prop. of variance due to hphysical
cat("R 10 1 2 3 4 5 6 7 8 9", "\n")       # line 11: compute prop. of variance due to genetics
cat("R 1 2 3 4 5 6 7 8 9 10", "\n")        # line 11: compute prop. of variance due to error
sink()
```

```{bash output for e_g_cov_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
     echo "Processing: $dir"
     cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
     grep -vwE '(LKH|h2)' domain_reml_ge_health_W1A.out > domain_reml_ge_health_W1A.out2
     grep -vwE '(LKH|h2)' domain_reml_ge_health_W1B.out > domain_reml_ge_health_W1B.out2
     grep -vwE '(LKH|h2)' domain_reml_ge_health_W4.out > domain_reml_ge_health_W4.out2

     ./../../../mtg2 -delta2 domain_reml_ge_health_W1A.do > domain_reml_ge_health_W1A.out3
     ./../../../mtg2 -delta2 domain_reml_ge_health_W1B.do > domain_reml_ge_health_W1B.out3
     ./../../../mtg2 -delta2 domain_reml_ge_health_W4.do > domain_reml_ge_health_W4.out3
done
```

### Output table

```{r output table for health w g}
domain_reml_ge_health_W1A.output.est.df <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "domain_reml_ge_health",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School",
                    "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", "Error")
)

domain_reml_ge_health_W1B.output.est.df <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "domain_reml_ge_health",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School",
                    "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", "Error")
)

domain_reml_ge_health_W4.output.est.df <- extract_delta_output(
  outcome = "W4",
  model_prefix = "domain_reml_ge_health",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School",
                    "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", "Error")
)

domain_reml_ge_health.output <- rbind(domain_reml_ge_health_W1A.output.est.df, domain_reml_ge_health_W1B.output.est.df, domain_reml_ge_health_W4.output.est.df)
domain_reml_ge_health.output
```

```{r significant domains only}
domain_reml_ge_health.output %>% filter(P_value < 0.05)

domain_reml_ge_health.output %>% select(outcome, domain, input, variance_std, P_value)
```

## 2. Health separated - dropped

### A. Dropped non-sig parameters

Significant domains:

W1A: Friends and family, Positive cognition, School, General health, 	Mental health, 	Physical health, Error
W1B: Friends and family, Positive cognition, School, General health, 	Mental health, 	Physical health, Genetic, Error
W4:  Demographics, Positive cognition, Stressful life events, Mental health, Physical health, Error

```{r create joint txt file}
# W1A
erm.grm.domains.health_droppedW1A.txt <- c("erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical")
  
write.table(erm.grm.domains.health_droppedW1A.txt, paste0(ancestry_sub_path, "domains/erm.grm.domains.health_droppedW1A.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# W1B 
erm.grm.domains.health_droppedW1B.txt <- c("erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat")
  
write.table(erm.grm.domains.health_droppedW1B.txt, paste0(ancestry_sub_path, "domains/erm.grm.domains.health_droppedW1B.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# W4
erm.grm.domains.health_droppedW4.txt <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical")
  
write.table(erm.grm.domains.health_droppedW4.txt, paste0(ancestry_sub_path, "domains/erm.grm.domains.health_droppedW4.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_g_e_reml_dom_health_sub_dropped.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann:

#### REML

```{bash g_e_reml_dom_health_sub_dropped}
# directories=(
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
# )
# 
# 
# # Loop through each directory
# for dir in "${directories[@]}"; do
#     echo "Processing directory: $dir"
#     
#     # Change to the target directory
#     cd "$dir" || { echo "Failed to enter directory: $dir"; continue; }
# 
#     # Run REML only after all ERM outputs are generated
#     echo "Running REML for directory: $dir"
#     ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health_droppedW1AB.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_dropped_W1A.out > domain_reml_ge_health_dropped_W1A.log
#     ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health_droppedW1AB.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_dropped_W1B.out > domain_reml_ge_health_dropped_W1B.log
#     ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health_droppedW4.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_dropped_W4.out > domain_reml_ge_health_dropped_W4.log
# done
```

#### Delta output prep

```{r do file for e_g_cov_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_dropped_W1A.do")                      # create file with this name
cat("domain_reml_ge_health_dropped_W1A.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(7, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7", "\n")       # line 3: compute prop. of variance due to frfam
cat("R 3 1 2 4 5 6 7", "\n")       # line 4: compute prop. of variance due to poscog
cat("R 4 1 2 3 5 6 7", "\n")       # line 5: compute prop. of variance due to school
cat("R 5 1 2 3 4 6 7", "\n")       # line 6: compute prop. of variance due to hgeneral
cat("R 6 1 2 3 4 5 7", "\n")       # line 7: compute prop. of variance due to hmental
cat("R 7 1 2 3 4 5 6", "\n")       # line 8: compute prop. of variance due to hphysical
cat("R 1 2 3 4 5 6 7", "\n")        # line 11: compute prop. of variance due to error
sink()
```

```{r do file for e_g_cov_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_dropped_W1B.do")                      # create file with this name
cat("domain_reml_ge_health_dropped_W1B.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(8, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8", "\n")       # line 3: compute prop. of variance due to frfam
cat("R 3 1 2 4 5 6 7 8", "\n")       # line 4: compute prop. of variance due to poscog
cat("R 4 1 2 3 5 6 7 8", "\n")       # line 5: compute prop. of variance due to school
cat("R 5 1 2 3 4 6 7 8", "\n")       # line 6: compute prop. of variance due to hgeneral
cat("R 6 1 2 3 4 5 7 8", "\n")       # line 7: compute prop. of variance due to hmental
cat("R 7 1 2 3 4 5 6 8", "\n")       # line 8: compute prop. of variance due to hphysical
cat("R 8 1 2 3 4 5 6 7", "\n")       # line 8: compute prop. of variance due to genetic
cat("R 1 2 3 4 5 6 7 8", "\n")        # line 11: compute prop. of variance due to error
sink()
```

```{r do file for e_g_cov_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_dropped_W4.do")                      # create file with this name
cat("domain_reml_ge_health_dropped_W4.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(6, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6", "\n")       # line 3: compute prop. of variance due to dem
cat("R 3 1 2 4 5 6", "\n")       # line 4: compute prop. of variance due to poscog
cat("R 4 1 2 3 5 6", "\n")       # line 5: compute prop. of variance due to sle
cat("R 5 1 2 3 4 6", "\n")       # line 6: compute prop. of variance due to hmental
cat("R 6 1 2 3 4 5", "\n")       # line 7: compute prop. of variance due to hphysical
cat("R 1 2 3 4 5 6", "\n")       # line 11: compute prop. of variance due to error
sink()
```

```{bash output for e_g_cov_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
     echo "Processing: $dir"
     cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
     grep -vwE '(LKH|h2)' domain_reml_ge_health_dropped_W1A.out > domain_reml_ge_health_dropped_W1A.out2
     grep -vwE '(LKH|h2)' domain_reml_ge_health_dropped_W1B.out > domain_reml_ge_health_dropped_W1B.out2
     grep -vwE '(LKH|h2)' domain_reml_ge_health_dropped_W4.out > domain_reml_ge_health_dropped_W4.out2

     ./../../../mtg2 -delta2 domain_reml_ge_health_dropped_W1A.do > domain_reml_ge_health_dropped_W1A.out3
     ./../../../mtg2 -delta2 domain_reml_ge_health_dropped_W1B.do > domain_reml_ge_health_dropped_W1B.out3
     ./../../../mtg2 -delta2 domain_reml_ge_health_dropped_W4.do > domain_reml_ge_health_dropped_W4.out3
done
```

#### Output table 

```{r output for dropped domains}
domain_reml_ge_health_dropped_W1A.output.est.df <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "domain_reml_ge_health_dropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", "Error")
)

domain_reml_ge_health_dropped_W1B.output.est.df <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "domain_reml_ge_health_dropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", "Genetic", "Error")
)

domain_reml_ge_health_dropped_W4.output.est.df <- extract_delta_output(
  outcome = "W4",
  model_prefix = "domain_reml_ge_health_dropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Positive cognition", "Stressful life events", "Mental health", "Physical health", "Error")
)

domain_reml_ge_health_dropped.output <- rbind(domain_reml_ge_health_dropped_W1A.output.est.df, domain_reml_ge_health_dropped_W1B.output.est.df, domain_reml_ge_health_dropped_W4.output.est.df)

domain_reml_ge_health_dropped.output %>%
  mutate(contribution = variance_std/variance_std_tot*100)
```

#### LRT

Number of parameters dropped: 
W1A = Demographics + Stressful life events + Genetic = 3
W1B = Demographics + Stressful life events = 2
W4  = Friends and family + School + General health + Genetic = 4

```{r LRT reduced E covE}
LRT_dome_domdrop_W1A <- LRT(path_full = "domains/domain_reml_ge_health_W1A.out", 
                    path_constrained = "domains/domain_reml_ge_health_dropped_W1A.out",
                    Outcome = "W1A",
                    Comparison = "Domains reduced E compared to full Domains",
                    df = 3)

LRT_dome_domdrop_W1B <- LRT(path_full = "domains/domain_reml_ge_health_W1B.out", 
                    path_constrained = "domains/domain_reml_ge_health_dropped_W1B.out",
                    Outcome = "W1B",
                    Comparison = "Domains reduced E compared to full Domains",
                    df = 2)

LRT_dome_domdrop_W4 <- LRT(path_full = "domains/domain_reml_ge_health_W4.out", 
                    path_constrained = "domains/domain_reml_ge_health_dropped_W4.out",
                    Outcome = "W4",
                    Comparison = "Domains reduced E compared to full Domains",
                    df = 4)

LRT_dome_domdrop <- rbind(LRT_dome_domdrop_W1A, LRT_dome_domdrop_W1B, LRT_dome_domdrop_W4)
LRT_dome_domdrop
```

### B. Sequentially drop based on effect size - SUPPLEMENT

We will drop parameters one at a time in order of effect size and significance

W1A: Demographics (non-sig), Stressful life events (non-sig just), Genetic (non-sig but this is reduced model above with all three dropped)

```{r create joint txt file}
# demographics dropped
erm.grm.domains.health_demdroppedW1A.txt <- c("erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat")
  
write.table(erm.grm.domains.health_demdroppedW1A.txt, paste0(ancestry_sub_path, "domains/erm.grm.domains.health_demdroppedW1A.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# dem and sle dropped 
erm.grm.domains.health_demsledroppedW1A.txt <- c("erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat")
  
write.table(erm.grm.domains.health_demsledroppedW1A.txt, paste0(ancestry_sub_path, "domains/erm.grm.domains.health_demsledroppedW1A.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

W1B: Stressful life events (non-sig), Demographics (non-sig this is same as model above)

```{r create joint txt file}
# sle dropped
erm.grm.domains.health_sledroppedW1B.txt <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat")
  
write.table(erm.grm.domains.health_sledroppedW1B.txt, paste0(ancestry_sub_path, "domains/erm.grm.domains.health_sledroppedW1B.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

W4: General health (non-sig), School (non-sig), Friends and family (non-sig), Genetic (non-sig same as model above)

```{r create joint txt file}
# general health dropped
erm.grm.domains.health_ghealthdroppedW4.txt <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat")
  
write.table(erm.grm.domains.health_ghealthdroppedW4.txt, paste0(ancestry_sub_path, "domains/erm.grm.domains.health_ghealthdroppedW4.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# ghealth and school dropped 
erm.grm.domains.health_ghealthschdroppedW4.txt <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat")
  
write.table(erm.grm.domains.health_ghealthschdroppedW4.txt, paste0(ancestry_sub_path, "domains/erm.grm.domains.health_ghealthschdroppedW4.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# ghealth and school and frfam dropped  
erm.grm.domains.health_ghealthschfrfdroppedW4.txt <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat")
  
write.table(erm.grm.domains.health_ghealthschfrfdroppedW4.txt, paste0(ancestry_sub_path, "domains/erm.grm.domains.health_ghealthschfrfdroppedW4.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_g_e_reml_dom_health_sub_seqdropped.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann:

#### REML

```{bash g_e_reml_dom_health_sub_seqdropped}
# directories=(
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
# )
# 
#  # Loop through each directory
#  for dir in "${directories[@]}"; do
#      echo "Processing directory: $dir"
#      
#      # Change to the target directory
#      cd "$dir" || { echo "Failed to enter directory: $dir"; continue; }
#  
#      # Run REML only after all ERM outputs are generated
#      echo "Running REML for directory: $dir"
#      ## W1A
#   ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health_demdroppedW1A.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_demdropped_W1A.out > domain_reml_ge_health_demdropped_W1A.log
#   ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health_demsledroppedW1A.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_demsledropped_W1A.out > domain_reml_ge_health_demsledropped_W1A.log
#      ## W1B 
#   ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health_sledroppedW1B.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_sledropped_W1B.out > domain_reml_ge_health_sledropped_W1B.log
     ## W4
#   ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health_ghealthdroppedW4.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_ghealthdropped_W4.out > domain_reml_ge_health_ghealthdropped_W4.log
#   ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health_ghealthschdroppedW4.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_ghealthschdropped_W4.out > domain_reml_ge_health_ghealthschdropped_W4.log
#   ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.domains.health_ghealthschfrfdroppedW4.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out domain_reml_ge_health_ghealthschfrfdropped_W4.out > domain_reml_ge_health_ghealthschfrfdropped_W4.log
#  
# done
```

#### Delta output prep

W1A

```{r do file for demdropped_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_demdropped_W1A.do")                      # create file with this name
cat("domain_reml_ge_health_demdropped_W1A.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(9, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9", "\n")        # line 3: compute prop. of variance due to frfam
cat("R 3 1 2 4 5 6 7 8 9", "\n")        # line 4: compute prop. of variance due to poscog
cat("R 4 1 2 3 5 6 7 8 9", "\n")        # line 5: compute prop. of variance due to school
cat("R 5 1 2 3 4 6 7 8 9", "\n")        # line 6: compute prop. of variance due to sle
cat("R 6 1 2 3 4 5 7 8 9", "\n")        # line 7: compute prop. of variance due to hgeneral
cat("R 7 1 2 3 4 5 6 8 9", "\n")        # line 8: compute prop. of variance due to hmental
cat("R 8 1 2 3 4 5 6 7 9", "\n")        # line 8: compute prop. of variance due to hphysical
cat("R 9 1 2 3 4 5 6 7 8", "\n")        # line 8: compute prop. of variance due to grm
cat("R 1 2 3 4 5 6 7 8 9", "\n")        # line 11: compute prop. of variance due to error
sink()
```

```{r do file for demsledropped_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_demsledropped_W1A.do")                      # create file with this name
cat("domain_reml_ge_health_demsledropped_W1A.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(8, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8", "\n")       # line 3: compute prop. of variance due to frfam
cat("R 3 1 2 4 5 6 7 8", "\n")       # line 4: compute prop. of variance due to poscog
cat("R 4 1 2 3 5 6 7 8", "\n")       # line 5: compute prop. of variance due to school
cat("R 5 1 2 3 4 6 7 8", "\n")       # line 6: compute prop. of variance due to hgeneral
cat("R 6 1 2 3 4 5 7 8", "\n")       # line 7: compute prop. of variance due to hmental
cat("R 7 1 2 3 4 5 6 8", "\n")       # line 8: compute prop. of variance due to hphysical
cat("R 8 1 2 3 4 5 6 7", "\n")       # line 8: compute prop. of variance due to genetic
cat("R 1 2 3 4 5 6 7 8", "\n")        # line 11: compute prop. of variance due to error
sink()
```

W1B

```{r do file for sledropped_W1B}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_sledropped_W1B.do")                      # create file with this name
cat("domain_reml_ge_health_sledropped_W1B.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(9, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9", "\n")        # line 3: compute prop. of variance due to dem
cat("R 3 1 2 4 5 6 7 8 9", "\n")        # line 4: compute prop. of variance due to frfam
cat("R 4 1 2 3 5 6 7 8 9", "\n")        # line 5: compute prop. of variance due to poscog
cat("R 5 1 2 3 4 6 7 8 9", "\n")        # line 6: compute prop. of variance due to school
cat("R 6 1 2 3 4 5 7 8 9", "\n")        # line 7: compute prop. of variance due to hgeneral
cat("R 7 1 2 3 4 5 6 8 9", "\n")        # line 8: compute prop. of variance due to hmental
cat("R 8 1 2 3 4 5 6 7 9", "\n")        # line 8: compute prop. of variance due to hphysical
cat("R 9 1 2 3 4 5 6 7 8", "\n")        # line 8: compute prop. of variance due to grm
cat("R 1 2 3 4 5 6 7 8 9", "\n")        # line 11: compute prop. of variance due to error
sink()
```

W4

```{r do file for ghealthdropped_W4}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_ghealthdropped_W4.do")                      # create file with this name
cat("domain_reml_ge_health_ghealthdropped_W4.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(9, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9", "\n")        # line 3: compute prop. of variance due to dem
cat("R 3 1 2 4 5 6 7 8 9", "\n")        # line 4: compute prop. of variance due to frfam
cat("R 4 1 2 3 5 6 7 8 9", "\n")        # line 5: compute prop. of variance due to poscog
cat("R 5 1 2 3 4 6 7 8 9", "\n")        # line 6: compute prop. of variance due to school
cat("R 6 1 2 3 4 5 7 8 9", "\n")        # line 7: compute prop. of variance due to sle
cat("R 7 1 2 3 4 5 6 8 9", "\n")        # line 8: compute prop. of variance due to hmental
cat("R 8 1 2 3 4 5 6 7 9", "\n")        # line 8: compute prop. of variance due to hphysical
cat("R 9 1 2 3 4 5 6 7 8", "\n")        # line 8: compute prop. of variance due to grm
cat("R 1 2 3 4 5 6 7 8 9", "\n")        # line 11: compute prop. of variance due to error
sink()
```

```{r do file for ghealthschdropped_W4}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_ghealthschdropped_W4.do")                      # create file with this name
cat("domain_reml_ge_health_ghealthschdropped_W4.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(8, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8", "\n")       # line 3: compute prop. of variance due to dem
cat("R 3 1 2 4 5 6 7 8", "\n")       # line 4: compute prop. of variance due to frfam
cat("R 4 1 2 3 5 6 7 8", "\n")       # line 5: compute prop. of variance due to poscog
cat("R 5 1 2 3 4 6 7 8", "\n")       # line 6: compute prop. of variance due to sle
cat("R 6 1 2 3 4 5 7 8", "\n")       # line 7: compute prop. of variance due to hmental
cat("R 7 1 2 3 4 5 6 8", "\n")       # line 8: compute prop. of variance due to hphysical
cat("R 8 1 2 3 4 5 6 7", "\n")       # line 8: compute prop. of variance due to genetic
cat("R 1 2 3 4 5 6 7 8", "\n")        # line 11: compute prop. of variance due to error
sink()
```

```{r do file for ghealthschfrfdropped_W4}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("domain_reml_ge_health_ghealthschfrfdropped_W4.do")                      # create file with this name
cat("domain_reml_ge_health_ghealthschfrfdropped_W4.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(7, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7", "\n")       # line 3: compute prop. of variance due to dem
cat("R 3 1 2 4 5 6 7", "\n")       # line 4: compute prop. of variance due to poscog
cat("R 4 1 2 3 5 6 7", "\n")       # line 5: compute prop. of variance due to sle
cat("R 5 1 2 3 4 6 7", "\n")       # line 6: compute prop. of variance due to hmental
cat("R 6 1 2 3 4 5 7", "\n")       # line 7: compute prop. of variance due to hphysical
cat("R 7 1 2 3 4 5 6", "\n")       # line 8: compute prop. of variance due to grm
cat("R 1 2 3 4 5 6 7", "\n")        # line 11: compute prop. of variance due to error
sink()
```

```{bash output for e_g_cov_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
     echo "Processing: $dir"
     cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
     grep -vwE '(LKH|h2)' domain_reml_ge_health_demdropped_W1A.out > domain_reml_ge_health_demdropped_W1A.out2
     grep -vwE '(LKH|h2)' domain_reml_ge_health_demsledropped_W1A.out > domain_reml_ge_health_demsledropped_W1A.out2
     
     grep -vwE '(LKH|h2)' domain_reml_ge_health_sledropped_W1B.out > domain_reml_ge_health_sledropped_W1B.out2
     
     grep -vwE '(LKH|h2)' domain_reml_ge_health_ghealthdropped_W4.out > domain_reml_ge_health_ghealthdropped_W4.out2
     grep -vwE '(LKH|h2)' domain_reml_ge_health_ghealthschdropped_W4.out > domain_reml_ge_health_ghealthschdropped_W4.out2
     grep -vwE '(LKH|h2)' domain_reml_ge_health_ghealthschfrfdropped_W4.out > domain_reml_ge_health_ghealthschfrfdropped_W4.out2

     ./../../../mtg2 -delta2 domain_reml_ge_health_demdropped_W1A.do > domain_reml_ge_health_demdropped_W1A.out3
     ./../../../mtg2 -delta2 domain_reml_ge_health_demsledropped_W1A.do > domain_reml_ge_health_demsledropped_W1A.out3
     
     ./../../../mtg2 -delta2 domain_reml_ge_health_sledropped_W1B.do > domain_reml_ge_health_sledropped_W1B.out3
     
     ./../../../mtg2 -delta2 domain_reml_ge_health_ghealthdropped_W4.do > domain_reml_ge_health_ghealthdropped_W4.out3
     ./../../../mtg2 -delta2 domain_reml_ge_health_ghealthschdropped_W4.do > domain_reml_ge_health_ghealthschdropped_W4.out3
     ./../../../mtg2 -delta2 domain_reml_ge_health_ghealthschfrfdropped_W4.do > domain_reml_ge_health_ghealthschfrfdropped_W4.out3
done
```

#### Output table 

W1A

```{r output for demdropped_W1A and demsledropped_W1A}
domain_reml_ge_health_demdropped_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "domain_reml_ge_health_demdropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", "Error")
)

domain_reml_ge_health_demsledropped_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "domain_reml_ge_health_demsledropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", "Genetic", "Error")
)
```

W1B

```{r output for sledropped_W1B}
domain_reml_ge_health_sledropped_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "domain_reml_ge_health_sledropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", "Genetic", "Error")
)
```

W4

```{r output for W4 ghealthdropped  }
domain_reml_ge_health_ghealthdropped_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "domain_reml_ge_health_ghealthdropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "Mental health", "Physical health", "Genetic", "Error")
)

domain_reml_ge_health_ghealthschdropped_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "domain_reml_ge_health_ghealthschdropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "Stressful life events", "Mental health", "Physical health", "Genetic", "Error")
)

domain_reml_ge_health_ghealthschfrfdropped_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "domain_reml_ge_health_ghealthschfrfdropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Positive cognition", "School", "Stressful life events", "Mental health", "Physical health", "Genetic", "Error")
)
```

```{r output for dropped domains}
domain_reml_ge_health_seqdropped.output <- rbind(
  domain_reml_ge_health_demdropped_W1A.output, 
  domain_reml_ge_health_demsledropped_W1A.output, 
  domain_reml_ge_health_sledropped_W1B.output,
  domain_reml_ge_health_ghealthdropped_W4.output,
  domain_reml_ge_health_ghealthschdropped_W4.output,
  domain_reml_ge_health_ghealthschfrfdropped_W4.output
  )

domain_reml_ge_health_seqdropped.output
```

#### LRT

W1A

Number of parameters dropped: 
domain_reml_ge_health_demdropped_W1A.output = 1
domain_reml_ge_health_demsledropped_W1A.output = 2
domain_reml_ge_health_dropped_W1A.out = 3

But we will use df = 1 as we will sequenctially drop these from the model to assess the decrease in fit from each parameter. 

```{r LRT}
LRT_domdrop_dem_W1A <- LRT(path_full = "domains/domain_reml_ge_health_W1A.out", 
                    path_constrained = "domains/domain_reml_ge_health_demdropped_W1A.out",
                    Outcome = "W1A",
                    Comparison = "Domains dropped dem compared to full Domains",
                    df = 1)

LRT_domdrop_demsle_W1A <- LRT(path_full = "domains/domain_reml_ge_health_demdropped_W1A.out", 
                    path_constrained = "domains/domain_reml_ge_health_demsledropped_W1A.out",
                    Outcome = "W1A",
                    Comparison = "Domains dropped demsle compared to domains dropped dem",
                    df = 1)

LRT_domdrop_demslegen_W1A <- LRT(path_full = "domains/domain_reml_ge_health_demsledropped_W1A.out", 
                    path_constrained = "domains/domain_reml_ge_health_dropped_W1A.out",
                    Outcome = "W1A",
                    Comparison = "Domains dropped compared to domains dropped demsle",
                    df = 1)

LRT_dom_seqdrop_W1A <- rbind(LRT_domdrop_dem_W1A, LRT_domdrop_demsle_W1A, LRT_domdrop_demslegen_W1A)
LRT_dom_seqdrop_W1A
```

W1B

Number of parameters dropped: 
domain_reml_ge_health_sledropped_W1B.output = 1
domain_reml_ge_health_dropped_W1B.out = 2

But we will use df = 1 as we will sequenctially drop these from the model to assess the decrease in fit from each parameter. 

```{r LRT}
LRT_domdrop_sle_W1B <- LRT(path_full = "domains/domain_reml_ge_health_W1B.out", 
                    path_constrained = "domains/domain_reml_ge_health_sledropped_W1B.out",
                    Outcome = "W1B",
                    Comparison = "Domains dropped sle compared to full Domains",
                    df = 1)

LRT_domdrop_W1B <- LRT(path_full = "domains/domain_reml_ge_health_sledropped_W1B.out", 
                    path_constrained = "domains/domain_reml_ge_health_dropped_W1B.out",
                    Outcome = "W1B",
                    Comparison = "Domains dropped compared to domains dropped sle",
                    df = 1)

LRT_dom_seqdrop_W1B <- rbind(LRT_domdrop_sle_W1B, LRT_domdrop_W1B)
LRT_dom_seqdrop_W1B
```

W4

Number of parameters dropped: 
domain_reml_ge_health_ghealthdropped_W4.output = 1
domain_reml_ge_health_ghealthschdropped_W4.output = 2
domain_reml_ge_health_ghealthschfrfdropped_W4.output = 3
domain_reml_ge_health_dropped_W4.out = 4

But we will use df = 1 as we will sequenctially drop these from the model to assess the decrease in fit from each parameter. 

```{r LRT}
LRT_domdrop_ghealth_W4 <- LRT(path_full = "domains/domain_reml_ge_health_W4.out", 
                    path_constrained = "domains/domain_reml_ge_health_ghealthdropped_W4.out",
                    Outcome = "W4",
                    Comparison = "Domains dropped ghealth compared to full Domains",
                    df = 1)

LRT_domdrop_ghealthsch_W4 <- LRT(path_full = "domains/domain_reml_ge_health_ghealthdropped_W4.out", 
                    path_constrained = "domains/domain_reml_ge_health_ghealthschdropped_W4.out",
                    Outcome = "W4",
                    Comparison = "Domains dropped ghealthsch compared to domains dropped ghealth",
                    df = 1)

LRT_domdrop_ghealthschfrf_W4 <- LRT(path_full = "domains/domain_reml_ge_health_ghealthschdropped_W4.out", 
                    path_constrained = "domains/domain_reml_ge_health_ghealthschfrfdropped_W4.out",
                    Outcome = "W4",
                    Comparison = "Domains dropped ghealthschfrf compared to domains dropped ghealthsch",
                    df = 1)

LRT_domdrop_W4 <- LRT(path_full = "domains/domain_reml_ge_health_ghealthschfrfdropped_W4.out", 
                    path_constrained = "domains/domain_reml_ge_health_dropped_W4.out",
                    Outcome = "W4",
                    Comparison = "Domains dropped compared to domains dropped ghealthschfrf",
                    df = 1)

LRT_dom_seqdrop_W4 <- rbind(LRT_domdrop_ghealth_W4, LRT_domdrop_ghealthsch_W4, LRT_domdrop_ghealthschfrf_W4, LRT_domdrop_W4)
LRT_dom_seqdrop_W4
```

## 3. Covariance models

### REML

To estimate the rGE - we need to specify in the model that we want to calculate the correlation between the two random effects. 

To do this - we first estimate Q
[transposed(sqrt(GRM)*sqrt(ERM)) + transposed(sqrt(GRM)*sqrt(ERMtransposed)]

See section 2.10 of the MTG2 manual for more details on NPD matrix bending.

We can do this in mtg2 through the following steps: 
* i. ensure positive definite kernel matrices - prerequisite for cholesky decomposition
* ii. Cholesky decomposition
* iii. compute Q

Step i and ii are done with the following script: 

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_rge_prep_dom_sub.txt

i. ensure positive definite kernel matrices

```{bash rge positive definite}
# cd /project/rche/data/datasets/addhealth/scratch/kthompson/mtg2_output/complete_dat_nosibs/domains/
# # GRM
# ./../../mtg2 -p IDs_IID.fam -g grm_full_out.grm_reformat -thread 80 -bend 2 # gives grm_full_out.grm_reformat.bend
# # ERM - dem 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_dem -thread 80 -bend 2 # erm_dat_theory1_clean_dummy_dem.bend
# # ERM - frfam 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_frfam -thread 80 -bend 2 # erm_dat_theory1_clean_dummy_frfam.bend
# # ERM - health_general 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_health_general -thread 80 -bend 2 # erm_dat_theory1_clean_dummy_health_general.bend
# # ERM - health_mental 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_health_mental -thread 80 -bend 2 # erm_dat_theory1_clean_dummy_health_mental.bend
# # ERM - health_physical 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_health_physical -thread 80 -bend 2 # erm_dat_theory1_clean_dummy_health_physical.bend
# # ERM - poscog 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_poscog -thread 80 -bend 2 # erm_dat_theory1_clean_dummy_poscog.bend
# # ERM - school 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_school -thread 80 -bend 2 # erm_dat_theory1_clean_dummy_school.bend
# # ERM - sle 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_sle -thread 80 -bend 2 # erm_dat_theory1_clean_dummy_sle.bend
```

ii. Cholesky decomposition

```{bash rge cholesky}
# cd /project/rche/data/datasets/addhealth/scratch/kthompson/mtg2_output/complete_dat_nosibs/domains/
# # GRM
# ./../../mtg2 -p IDs_IID.fam -g grm_full_out.grm_reformat.bend -thread 80 -chol 1 # output grm_full_out.grm_reformat.bend.chol
# # ERM
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_dem.bend -thread 80 -chol 1 # output erm_dat_theory1_clean_dummy_dem.bend.chol
# # ERM - frfam 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_frfam.bend -thread 80 -chol 1 # erm_dat_theory1_clean_dummy_frfam.bend.chol
# # ERM - health_general 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_health_general.bend -thread 80 -chol 1 # erm_dat_theory1_clean_dummy_health_general.bend.chol
# # ERM - health_mental 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_health_mental.bend -thread 80 -chol 1 # erm_dat_theory1_clean_dummy_health_mental.bend.chol
# # ERM - health_physical 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_health_physical.bend -thread 80 -chol 1 # erm_dat_theory1_clean_dummy_health_physical.bend.chol
# # ERM - poscog 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_poscog.bend -thread 80 -chol 1 # erm_dat_theory1_clean_dummy_poscog.bend.chol
# # ERM - school 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_school.bend -thread 80 -chol 1 # erm_dat_theory1_clean_dummy_school.bend.chol
# # ERM - sle 
# ./../../mtg2 -p IDs_IID.fam -g erm_dat_theory1_clean_dummy_sle.bend -thread 80 -chol 1 # erm_dat_theory1_clean_dummy_sle.bend.chol
```

iii. compute Q 

Create txt file with choleksy output. 

Also created below is the txt files that contain all the correlation matrices for the different rGE models. 

```{r loop to create pair txt files of chol files}
# Define the list of file names
domain_chol_file_names <- c(
  "erm_dat_theory1_clean_dummy_dem.bend.chol",
  "erm_dat_theory1_clean_dummy_frfam.bend.chol",
  "erm_dat_theory1_clean_dummy_health_general.bend.chol",
  "erm_dat_theory1_clean_dummy_health_mental.bend.chol",
  "erm_dat_theory1_clean_dummy_health_physical.bend.chol",
  "erm_dat_theory1_clean_dummy_poscog.bend.chol",
  "erm_dat_theory1_clean_dummy_school.bend.chol",
  "erm_dat_theory1_clean_dummy_sle.bend.chol"
)

# Define the output directories
output_cov_dir <- c(
  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/",
  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/",
  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/",
  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/",
  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

# Extract the meaningful names from file names
domain_names <- gsub(".*_dummy_|.*_reformat\\.|\\.bend\\.chol", "", domain_chol_file_names)

# Generate all pairwise combinations
combinations <- combn(seq_along(domain_chol_file_names), 2, simplify = FALSE)

# Loop through all directories
for (dir in output_cov_dir){

# Loop through each combination and write to files
for (pair_indices in combinations) {
  # Get the file paths for the pair
  pair_files <- domain_chol_file_names[pair_indices]

  # Get the extracted domain names for the pair
  pair_names <- domain_names[pair_indices]

  # Create a meaningful output file name
  output_file <- paste0(dir, paste(pair_names, collapse = "_"), ".chol")

  # Write the pair vector to the file
  write.table(pair_files, file = output_file, quote = FALSE, row.names = FALSE, col.names = FALSE, sep = "\t")
}

# Loop through each domain file name (excluding the first, which is "grm_full_out")
for (file_name in domain_chol_file_names) {
  # Create a pair vector with the first file and the current file
  pair_vector <- c("grm_full_out.grm_reformat.bend.chol", file_name)

  # Extract the domain-specific name for the output file
  domain_name <- gsub(".*_dummy_|.*_reformat\\.|\\.bend\\.chol", "", file_name)

  # Create the output file name
  output_file <- paste0(dir, domain_name, ".grm.chol")

  # Write the pair vector to the file
  write.table(pair_vector, file = output_file, quote = FALSE, row.names = FALSE, col.names = FALSE, sep = "\t")
}
}
```

### Create cov txt files 

```{r create cov txt files}
for (dir in output_cov_dir){
  
  # ALL covariances (nothing dropped)
domains_cov_all.txt <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat",
                 "dem_frfam.chol.matmat2", "dem_poscog.chol.matmat2", "dem_school.chol.matmat2", "dem_sle.chol.matmat2", "dem_health_general.chol.matmat2", "dem_health_mental.chol.matmat2", "dem_health_physical.chol.matmat2",           
                 "frfam_poscog.chol.matmat2", "frfam_school.chol.matmat2", "frfam_sle.chol.matmat2", "frfam_health_general.chol.matmat2",  "frfam_health_mental.chol.matmat2", "frfam_health_physical.chol.matmat2",  
                 "poscog_school.chol.matmat2", "poscog_sle.chol.matmat2", "health_general_poscog.chol.matmat2", "health_mental_poscog.chol.matmat2", "health_physical_poscog.chol.matmat2", 
                 "school_sle.chol.matmat2", "health_general_school.chol.matmat2", "health_mental_school.chol.matmat2", "health_physical_school.chol.matmat2",
                 "health_general_sle.chol.matmat2", "health_mental_sle.chol.matmat2", "health_physical_sle.chol.matmat2",
                 "health_general_health_mental.chol.matmat2", "health_general_health_physical.chol.matmat2",   
                  "health_mental_health_physical.chol.matmat2",  
                 "dem.grm.chol.matmat2", "frfam.grm.chol.matmat2", "poscog.grm.chol.matmat2", "school.grm.chol.matmat2", "sle.grm.chol.matmat2", "health_general.grm.chol.matmat2", "health_mental.grm.chol.matmat2", "health_physical.grm.chol.matmat2")

write.table(domains_cov_all.txt, paste0(dir, "domains_cov_all.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

### Create txt files for the "dropped" models - different for W1AB and W4

# W1A
rge_all_erm.grm.domains.health_droppedW1A.txt <- c("erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical",
                                                   "health_mental_health_physical.chol.matmat2", "health_general_health_mental.chol.matmat2", "health_general_health_physical.chol.matmat2", "frfam_poscog.chol.matmat2", "frfam_school.chol.matmat2", "frfam_health_mental.chol.matmat2", "frfam_health_physical.chol.matmat2", "frfam_health_general.chol.matmat2", "poscog_school.chol.matmat2", "health_mental_poscog.chol.matmat2", "health_physical_poscog.chol.matmat2", "health_general_poscog.chol.matmat2", "health_mental_school.chol.matmat2", "health_physical_school.chol.matmat2", "health_general_school.chol.matmat2")

write.table(rge_all_erm.grm.domains.health_droppedW1A.txt, paste0(dir, "rge_all_erm.grm.domains.health_droppedW1A.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# W1B
rge_all_erm.grm.domains.health_droppedW1B.txt <- c("erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat",
                                                   "health_mental_health_physical.chol.matmat2", "health_general_health_mental.chol.matmat2", "health_general_health_physical.chol.matmat2", "frfam_poscog.chol.matmat2", "frfam_school.chol.matmat2", "frfam_health_mental.chol.matmat2", "frfam_health_physical.chol.matmat2", "frfam_health_general.chol.matmat2", "poscog_school.chol.matmat2", "health_mental_poscog.chol.matmat2", "health_physical_poscog.chol.matmat2", "health_general_poscog.chol.matmat2", "health_mental_school.chol.matmat2", "health_physical_school.chol.matmat2", "health_general_school.chol.matmat2", "health_mental.grm.chol.matmat2", "health_physical.grm.chol.matmat2", "health_general.grm.chol.matmat2", "frfam.grm.chol.matmat2", "poscog.grm.chol.matmat2", "school.grm.chol.matmat2")

write.table(rge_all_erm.grm.domains.health_droppedW1B.txt, paste0(dir, "rge_all_erm.grm.domains.health_droppedW1B.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# W4
rge_all_erm.grm.domains.health_droppedW4.txt <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", 
                                                  "health_mental_health_physical.chol.matmat2",   "dem_poscog.chol.matmat2", "dem_sle.chol.matmat2", "dem_health_mental.chol.matmat2", "dem_health_physical.chol.matmat2", "poscog_sle.chol.matmat2", "health_mental_poscog.chol.matmat2", "health_physical_poscog.chol.matmat2", "health_mental_sle.chol.matmat2", "health_physical_sle.chol.matmat2")

write.table(rge_all_erm.grm.domains.health_droppedW4.txt, paste0(dir, "rge_all_erm.grm.domains.health_droppedW4.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
}
```

Create Q matrix and run REML 

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_rge_reml_dom_sub.txt

This script contains the following code for all the rGE models - see script for this - *I have given the below as an example*. 

```{bash}
# cd /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/
# for chol_file in *.chol; do
#     [[ "$chol_file" == *.bend.chol ]] && continue
#    ./../../../mtg2 -p IDs_IID.fam -mg "$chol_file" -thread 80 -matmat 2
# done
```

```{bash compute ereml and greml model}
#      # REML for dropped models
#      ./../../../mtg2 -p IDs_IID.fam -mg rge_all_erm.grm.domains.health_droppedW1A.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_health_all_reml_dropped_W1A.out > e_g_cov_health_all_reml_dropped_W1A.log
#      ./../../../mtg2 -p IDs_IID.fam -mg rge_all_erm.grm.domains.health_droppedW1B.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_health_all_reml_dropped_W1B.out > e_g_cov_health_all_reml_dropped_W1B.log
#      ./../../../mtg2 -p IDs_IID.fam -mg rge_all_erm.grm.domains.health_droppedW4.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_health_all_reml_dropped_W4.out > e_g_cov_health_all_reml_dropped_W4.log
#      
#      # remove not needed columns for output
#      grep -vwE '(LKH|h2)' e_g_cov_health_all_reml_dropped_W1A.out > e_g_cov_health_all_reml_dropped_W1A.out2
#      grep -vwE '(LKH|h2)' e_g_cov_health_all_reml_dropped_W1B.out > e_g_cov_health_all_reml_dropped_W1B.out2
#      grep -vwE '(LKH|h2)' e_g_cov_health_all_reml_dropped_W4.out > e_g_cov_health_all_reml_dropped_W4.out2
```

### A. All cor included - SUPPLEMENT

#### Delta output prep

```{r do file for e_g_cov_all_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_cov_all_reml_W1A.do")                      # create file with this name
cat("e_g_cov_all_reml_W1A.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(46, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # frfam
cat("R 4 1 2 3 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # school
cat("R 6 1 2 3 4 5 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # hgeneral
cat("R 8 1 2 3 4 5 6 7 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # hmental
cat("R 9 1 2 3 4 5 6 7 8 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # hphysical
cat("R 10 1 2 3 4 5 6 7 8 9 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # genetics
cat("R 1 2 3 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # error
cat("C 11 2 3", "\n")                                      # correlation between dem~frf
cat("C 12 2 4", "\n")                                      # correlation between dem~pcg
cat("C 13 2 5", "\n")                                      # correlation between dem~sch
cat("C 14 2 6", "\n")                                      # correlation between dem~sle
cat("C 15 2 7", "\n")                                      # correlation between dem~g
cat("C 16 2 8", "\n")                                      # correlation between dem~m
cat("C 17 2 9", "\n")                                      # correlation between dem~p
cat("C 18 3 4", "\n")                                      # correlation between frf~pcg
cat("C 19 3 5", "\n")                                      # correlation between frf~sch
cat("C 20 3 6", "\n")                                      # correlation between frf~sle
cat("C 21 3 7", "\n")                                      # correlation between frf~g
cat("C 22 3 8", "\n")                                      # correlation between frf~m
cat("C 23 3 9", "\n")                                      # correlation between frf~p
cat("C 24 4 5", "\n")                                      # correlation between pcg~sch
cat("C 25 4 6", "\n")                                      # correlation between pcg~sle
cat("C 26 4 7", "\n")                                      # correlation between pcg~g
cat("C 27 4 8", "\n")                                      # correlation between pcg~m
cat("C 28 4 9", "\n")                                      # correlation between pcg~p
cat("C 29 5 6", "\n")                                      # correlation between sch~sle
cat("C 30 5 7", "\n")                                      # correlation between sch~g
cat("C 31 5 8", "\n")                                      # correlation between sch~m
cat("C 32 5 9", "\n")                                      # correlation between sch~p
cat("C 33 6 7", "\n")                                      # correlation between sle~g
cat("C 34 6 8", "\n")                                      # correlation between sle~m
cat("C 35 6 9", "\n")                                      # correlation between sle~p
cat("C 36 7 8", "\n")                                      # correlation between hmental hgeneral
cat("C 37 7 9", "\n")                                      # correlation between hphys hgeneral
cat("C 38 8 9", "\n")                                      # correlation between hphys hmental
cat("C 39 10 2", "\n")                                      # correlation between grm~dem
cat("C 40 10 3", "\n")                                      # correlation between grm~frf
cat("C 41 10 4", "\n")                                      # correlation between grm~pcg
cat("C 42 10 5", "\n")                                      # correlation between grm~sch
cat("C 43 10 6", "\n")                                      # correlation between grm~sle
cat("C 44 10 7", "\n")                                      # correlation between grm~hgeneral
cat("C 45 10 8", "\n")                                      # correlation between grm~hmental
cat("C 46 10 9", "\n")                                      # correlation between grm~hphysical
sink()
```

```{r do file for e_g_cov_all_reml_W1B}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_cov_all_reml_W1B.do")                      # create file with this name
cat("e_g_cov_all_reml_W1B.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(46, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # frfam
cat("R 4 1 2 3 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # school
cat("R 6 1 2 3 4 5 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # hgeneral
cat("R 8 1 2 3 4 5 6 7 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # hmental
cat("R 9 1 2 3 4 5 6 7 8 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # hphysical
cat("R 10 1 2 3 4 5 6 7 8 9 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # genetics
cat("R 1 2 3 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # error
cat("C 11 2 3", "\n")                                      # correlation between dem~frf
cat("C 12 2 4", "\n")                                      # correlation between dem~pcg
cat("C 13 2 5", "\n")                                      # correlation between dem~sch
cat("C 14 2 6", "\n")                                      # correlation between dem~sle
cat("C 15 2 7", "\n")                                      # correlation between dem~g
cat("C 16 2 8", "\n")                                      # correlation between dem~m
cat("C 17 2 9", "\n")                                      # correlation between dem~p
cat("C 18 3 4", "\n")                                      # correlation between frf~pcg
cat("C 19 3 5", "\n")                                      # correlation between frf~sch
cat("C 20 3 6", "\n")                                      # correlation between frf~sle
cat("C 21 3 7", "\n")                                      # correlation between frf~g
cat("C 22 3 8", "\n")                                      # correlation between frf~m
cat("C 23 3 9", "\n")                                      # correlation between frf~p
cat("C 24 4 5", "\n")                                      # correlation between pcg~sch
cat("C 25 4 6", "\n")                                      # correlation between pcg~sle
cat("C 26 4 7", "\n")                                      # correlation between pcg~g
cat("C 27 4 8", "\n")                                      # correlation between pcg~m
cat("C 28 4 9", "\n")                                      # correlation between pcg~p
cat("C 29 5 6", "\n")                                      # correlation between sch~sle
cat("C 30 5 7", "\n")                                      # correlation between sch~g
cat("C 31 5 8", "\n")                                      # correlation between sch~m
cat("C 32 5 9", "\n")                                      # correlation between sch~p
cat("C 33 6 7", "\n")                                      # correlation between sle~g
cat("C 34 6 8", "\n")                                      # correlation between sle~m
cat("C 35 6 9", "\n")                                      # correlation between sle~p
cat("C 36 7 8", "\n")                                      # correlation between hmental hgeneral
cat("C 37 7 9", "\n")                                      # correlation between hphys hgeneral
cat("C 38 8 9", "\n")                                      # correlation between hphys hmental
cat("C 39 10 2", "\n")                                      # correlation between grm~dem
cat("C 40 10 3", "\n")                                      # correlation between grm~frf
cat("C 41 10 4", "\n")                                      # correlation between grm~pcg
cat("C 42 10 5", "\n")                                      # correlation between grm~sch
cat("C 43 10 6", "\n")                                      # correlation between grm~sle
cat("C 44 10 7", "\n")                                      # correlation between grm~hgeneral
cat("C 45 10 8", "\n")                                      # correlation between grm~hmental
cat("C 46 10 9", "\n")                                      # correlation between grm~hphysical
sink()
```

```{r do file for e_g_cov_all_reml_W4}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_cov_all_reml_W4.do")                      # create file with this name
cat("e_g_cov_all_reml_W4.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(46, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # frfam
cat("R 4 1 2 3 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # school
cat("R 6 1 2 3 4 5 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # hgeneral
cat("R 8 1 2 3 4 5 6 7 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # hmental
cat("R 9 1 2 3 4 5 6 7 8 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # hphysical
cat("R 10 1 2 3 4 5 6 7 8 9 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # genetics
cat("R 1 2 3 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29 30 30 31 31 32 32 33 33 34 34 35 35 36 36 37 37 38 38 39 39 40 40 41 41 42 42 43 43 44 44 45 45 46 46", "\n")   # error
cat("C 11 2 3", "\n")                                      # correlation between dem~frf
cat("C 12 2 4", "\n")                                      # correlation between dem~pcg
cat("C 13 2 5", "\n")                                      # correlation between dem~sch
cat("C 14 2 6", "\n")                                      # correlation between dem~sle
cat("C 15 2 7", "\n")                                      # correlation between dem~g
cat("C 16 2 8", "\n")                                      # correlation between dem~m
cat("C 17 2 9", "\n")                                      # correlation between dem~p
cat("C 18 3 4", "\n")                                      # correlation between frf~pcg
cat("C 19 3 5", "\n")                                      # correlation between frf~sch
cat("C 20 3 6", "\n")                                      # correlation between frf~sle
cat("C 21 3 7", "\n")                                      # correlation between frf~g
cat("C 22 3 8", "\n")                                      # correlation between frf~m
cat("C 23 3 9", "\n")                                      # correlation between frf~p
cat("C 24 4 5", "\n")                                      # correlation between pcg~sch
cat("C 25 4 6", "\n")                                      # correlation between pcg~sle
cat("C 26 4 7", "\n")                                      # correlation between pcg~g
cat("C 27 4 8", "\n")                                      # correlation between pcg~m
cat("C 28 4 9", "\n")                                      # correlation between pcg~p
cat("C 29 5 6", "\n")                                      # correlation between sch~sle
cat("C 30 5 7", "\n")                                      # correlation between sch~g
cat("C 31 5 8", "\n")                                      # correlation between sch~m
cat("C 32 5 9", "\n")                                      # correlation between sch~p
cat("C 33 6 7", "\n")                                      # correlation between sle~g
cat("C 34 6 8", "\n")                                      # correlation between sle~m
cat("C 35 6 9", "\n")                                      # correlation between sle~p
cat("C 36 7 8", "\n")                                      # correlation between hmental hgeneral
cat("C 37 7 9", "\n")                                      # correlation between hphys hgeneral
cat("C 38 8 9", "\n")                                      # correlation between hphys hmental
cat("C 39 10 2", "\n")                                      # correlation between grm~dem
cat("C 40 10 3", "\n")                                      # correlation between grm~frf
cat("C 41 10 4", "\n")                                      # correlation between grm~pcg
cat("C 42 10 5", "\n")                                      # correlation between grm~sch
cat("C 43 10 6", "\n")                                      # correlation between grm~sle
cat("C 44 10 7", "\n")                                      # correlation between grm~hgeneral
cat("C 45 10 8", "\n")                                      # correlation between grm~hmental
cat("C 46 10 9", "\n")                                      # correlation between grm~hphysical
sink()
```

Output using Delta

```{bash output for e_g_cov_all_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
     echo "Processing: $dir"
     cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
     ./../../../mtg2 -delta2 e_g_cov_all_reml_W1A.do > e_g_cov_all_reml_W1A.out3
     ./../../../mtg2 -delta2 e_g_cov_all_reml_W1B.do > e_g_cov_all_reml_W1B.out3
     ./../../../mtg2 -delta2 e_g_cov_all_reml_W4.do > e_g_cov_all_reml_W4.out3
done
```

#### Output table

```{r output for dropped cov model}
e_g_cov_all_reml_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "e_g_cov_all_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", "Error")
)

e_g_cov_all_reml_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "e_g_cov_all_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", "Error")
)

e_g_cov_all_reml_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "e_g_cov_all_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", "Error")
)

e_g_cov_all_reml.output <- rbind(e_g_cov_all_reml_W1A.output, e_g_cov_all_reml_W1B.output, e_g_cov_all_reml_W4.output)

e_g_cov_all_reml.output %>% select(outcome, domain, parameter, variance_std, flag_corr_out_of_bounds)
```

Test to calculate covariance contributions

```{r compare dom and cov main effects}
domain_reml_ge_health.output
e_g_cov_all_reml.output
```

Significant and not out of bounds correlations: 
W1A	dem ~ poscog	
W1A	dem ~ mhealth	
W1A	dem ~ phealth	
W1A	frfam ~ poscog 	
W1A	frfam ~ school 	
W1A	frfam ~ mhealth 	
W1A	frfam ~ phealth 	
W1A	poscog ~ school 	
W1A	ghealth ~ poscog 
W1A	mhealth ~ poscog 
W1A	phealth ~ poscog 
W1A	ghealth ~ school 
W1A	mhealth ~ school 
W1A	phealth ~ school 
W1A	phealth ~ ghealth	
W1A	mhealth ~ phealth	
W1B	dem ~ poscog	
W1B	dem ~ school	
W1B	dem ~ mhealth	
W1B	frfam ~ poscog 	
W1B	frfam ~ school 	
W1B	frfam ~ mhealth 	
W1B	frfam ~ phealth 	
W1B	poscog ~ school 	
W1B	ghealth ~ poscog 
W1B	mhealth ~ poscog 
W1B	phealth ~ poscog 
W1B	mhealth ~ school 
W1B	phealth ~ school 
W1B	phealth ~ ghealth	
W1B	mhealth ~ phealth

#### Plot correlations 

```{r include correlations in output}
color_blind_palette <- c("#56B4E9", "#E69F00")


e_g_cov_all_reml.cor <- e_g_cov_all_reml.output %>% 
  mutate(
    significance = ifelse(P_value < 0.05, "Significant", "Not significant"),
    shape_group = case_when(flag_corr_out_of_bounds ~ "Out of bounds",
                            TRUE ~ as.character(significance))) %>%
  filter(domain == "Correlation",
         flag_corr_out_of_bounds == "FALSE")

e_g_cov_all_reml.cor %>% filter(shape_group == "Significant")
  
all.cor <- ggplot(e_g_cov_all_reml.cor, 
                      aes(x = variance_std, 
                          y = reorder(covariance, variance_std), 
                          color = outcome, 
                          shape = shape_group)) +   # <- use the new variable
  geom_point(size = 3, position = position_dodge(width = 1)) +
  geom_errorbarh(aes(xmin = CI_lower, xmax = CI_upper), height = 0.2, position = position_dodge(width = 1)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "gray50") +
  scale_shape_manual(values = c(
    "Out of bounds" = 1,      # hollow circle
    "Significant" = 16,            # solid circle
    "Not significant" = 17            # solid triangle, adjust to taste
  )) +
  scale_color_manual(values = color_blind_palette) +
  labs(
    x = "Correlation Estimate",
    y = "",
    color = "Outcome",
    shape = "Flagged / Significant"
  ) +
  theme_classic(base_size = 14) +
  theme(
    legend.position = "bottom",
    legend.title = element_blank(),
    legend.text = element_text(size = 10),
    legend.key.width = unit(1.5, "cm"),
    axis.title.y = element_text(size = 16, margin = margin(r = 10)),
    axis.text.x = element_text(angle = 30, hjust = 1),
    plot.title = element_text(size = 16, face = "bold"),
    plot.subtitle = element_text(size = 14),
    plot.margin = margin(10, 10, 10, 10)
  )

# ggsave(all.cor, file = "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/plot_all_corGE.png", height = 6, width = 8)
```

### B. Dropped/reduced rge  - SUPPLEMENT

#### Delta output prep 

* ‘R’ is to get the ratio of variance
* R 2 1 3 4 (i.e. 2 / (1 + 2 + 3 + 4))

var(y) = var(e1) + var(e2) + 2cov(e1e2) + var(error)

* ‘C’ is to get the correlation of covariance
* C 5 2 3 (i.e. 5 / sqrt(2 * 3))
* Here we have the covariance estimate first, followed by the two variances used for the covariance

```{r do file for e_g_cov_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_cov_health_all_reml_dropped_W1A.do")                      # create file with this name
cat("e_g_cov_health_all_reml_dropped_W1A.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(22, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22", "\n")   # frfam
cat("R 3 1 2 4 5 6 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22", "\n")   # poscog
cat("R 4 1 2 3 5 6 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22", "\n")   # school
cat("R 5 1 2 3 4 6 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22", "\n")   # hgeneral
cat("R 6 1 2 3 4 5 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22", "\n")   # hmental
cat("R 7 1 2 3 4 5 6 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22", "\n")   # hphysical
cat("R 1 2 3 4 5 6 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22", "\n")   # error
cat("C 8 6 7", "\n")                                      # correlation between hphys hmental
cat("C 9 5 6", "\n")                                      # correlation between hmental hgeneral
cat("C 10 7 5", "\n")                                      # correlation between hphys hgeneral
cat("C 11 2 3", "\n")                                      # correlation between frf~pcg          
cat("C 12 2 4", "\n")                                      # correlation between frf~sch        
cat("C 13 2 6", "\n")                                      # correlation between frf~m            
cat("C 14 2 7", "\n")                                      # correlation between frf~p              
cat("C 15 2 5", "\n")                                      # correlation between frf~g              
cat("C 16 3 4", "\n")                                      # correlation between pcg~sch             
cat("C 17 3 6", "\n")                                      # correlation between pcg~m               
cat("C 18 3 7", "\n")                                      # correlation between pcg~p               
cat("C 19 3 5", "\n")                                      # correlation between pcg~g               
cat("C 20 4 6", "\n")                                      # correlation between sch~m               
cat("C 21 4 7", "\n")                                      # correlation between sch~p               
cat("C 22 4 5", "\n")                                      # correlation between sch~g               
sink()
```

```{r do file for e_g_cov_reml_W1B}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_cov_health_all_reml_dropped_W1B.do")                      # create file with this name
cat("e_g_cov_health_all_reml_dropped_W1B.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(29, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29", "\n")   # frfam
cat("R 3 1 2 4 5 6 7 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29", "\n")   # poscog
cat("R 4 1 2 3 5 6 7 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29", "\n")   # school
cat("R 5 1 2 3 4 6 7 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29", "\n")   # hgeneral
cat("R 6 1 2 3 4 5 7 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29", "\n")   # hmental
cat("R 7 1 2 3 4 5 6 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29", "\n")   # hphysical
cat("R 8 1 2 3 4 5 6 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29", "\n")   # genetic
cat("R 1 2 3 4 5 6 7 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26 27 27 28 28 29 29", "\n")   # error
cat("C 9 6 7", "\n")                                       # correlation between hphys hmental      
cat("C 10 5 6", "\n")                                      # correlation between hmental hgeneral
cat("C 11 7 5", "\n")                                      # correlation between hphys hgeneral        
cat("C 12 2 3", "\n")                                      # correlation between frf~pcg                                      
cat("C 13 2 4", "\n")                                      # correlation between frf~sch                     
cat("C 14 2 6", "\n")                                      # correlation between frf~m                      
cat("C 15 2 7", "\n")                                      # correlation between frf~p                               
cat("C 16 2 5", "\n")                                      # correlation between frf~g                            
cat("C 17 3 4", "\n")                                      # correlation between pcg~sch                     
cat("C 18 3 6", "\n")                                      # correlation between pcg~m                            
cat("C 19 3 7", "\n")                                      # correlation between pcg~p                            
cat("C 20 3 5", "\n")                                      # correlation between pcg~g                            
cat("C 21 4 6", "\n")                                      # correlation between sch~m                            
cat("C 22 4 7", "\n")                                      # correlation between sch~p                            
cat("C 23 4 5", "\n")                                      # correlation between sch~g                                               
cat("C 24 8 6", "\n")                                      # correlation between grm~m                            
cat("C 25 8 7", "\n")                                      # correlation between grm~p                            
cat("C 26 8 5", "\n")                                      # correlation between grm~g                            
cat("C 27 8 2", "\n")                                      # correlation between grm~frf                            
cat("C 28 8 3", "\n")                                      # correlation between grm~poscog                               
cat("C 29 8 4", "\n")                                      # correlation between grm~school                              
sink()
```

```{r do file for e_g_cov_reml_W4}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_cov_health_all_reml_dropped_W4.do")                      # create file with this name
cat("e_g_cov_health_all_reml_dropped_W4.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(16, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15", "\n")   # dem
cat("R 3 1 2 4 5 6 7 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15", "\n")   # poscog
cat("R 4 1 2 3 5 6 7 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15", "\n")   # sle
cat("R 5 1 2 3 4 6 7 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15", "\n")   # hmental
cat("R 6 1 2 3 4 5 7 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15", "\n")   # hphysical
cat("R 1 2 3 4 5 6 7 7 8 8 9 9 10 10 11 11 12 12 13 13 14 14 15 15", "\n")   # error
cat("C 7 6 5", "\n")                                       # correlation between hphys hmental
cat("C 8 2 3", "\n")                                       # correlation between dem~pcg            
cat("C 9 2 4", "\n")                                       # correlation between dem~sle          
cat("C 10 2 5", "\n")                                      # correlation between dem~m            
cat("C 11 2 6", "\n")                                      # correlation between dem~p         
cat("C 12 3 4", "\n")                                      # correlation between pcg~sle      
cat("C 13 3 5", "\n")                                      # correlation between pcg~m        
cat("C 14 3 6", "\n")                                      # correlation between pcg~p       
cat("C 15 4 5", "\n")                                      # correlation between sle~m           
cat("C 16 4 6", "\n")                                      # correlation between sle~p           
sink()
```

Output using Delta

```{bash output for e_g_cov_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
     echo "Processing: $dir"
     cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
     ./../../../mtg2 -delta2 e_g_cov_health_all_reml_dropped_W1A.do > e_g_cov_health_all_reml_dropped_W1A.out3
     ./../../../mtg2 -delta2 e_g_cov_health_all_reml_dropped_W1B.do > e_g_cov_health_all_reml_dropped_W1B.out3
     ./../../../mtg2 -delta2 e_g_cov_health_all_reml_dropped_W4.do > e_g_cov_health_all_reml_dropped_W4.out3
done
```

#### Output table

```{r output for dropped cov model}
e_g_cov_health_all_reml_dropped_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "e_g_cov_health_all_reml_dropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", "Error")
)

e_g_cov_health_all_reml_dropped_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "e_g_cov_health_all_reml_dropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", "Genetic", "Error")
)

e_g_cov_health_all_reml_dropped_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "e_g_cov_health_all_reml_dropped",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Positive cognition", "Stressful life events", "Mental health", "Physical health", "Error")
)

e_g_cov_health_dropped_reml.output <- rbind(e_g_cov_health_all_reml_dropped_W1A.output, e_g_cov_health_all_reml_dropped_W1B.output, e_g_cov_health_all_reml_dropped_W4.output)

# contributiont
e_g_cov_health_dropped_reml.output %>%
  mutate(contribution = variance_std/variance_std_tot*100)
```

#### Plot correlations

```{r include correlations in output}
e_g_cov_health_dropped_reml.output <- e_g_cov_health_dropped_reml.output %>% 
  mutate(
    significance = ifelse(P_value < 0.05, "Significant", "Not significant"),
    shape_group = case_when(flag_corr_out_of_bounds ~ "Out of bounds",
                            TRUE ~ as.character(significance)),
    covariance = 
           c( # W1A
              "Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", "Error", "Covariance",
              "hphys ~ hmental",
              "hmental ~ hgeneral",
              "hphys ~ hgeneral",
              "frf ~ pcg",
              "frf ~ sch",
              "frf ~ hmental",  
              "frf ~ hphys",  
              "frf ~ hgeneral",  
              "pcg ~ sch",
              "pcg ~ hmental",  
              "pcg ~ hphys",  
              "pcg ~ hgeneral",  
              "sch ~ hmental",  
              "sch ~ hphys",  
              "sch ~ hgeneral",
             # W1B 
             "Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", "Genetic", "Error", "Covariance",
              "hphys ~ hmental",      
              "hmental ~ hgeneral",
              "hphys ~ hgeneral", 
              "frf ~ pcg",               
              "frf ~ sch",                 
              "frf ~ hmental",                    
              "frf ~ hphys",                    
              "frf ~ hgeneral",                    
              "pcg ~ sch",               
              "pcg ~ hmental",                    
              "pcg ~ hphys",                     
              "pcg ~ hgeneral",                     
              "sch ~ hmental",                     
              "sch ~ hphys",                     
              "sch ~ hgeneral",                     
              "grm ~ hmental",                     
              "grm ~ hphys",                     
              "grm ~ hgeneral",                     
              "grm ~ frf",                
              "grm ~ poscog",         
              "grm ~ school ",
             # W4
              "Demographics", "Positive cognition", "Stressful life events", "Mental health", "Physical health", "Error", "Covariance",
              "hphys ~ hmental",
              "dem ~ pcg",
              "dem ~ sle",
              "dem ~ hmental",
              "dem ~ hphys",
              "pcg ~ sle",
              "pcg ~ hmental",  
              "pcg ~ hphys",
              "sle ~ hmental",
              "sle ~ hphys")
             ) 
  
e_g_cov_health_dropped_reml.outputcor <- e_g_cov_health_dropped_reml.output %>% 
  filter(domain == "Correlation") %>%
  select(-variance_std_tot, -model, -parameter)

# include in dropped model
e_g_cov_health_dropped_reml.outputcor %>% 
  filter(shape_group == "Significant") %>%
  select(outcome, covariance, variance_std, P_value)
```

```{r ggplot domcov}
dropped.cor <- ggplot(e_g_cov_health_dropped_reml.outputcor, 
                      aes(x = variance_std, 
                          y = reorder(covariance, variance_std), 
                          color = outcome, 
                          shape = shape_group)) +   # <- use the new variable
  geom_point(size = 3, position = position_dodge(width = 1)) +
  geom_errorbarh(aes(xmin = CI_lower, xmax = CI_upper), height = 0.2, position = position_dodge(width = 1)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "gray50") +
  scale_shape_manual(values = c(
    "Out of bounds" = 1,      # hollow circle
    "Significant" = 16,            # solid circle
    "Not significant" = 17            # solid triangle, adjust to taste
  )) +
  labs(
    x = "Correlation Estimate",
    y = "Variable",
    color = "Outcome",
    shape = "Flagged / Significant"
  ) +
  theme_minimal(base_size = 14) +
  theme(
    legend.position = "right",
    axis.text.y = element_text(size = 10),
    panel.grid.major.y = element_blank()
  )

# ggsave(dropped.cor, file = "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/plot_domainsdropped_corE.png", height = 6, width = 10)
```

### C. All main effects with sig/OOB cors

Correlations that are significant and not out of bounds (OOB):

W1A	dem ~ poscog	
W1A	dem ~ mhealth	
W1A	dem ~ phealth	
W1A	frfam ~ poscog 	
W1A	frfam ~ school 	
W1A	frfam ~ mhealth 	
W1A	frfam ~ phealth 	
W1A	poscog ~ school 	
W1A	ghealth ~ poscog 
W1A	mhealth ~ poscog 
W1A	phealth ~ poscog 
W1A	ghealth ~ school 
W1A	mhealth ~ school 
W1A	phealth ~ school 
W1A	phealth ~ ghealth	
W1A	mhealth ~ phealth	

W1B	dem ~ poscog	
W1B	dem ~ school	
W1B	dem ~ mhealth	
W1B	frfam ~ poscog 	
W1B	frfam ~ school 	
W1B	frfam ~ mhealth 	
W1B	frfam ~ phealth 	
W1B	poscog ~ school 	
W1B	ghealth ~ poscog 
W1B	mhealth ~ poscog 
W1B	phealth ~ poscog 
W1B	mhealth ~ school 
W1B	phealth ~ school 
W1B	phealth ~ ghealth	
W1B	mhealth ~ phealth

W4 - none (all out of bounds)

```{r define lists excluding non-sig}
# W1A
rge_all_OOB_W1A.txt <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat",
                 "dem_poscog.chol.matmat2", "dem_health_mental.chol.matmat2", "dem_health_physical.chol.matmat2",           
                 "frfam_poscog.chol.matmat2", "frfam_school.chol.matmat2",  "frfam_health_mental.chol.matmat2", "frfam_health_physical.chol.matmat2",  
                 "poscog_school.chol.matmat2", "health_general_poscog.chol.matmat2", "health_mental_poscog.chol.matmat2", "health_physical_poscog.chol.matmat2", 
                 "health_general_school.chol.matmat2", "health_mental_school.chol.matmat2", "health_physical_school.chol.matmat2",
                 "health_general_health_physical.chol.matmat2",   
                 "health_mental_health_physical.chol.matmat2")

write.table(rge_all_OOB_W1A.txt, paste0(ancestry_sub_path, "domains/rge_all_OOB_W1A.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# W1B
rge_all_OOB_W1B.txt <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat",
                 "dem_poscog.chol.matmat2", "dem_school.chol.matmat2", "dem_health_mental.chol.matmat2",         
                 "frfam_poscog.chol.matmat2", "frfam_school.chol.matmat2", "frfam_health_mental.chol.matmat2", "frfam_health_physical.chol.matmat2",  
                 "poscog_school.chol.matmat2", "health_general_poscog.chol.matmat2", "health_mental_poscog.chol.matmat2", "health_physical_poscog.chol.matmat2",
                 "health_mental_school.chol.matmat2", "health_physical_school.chol.matmat2",
                 "health_general_health_physical.chol.matmat2",   
                 "health_mental_health_physical.chol.matmat2")

write.table(rge_all_OOB_W1B.txt, paste0(ancestry_sub_path, "domains/rge_all_OOB_W1B.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_rge_reml_dom_sub_OOB.txt

```{bash compute ereml and greml model}
# directories=(
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
# )
# 
# for dir in "${directories[@]}"; do
#      echo "Processing: $dir"
#      cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
# 
#       # REML for OOB dropped models
#       ./../../../mtg2 -p IDs_IID.fam -mg rge_all_OOB_W1A.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_all_reml_OOB_W1A.out > e_g_cov_all_reml_OOB_W1A.log
#       ./../../../mtg2 -p IDs_IID.fam -mg rge_all_OOB_W1B.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_cov_all_reml_OOB_W1B.out > e_g_cov_all_reml_OOB_W1B.log
#       # wave 4 had no in bounds cors
#       
#       # remove not needed columns for output
#       grep -vwE '(LKH|h2)' e_g_cov_all_reml_OOB_W1A.out > e_g_cov_all_reml_OOB_W1A.out2
#       grep -vwE '(LKH|h2)' e_g_cov_all_reml_OOB_W1B.out > e_g_cov_all_reml_OOB_W1B.out2
#   
#      echo "Finished processing $dir"
# done
```

##### Delta output prep

* ‘R’ is to get the ratio of variance
* R 2 1 3 4 (i.e. 2 / (1 + 2 + 3 + 4))

var(y) = var(e1) + var(e2) + 2cov(e1e2) + var(error)

* ‘C’ is to get the correlation of covariance
* C 5 2 3 (i.e. 5 / sqrt(2 * 3))
* Here we have the covariance estimate first, followed by the two variances used for the covariance

```{r do file for e_g_cov_all_reml_OOB_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_cov_all_reml_OOB_W1A.do")                      # create file with this name
cat("e_g_cov_all_reml_OOB_W1A.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(26, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # frfam
cat("R 4 1 2 3 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # school
cat("R 6 1 2 3 4 5 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # hmental
cat("R 8 1 2 3 4 5 6 7 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # hphysical
cat("R 9 1 2 3 4 5 6 7 8 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # hgeneral
cat("R 10 1 2 3 4 5 6 7 8 9 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # grm
cat("R 1 2 3 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # error
cat("C 11 2 4", "\n")                                      # correlation between dem~poscog
cat("C 12 2 7", "\n")                                      # correlation between dem~mhealth 
cat("C 13 2 8", "\n")                                      # correlation between dem~phealth
cat("C 14 3 4", "\n")                                      # correlation between frf~poscog          
cat("C 15 3 5", "\n")                                      # correlation between frf~sch        
cat("C 16 3 7", "\n")                                      # correlation between frf~m            
cat("C 17 3 8", "\n")                                      # correlation between frf~p   
cat("C 18 4 5", "\n")                                      # correlation between pcg~sch              
cat("C 19 4 9", "\n")                                      # correlation between pcg~g            
cat("C 20 4 7", "\n")                                      # correlation between pcg~m               
cat("C 21 4 8", "\n")                                      # correlation between pcg~p
cat("C 22 5 9", "\n")                                      # correlation between sch~g               
cat("C 23 5 7", "\n")                                      # correlation between sch~m  
cat("C 24 5 8", "\n")                                      # correlation between sch~p
cat("C 25 8 9", "\n")                                      # correlation between p~g  
cat("C 26 8 7", "\n")                                      # correlation between p~m  
sink()
```

                                                   
```{r do file for e_g_cov_all_reml_OOB_W1B}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_cov_all_reml_OOB_W1B.do")                      # create file with this name
cat("e_g_cov_all_reml_OOB_W1B.out2", "\n")               # line 1: specify the file with parameter estimates (above)
cat(25, "\n")                                               # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # frfam
cat("R 4 1 2 3 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # school
cat("R 6 1 2 3 4 5 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # hmental
cat("R 8 1 2 3 4 5 6 7 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # hphysical
cat("R 9 1 2 3 4 5 6 7 8 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # hgeneral
cat("R 10 1 2 3 4 5 6 7 8 9 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # grm
cat("R 1 2 3 4 5 6 7 8 9 10 11 11 12 12 13 13 14 14 15 15 16 16 17 17 18 18 19 19 20 20 21 21 22 22 23 23 24 24 25 25 26 26", "\n")   # error
cat("C 11 2 4", "\n")                                      # correlation between dem~poscog
cat("C 12 2 5", "\n")                                      # correlation between dem~sch 
cat("C 13 2 7", "\n")                                      # correlation between dem~mhealth
cat("C 14 3 4", "\n")                                      # correlation between frf~poscog          
cat("C 15 3 5", "\n")                                      # correlation between frf~sch        
cat("C 16 3 7", "\n")                                      # correlation between frf~m            
cat("C 17 3 8", "\n")                                      # correlation between frf~p   
cat("C 18 4 5", "\n")                                      # correlation between pcg~sch              
cat("C 19 4 9", "\n")                                      # correlation between pcg~g            
cat("C 20 4 7", "\n")                                      # correlation between pcg~m               
cat("C 21 4 8", "\n")                                      # correlation between pcg~p
cat("C 22 5 7", "\n")                                      # correlation between sch~m               
cat("C 23 5 8", "\n")                                      # correlation between sch~p
cat("C 24 8 9", "\n")                                      # correlation between p~g  
cat("C 25 8 7", "\n")                                      # correlation between p~m  
sink()
```

Output using Delta

```{bash output for OOB cov}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
     echo "Processing: $dir"
     cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
     ./../../../mtg2 -delta2 e_g_cov_all_reml_OOB_W1A.do > e_g_cov_all_reml_OOB_W1A.out3
     ./../../../mtg2 -delta2 e_g_cov_all_reml_OOB_W1B.do > e_g_cov_all_reml_OOB_W1B.out3
done
```

##### Output table

```{r lists for final labels}
r_list_W1A <- c("Demographics ~ Positive cognition", "Demographics ~ Mental health", "Demographics ~ Physical health",           
                 "Friends and family ~ Positive cognition", "Friends and family ~ School",  "Friends and family ~ Mental health", "Friends and family ~ Physical health",  
                 "Positive cognition ~ School", "General health ~ Positive cognition", "Mental health ~ Positive cognition", "Physical health ~ Positive cognition", 
                 "General health ~ School", "Mental health ~ School", "Physical health ~ School",
                 "General health ~ Physical health",   
                 "Mental health ~ Physical health")

r_list_W1B <- c("Demographics ~ Positive cognition", "Demographics ~ School", "Demographics ~ Mental health",           
                 "Friends and family ~ Positive cognition", "Friends and family ~ School",  "Friends and family ~ Mental health", "Friends and family ~ Physical health",  
                 "Positive cognition ~ School", "General health ~ Positive cognition", "Mental health ~ Positive cognition", "Physical health ~ Positive cognition", 
                 "Mental health ~ School", "Physical health ~ School",
                 "General health ~ Physical health",   
                 "Mental health ~ Physical health")
```

```{r output for dropped cov model}
 e_g_cov_all_reml_OOB_W1A.output <- extract_delta_output(
   outcome = "W1A",
   model_prefix = "e_g_cov_all_reml_OOB",
   ancestry_sub_path = ancestry_sub_path,
   domain_model = TRUE,
   domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", "Error"),
   manual_correlation_labels = r_list_W1A
 )

 e_g_cov_all_reml_OOB_W1B.output <- extract_delta_output(
   outcome = "W1B",
   model_prefix = "e_g_cov_all_reml_OOB",
   ancestry_sub_path = ancestry_sub_path,
   domain_model = TRUE,
   domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", "Error"),
   manual_correlation_labels = r_list_W1B
 )
 
# W4 IS THE SAME AS THE NO CORRELATIONS
e_g_cov_all_reml_OOB_W4.output <- domain_reml_ge_health_W4.output.est.df %>% 
  mutate(input = "Environment and genetic, rD",
         flag_neg_variance = NA,
         flag_corr_out_of_bounds = NA) 
 
e_g_cov_all_OOB_reml.output <- rbind(e_g_cov_all_reml_OOB_W1A.output, e_g_cov_all_reml_OOB_W1B.output, e_g_cov_all_reml_OOB_W4.output)
 
e_g_cov_all_OOB_reml.output %>% 
  filter(flag_corr_out_of_bounds == "FALSE", P_value < 0.05) %>%
  select(outcome, parameter, correlation_r, variance_std, P_value)
```

#### LRT

Compared to 

```{r LRT reduced E covE}
LRT_dom_cov_W1A <- LRT(path_full = "domains/e_g_cov_all_reml_OOB_W1A.out", 
                    path_constrained = "domains/domain_reml_ge_health_W1A.out",
                    Outcome = "W1A",
                    Comparison = "Domains compared to Domains covE",
                    df = 16)

LRT_dom_cov_W1B <- LRT(path_full = "domains/e_g_cov_all_reml_OOB_W1B.out", 
                    path_constrained = "domains/domain_reml_ge_health_W1B.out",
                    Outcome = "W1B",
                    Comparison = "Domains compared to Domains covE",
                    df = 15)

LRT_dom_cov <- rbind(LRT_dom_cov_W1A, LRT_dom_cov_W1B)
LRT_dom_cov
```

Nested model is y = g + e + error (siall.grmler, fewer predictors)
If significant (<0.05) = the full model (y = g + e + rge + error) provides a significantly better fit than the nested model
If non-significant (>0.05) = the nested model fits the data as well as the full model (more complex model)


## 4. ExE and GxE

### Create domains with no pca

Match the domain data with NO PCA to the N in the files 

```{r match IDs for no pca domain files}
df_list <- list(
  dem = dat.erm.theory1.dummy_dem,
  frfam = dat.erm.theory1.dummy_frfam,
  health_mental = dat.erm.theory1.dummy_health_mental,
  health_general = dat.erm.theory1.dummy_health_general,
  health_physical = dat.erm.theory1.dummy_health_physical,
  poscog = dat.erm.theory1.dummy_poscog,
  school = dat.erm.theory1.dummy_school,
  sle = dat.erm.theory1.dummy_sle
)

for(name in names(df_list)) {
  file <- df_list[[name]]

# Apply corrections (filter and attach ID variable) - removed to include full sample
filtered_df <- file %>%
  filter(AID %in% ALL_matching_ids_nosibs)   # Filter based on matching IDs

# take out ID 
filtered_df_noID <- as.matrix(filtered_df[,-1]) 

# scale the data
filtered_df_scaled <- scale(filtered_df_noID) # normalise this matrix - minus the mean and divide by SD 

# Need to remove any NA values
filtered_df_scaled <- as.data.frame(filtered_df_scaled) %>% select(where(~ !all(is.na(.))))

# make into matrix
filtered_df_scaled <- as.matrix(filtered_df_scaled)

# get number of variables
m <- dim(filtered_df_scaled)[2] # m is the number of variables - NOW 395

# divide by sqrt(m)
filtered_df_scaled <- filtered_df_scaled/sqrt(m)

head(as.data.frame(filtered_df_scaled))

# attach additional AID variable - mtg 
filtered_df_scaled <- cbind(filtered_df$AID, filtered_df$AID, filtered_df_scaled)

# Convert to numeric matrix (ensures all values are numeric)
filtered_numeric <- as.data.frame(filtered_df_scaled) %>% dplyr::rename(AID2 = V2, AID = V1) 

filtered_numeric <- filtered_numeric[, colSums(!is.na(filtered_numeric)) > 0]
# Reorder the file to match the ID file
filtered_reordered <- filtered_numeric[order(match(filtered_numeric$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

# Save the current dataframe to a file
  temp_filename <- file.path(ancestry_sub_path, paste0("/domains/dat_nopca_", name, ".txt"))
  write.table(filtered_reordered, temp_filename, row.names = FALSE, col.names = FALSE, quote = FALSE, sep="\t")
}
```

### Create interaction files 

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_domexe_sub.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann:

```{bash erm and reml per domain}
# # Define directories
#  directories=(
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
#  )
#  
#  # Loop through each directory
#  for dir in "${directories[@]}"; do
#      echo "Processing directory: $dir"
#      
#      # Change to the target directory
#      cd "$dir" || { echo "Failed to enter directory: $dir"; continue; }
#  
#      # Step 1: Run all ERM commands for all files first
#      ./../../../mtg2 -p IDs_IID.fam -pdmx dat_nopca_dem.txt -thread 100 -out theory1_clean_dummy_dem_nopca.e
#      ./../../../mtg2 -p IDs_IID.fam -pdmx dat_nopca_frfam.txt -thread 100 -out theory1_clean_dummy_frfam_nopca.e
#      ./../../../mtg2 -p IDs_IID.fam -pdmx dat_nopca_health_mental.txt -thread 100 -out theory1_clean_dummy_health_mental_nopca.e
#      ./../../../mtg2 -p IDs_IID.fam -pdmx dat_nopca_health_physical.txt -thread 100 -out theory1_clean_dummy_health_physical_nopca.e
#      ./../../../mtg2 -p IDs_IID.fam -pdmx dat_nopca_health_general.txt -thread 100 -out theory1_clean_dummy_health_general_nopca.e
#      ./../../../mtg2 -p IDs_IID.fam -pdmx dat_nopca_poscog.txt -thread 100 -out theory1_clean_dummy_poscog_nopca.e
#      ./../../../mtg2 -p IDs_IID.fam -pdmx dat_nopca_school.txt -thread 100 -out theory1_clean_dummy_school_nopca.e
#      ./../../../mtg2 -p IDs_IID.fam -pdmx dat_nopca_sle.txt -thread 100 -out theory1_clean_dummy_sle_nopca.e
#      
#      # create interaction term from erm x erm for each domain combination 
#      paste theory1_clean_dummy_dem_nopca.e theory1_clean_dummy_frfam_nopca.e | awk '{ print $1, $2, $3 * $7 }' > dem_frfam.exe
#      paste theory1_clean_dummy_dem_nopca.e theory1_clean_dummy_health_mental_nopca.e | awk '{ print $1, $2, $3 * $7 }' > dem_mhealth.exe
#      paste theory1_clean_dummy_dem_nopca.e theory1_clean_dummy_health_physical_nopca.e | awk '{ print $1, $2, $3 * $7 }' > dem_phealth.exe
#      paste theory1_clean_dummy_dem_nopca.e theory1_clean_dummy_health_general_nopca.e | awk '{ print $1, $2, $3 * $7 }' > dem_ghealth.exe
#      paste theory1_clean_dummy_dem_nopca.e theory1_clean_dummy_poscog_nopca.e | awk '{ print $1, $2, $3 * $7 }' > dem_poscog.exe
#      paste theory1_clean_dummy_dem_nopca.e theory1_clean_dummy_school_nopca.e | awk '{ print $1, $2, $3 * $7 }' > dem_school.exe
#      paste theory1_clean_dummy_dem_nopca.e theory1_clean_dummy_sle_nopca.e | awk '{ print $1, $2, $3 * $7 }' > dem_sle.exe
#      paste theory1_clean_dummy_frfam_nopca.e theory1_clean_dummy_health_mental_nopca.e | awk '{ print $1, $2, $3 * $7 }' > frfam_mhealth.exe
#      paste theory1_clean_dummy_frfam_nopca.e theory1_clean_dummy_health_physical_nopca.e | awk '{ print $1, $2, $3 * $7 }' > frfam_phealth.exe
#      paste theory1_clean_dummy_frfam_nopca.e theory1_clean_dummy_health_general_nopca.e | awk '{ print $1, $2, $3 * $7 }' > frfam_ghealth.exe
#      paste theory1_clean_dummy_frfam_nopca.e theory1_clean_dummy_poscog_nopca.e | awk '{ print $1, $2, $3 * $7 }' > frfam_poscog.exe
#      paste theory1_clean_dummy_frfam_nopca.e theory1_clean_dummy_school_nopca.e | awk '{ print $1, $2, $3 * $7 }' > frfam_school.exe
#      paste theory1_clean_dummy_frfam_nopca.e theory1_clean_dummy_sle_nopca.e | awk '{ print $1, $2, $3 * $7 }' > frfam_sle.exe
#      paste theory1_clean_dummy_health_mental_nopca.e theory1_clean_dummy_health_physical_nopca.e | awk '{ print $1, $2, $3 * $7 }' > mhealth_phealth.exe
#      paste theory1_clean_dummy_health_mental_nopca.e theory1_clean_dummy_health_general_nopca.e | awk '{ print $1, $2, $3 * $7 }' > mhealth_ghealth.exe
#      paste theory1_clean_dummy_health_mental_nopca.e theory1_clean_dummy_poscog_nopca.e | awk '{ print $1, $2, $3 * $7 }' > mhealth_poscog.exe
#      paste theory1_clean_dummy_health_mental_nopca.e theory1_clean_dummy_school_nopca.e | awk '{ print $1, $2, $3 * $7 }' > mhealth_school.exe
#      paste theory1_clean_dummy_health_mental_nopca.e theory1_clean_dummy_sle_nopca.e | awk '{ print $1, $2, $3 * $7 }' > mhealth_sle.exe
#      paste theory1_clean_dummy_health_physical_nopca.e theory1_clean_dummy_health_general_nopca.e | awk '{ print $1, $2, $3 * $7 }' > phealth_ghealth.exe
#      paste theory1_clean_dummy_health_physical_nopca.e theory1_clean_dummy_poscog_nopca.e | awk '{ print $1, $2, $3 * $7 }' > phealth_poscog.exe
#      paste theory1_clean_dummy_health_physical_nopca.e theory1_clean_dummy_school_nopca.e | awk '{ print $1, $2, $3 * $7 }' > phealth_school.exe
#      paste theory1_clean_dummy_health_physical_nopca.e theory1_clean_dummy_sle_nopca.e | awk '{ print $1, $2, $3 * $7 }' > phealth_sle.exe
#      paste theory1_clean_dummy_health_general_nopca.e theory1_clean_dummy_poscog_nopca.e | awk '{ print $1, $2, $3 * $7 }' > ghealth_poscog.exe
#      paste theory1_clean_dummy_health_general_nopca.e theory1_clean_dummy_school_nopca.e | awk '{ print $1, $2, $3 * $7 }' > ghealth_school.exe
#      paste theory1_clean_dummy_health_general_nopca.e theory1_clean_dummy_sle_nopca.e | awk '{ print $1, $2, $3 * $7 }' > ghealth_sle.exe
#      paste theory1_clean_dummy_poscog_nopca.e theory1_clean_dummy_school_nopca.e | awk '{ print $1, $2, $3 * $7 }' > poscog_school.exe
#      paste theory1_clean_dummy_poscog_nopca.e theory1_clean_dummy_sle_nopca.e | awk '{ print $1, $2, $3 * $7 }' > poscog_sle.exe
#      paste theory1_clean_dummy_school_nopca.e theory1_clean_dummy_sle_nopca.e | awk '{ print $1, $2, $3 * $7 }' > school_sle.exe
#      
#      # create interaction term from domE x grm 
#      paste theory1_clean_dummy_dem_nopca.e grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > dem_grm.gxe
#      paste theory1_clean_dummy_frfam_nopca.e grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > frfam_grm.gxe
#      paste theory1_clean_dummy_health_mental_nopca.e grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > mhealth_grm.gxe
#      paste theory1_clean_dummy_health_physical_nopca.e grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > phealth_grm.gxe
#      paste theory1_clean_dummy_health_general_nopca.e  grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > ghealth_grm.gxe
#      paste theory1_clean_dummy_poscog_nopca.e grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > poscog_grm.gxe
#      paste theory1_clean_dummy_school_nopca.e grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > school_grm.gxe
#      paste theory1_clean_dummy_sle_nopca.e grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > sle_grm.gxe
#      
#  done
```

### Create lists of interactions

Create txt files with lists of domain interactions.

```{r create e_g txt file}
exe_dom_W1A <- c("erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical",
             "frfam_poscog.exe", "frfam_school.exe", "frfam_ghealth.exe", "frfam_mhealth.exe", "frfam_phealth.exe", 
             "poscog_school.exe", "ghealth_poscog.exe", "mhealth_poscog.exe", "phealth_poscog.exe", 
             "ghealth_school.exe", "mhealth_school.exe", "phealth_school.exe", 
             "mhealth_ghealth.exe", "phealth_ghealth.exe", 
             "mhealth_phealth.exe") 

write.table(exe_dom_W1A, paste0(ancestry_sub_path, "domains/exe_dom_W1A.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

exe_dom_W1B <- c("erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat",
             "frfam_poscog.exe", "frfam_school.exe", "frfam_ghealth.exe", "frfam_mhealth.exe", "frfam_phealth.exe", "frfam_grm.gxe",
             "poscog_school.exe", "ghealth_poscog.exe", "mhealth_poscog.exe", "phealth_poscog.exe", "poscog_grm.gxe",
             "ghealth_school.exe", "mhealth_school.exe", "phealth_school.exe", "school_grm.gxe",
             "mhealth_ghealth.exe", "phealth_ghealth.exe", "ghealth_grm.gxe",
             "mhealth_phealth.exe", "mhealth_grm.gxe",
             "phealth_grm.gxe") 

write.table(exe_dom_W1B, paste0(ancestry_sub_path, "domains/exe_dom_W1B.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

exe_dom_W4 <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical",
                "dem_poscog.exe", "dem_sle.exe", "dem_mhealth.exe", "dem_phealth.exe",
                "poscog_sle.exe", "mhealth_poscog.exe", "phealth_poscog.exe",
                "mhealth_sle.exe", "phealth_sle.exe", 
                "mhealth_phealth.exe") 

write.table(exe_dom_W4, paste0(ancestry_sub_path, "domains/exe_dom_W4.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# all interactions 
exe_dom_ALL <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat",
                 "dem_frfam.exe", "dem_mhealth.exe", "dem_phealth.exe", "dem_ghealth.exe", "dem_poscog.exe", "dem_school.exe", "dem_sle.exe",         
                 "frfam_mhealth.exe", "frfam_phealth.exe", "frfam_ghealth.exe", "frfam_poscog.exe", "frfam_school.exe", "frfam_sle.exe", 
                 "mhealth_phealth.exe", "mhealth_ghealth.exe", "mhealth_poscog.exe", "mhealth_school.exe", "mhealth_sle.exe", 
                 "phealth_ghealth.exe", "phealth_poscog.exe", "phealth_school.exe", "phealth_sle.exe",
                 "ghealth_poscog.exe", "ghealth_school.exe", "ghealth_sle.exe", 
                 "poscog_school.exe", "poscog_sle.exe",
                 "school_sle.exe", 
                 "dem_grm.gxe", "frfam_grm.gxe", "mhealth_grm.gxe", "phealth_grm.gxe", "ghealth_grm.gxe", "poscog_grm.gxe", "school_grm.gxe", "sle_grm.gxe"
                 ) 

write.table(exe_dom_ALL, paste0(ancestry_sub_path, "domains/exe_dom_ALL.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

### REML

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_domexe_reml_dem_sub.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann:

```{bash domexe_reml_dem}
#  # Define directories
#  directories=(
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
#  )
#  
# # Loop through each directory
# for dir in "${directories[@]}"; do
#     echo "Processing directory: $dir"
#     
#     # Change to the target directory
#     cd "$dir" || { echo "Failed to enter directory: $dir"; continue; }
#     
#     # Compute E-REML
#     ./../../../mtg2 -p IDs_IID.fam -mg exe_dom_W1A.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_domexe_reml_W1A.out > e_g_domexe_reml_W1A.log
#     ./../../../mtg2 -p IDs_IID.fam -mg exe_dom_W1B.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_domexe_reml_W1B.out > e_g_domexe_reml_W1B.log
#     ./../../../mtg2 -p IDs_IID.fam -mg exe_dom_W4.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out  e_g_domexe_reml_W4.out >  e_g_domexe_reml_W4.log
#     
# done
```

### A. All exe and gxe - SUPPLEMENT

#### Delta output prep 

```{bash pick out LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_domexe_ALL_reml_W1A.out > e_g_domexe_ALL_reml_W1A.out2
       grep -vwE '(LKH|h2)' e_g_domexe_ALL_reml_W1B.out > e_g_domexe_ALL_reml_W1B.out2
       grep -vwE '(LKH|h2)' e_g_domexe_ALL_reml_W4.out > e_g_domexe_ALL_reml_W4.out2
done
```

```{r do file for e_g_domexe_ALL_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_domexe_ALL_reml_W1A.do")          # create file with this name
cat("e_g_domexe_ALL_reml_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(46, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frf
cat("R 4 1 2 3 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to sch
cat("R 6 1 2 3 4 5 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to hgeneral
cat("R 8 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to hmental
cat("R 9 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to hphysical
cat("R 10 1 2 3 4 5 6 7 8 9 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to grm
cat("R 11 1 2 3 4 5 6 7 8 9 10 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_frfam exe
cat("R 12 1 2 3 4 5 6 7 8 9 10 11 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_mhealth exe
cat("R 13 1 2 3 4 5 6 7 8 9 10 11 12 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_phealth exe
cat("R 14 1 2 3 4 5 6 7 8 9 10 11 12 13 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_ghealth exe
cat("R 15 1 2 3 4 5 6 7 8 9 10 11 12 13 14 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_poscog exe
cat("R 16 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_school exe
cat("R 17 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_sle exe
cat("R 18 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_mhealth exe
cat("R 19 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_phealth exe
cat("R 20 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_ghealth exe
cat("R 21 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_poscog exe
cat("R 22 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to frfam_school exe
cat("R 23 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to frfam_sle exe
cat("R 24 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_phealth exe
cat("R 25 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_ghealth exe
cat("R 26 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_poscog exe
cat("R 27 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_school exe
cat("R 28 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_sle exe
cat("R 29 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_ghealth exe
cat("R 30 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_poscog exe
cat("R 31 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_school exe
cat("R 32 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_sle exe
cat("R 33 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to ghealth_poscog exe
cat("R 34 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to ghealth_school exe
cat("R 35 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to ghealth_sle exe
cat("R 36 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 37 38 39 40 41 42 43 44 45 46", "\n")     # due to poscog_school exe
cat("R 37 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 38 39 40 41 42 43 44 45 46", "\n")     # due to poscog_sle exe
cat("R 38 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 39 40 41 42 43 44 45 46", "\n")     # due to school_sle exe
cat("R 39 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 40 41 42 43 44 45 46", "\n")     # due to dem_grm gxe 
cat("R 40 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 41 42 43 44 45 46", "\n")     # due to frf_grm gxe
cat("R 41 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 42 43 44 45 46", "\n")     # due to mhealth_grm gxe
cat("R 42 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 43 44 45 46", "\n")     # due to phealth_grm gxe
cat("R 43 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 44 45 46", "\n")     # due to ghealth_grm gxe
cat("R 44 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 45 46", "\n")     # due to poscog_grm gxe
cat("R 45 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 46", "\n")     # due to school_grm gxe
cat("R 46 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45", "\n")     # due to sle_grm gxe
cat("R 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to error
sink()
```

```{r do file for e_g_domexe_ALL_reml_W1B}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_domexe_ALL_reml_W1B.do")          # create file with this name
cat("e_g_domexe_ALL_reml_W1B.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(46, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frf
cat("R 4 1 2 3 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to sch
cat("R 6 1 2 3 4 5 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to hgeneral
cat("R 8 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to hmental
cat("R 9 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to hphysical
cat("R 10 1 2 3 4 5 6 7 8 9 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to grm
cat("R 11 1 2 3 4 5 6 7 8 9 10 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_frfam exe
cat("R 12 1 2 3 4 5 6 7 8 9 10 11 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_mhealth exe
cat("R 13 1 2 3 4 5 6 7 8 9 10 11 12 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_phealth exe
cat("R 14 1 2 3 4 5 6 7 8 9 10 11 12 13 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_ghealth exe
cat("R 15 1 2 3 4 5 6 7 8 9 10 11 12 13 14 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_poscog exe
cat("R 16 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_school exe
cat("R 17 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_sle exe
cat("R 18 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_mhealth exe
cat("R 19 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_phealth exe
cat("R 20 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_ghealth exe
cat("R 21 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_poscog exe
cat("R 22 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to frfam_school exe
cat("R 23 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to frfam_sle exe
cat("R 24 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_phealth exe
cat("R 25 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_ghealth exe
cat("R 26 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_poscog exe
cat("R 27 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_school exe
cat("R 28 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_sle exe
cat("R 29 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_ghealth exe
cat("R 30 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_poscog exe
cat("R 31 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_school exe
cat("R 32 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_sle exe
cat("R 33 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to ghealth_poscog exe
cat("R 34 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to ghealth_school exe
cat("R 35 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to ghealth_sle exe
cat("R 36 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 37 38 39 40 41 42 43 44 45 46", "\n")     # due to poscog_school exe
cat("R 37 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 38 39 40 41 42 43 44 45 46", "\n")     # due to poscog_sle exe
cat("R 38 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 39 40 41 42 43 44 45 46", "\n")     # due to school_sle exe
cat("R 39 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 40 41 42 43 44 45 46", "\n")     # due to dem_grm gxe 
cat("R 40 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 41 42 43 44 45 46", "\n")     # due to frf_grm gxe
cat("R 41 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 42 43 44 45 46", "\n")     # due to mhealth_grm gxe
cat("R 42 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 43 44 45 46", "\n")     # due to phealth_grm gxe
cat("R 43 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 44 45 46", "\n")     # due to ghealth_grm gxe
cat("R 44 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 45 46", "\n")     # due to poscog_grm gxe
cat("R 45 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 46", "\n")     # due to school_grm gxe
cat("R 46 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45", "\n")     # due to sle_grm gxe
cat("R 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to error
sink()
```

```{r do file for e_g_domexe_ALL_reml_W4}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_domexe_ALL_reml_W4.do")          # create file with this name
cat("e_g_domexe_ALL_reml_W4.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(46, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frf
cat("R 4 1 2 3 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to sch
cat("R 6 1 2 3 4 5 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to hgeneral
cat("R 8 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to hmental
cat("R 9 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to hphysical
cat("R 10 1 2 3 4 5 6 7 8 9 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to grm
cat("R 11 1 2 3 4 5 6 7 8 9 10 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_frfam exe
cat("R 12 1 2 3 4 5 6 7 8 9 10 11 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_mhealth exe
cat("R 13 1 2 3 4 5 6 7 8 9 10 11 12 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_phealth exe
cat("R 14 1 2 3 4 5 6 7 8 9 10 11 12 13 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_ghealth exe
cat("R 15 1 2 3 4 5 6 7 8 9 10 11 12 13 14 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_poscog exe
cat("R 16 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_school exe
cat("R 17 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to dem_sle exe
cat("R 18 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_mhealth exe
cat("R 19 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_phealth exe
cat("R 20 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_ghealth exe
cat("R 21 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")  # due to frfam_poscog exe
cat("R 22 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to frfam_school exe
cat("R 23 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to frfam_sle exe
cat("R 24 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_phealth exe
cat("R 25 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_ghealth exe
cat("R 26 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_poscog exe
cat("R 27 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_school exe
cat("R 28 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to mhealth_sle exe
cat("R 29 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_ghealth exe
cat("R 30 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_poscog exe
cat("R 31 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_school exe
cat("R 32 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to phealth_sle exe
cat("R 33 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to ghealth_poscog exe
cat("R 34 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to ghealth_school exe
cat("R 35 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to ghealth_sle exe
cat("R 36 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 37 38 39 40 41 42 43 44 45 46", "\n")     # due to poscog_school exe
cat("R 37 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 38 39 40 41 42 43 44 45 46", "\n")     # due to poscog_sle exe
cat("R 38 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 39 40 41 42 43 44 45 46", "\n")     # due to school_sle exe
cat("R 39 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 40 41 42 43 44 45 46", "\n")     # due to dem_grm gxe 
cat("R 40 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 41 42 43 44 45 46", "\n")     # due to frf_grm gxe
cat("R 41 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 42 43 44 45 46", "\n")     # due to mhealth_grm gxe
cat("R 42 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 43 44 45 46", "\n")     # due to phealth_grm gxe
cat("R 43 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 44 45 46", "\n")     # due to ghealth_grm gxe
cat("R 44 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 45 46", "\n")     # due to poscog_grm gxe
cat("R 45 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 46", "\n")     # due to school_grm gxe
cat("R 46 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45", "\n")     # due to sle_grm gxe
cat("R 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46", "\n")     # due to error
sink()
```

```{bash output for e_g_gxe_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_domexe_ALL_reml_W1A.do > e_g_domexe_ALL_reml_W1A.out3
    ./../../../mtg2 -delta2 e_g_domexe_ALL_reml_W1B.do > e_g_domexe_ALL_reml_W1B.out3
    ./../../../mtg2 -delta2 e_g_domexe_ALL_reml_W4.do > e_g_domexe_ALL_reml_W4.out3
done
```

```{r labels for interactions}
exe_ALL_labels <- c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic",
                    "Demographics x Friends and family", "Demographics x Mental health", "Demographics x Physical health", "Demographics x General health", "Demographics x Positive cognition", "Demographics x School", "Demographics x Stressful life events",         
                    "Friends and family x Mental health", "Friends and family x Physical health", "Friends and family x General health", "Friends and family x Positive cognition", "Friends and family x School", "Friends and family x Stressful life events", 
                    "Mental health x Physical health", "Mental health x General health", "Mental health x Positive cognition", "Mental health x School", "Mental health x Stressful life events", 
                    "Physical health x General health", "Physical health x Positive cognition", "Physical health x School", "Physical health x Stressful life events",
                    "General health x Positive cognition", "General health x School", "General health x Stressful life events", 
                    "Positive cognition x School", "Positive cognition x Stressful life events",
                    "School x Stressful life events", 
                    "Demographics x Genetic", "Friends and family x Genetic", "Mental health x Genetic", "Physical health x Genetic", "General health x Genetic", "Positive cognition x Genetic", "School x Genetic", "Stressful life events x Genetic",
                    "Error") 
```

#### Output table

```{r output for dropped domains}
# W1A
e_g_domexe_ALL_reml_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "e_g_domexe_ALL_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", rep("Environment-environment interaction", 28), rep("Gene-environment interaction", 8), "Error")
)

# add interaction description
e_g_domexe_ALL_reml_W1A.output <- e_g_domexe_ALL_reml_W1A.output %>%
  mutate(interaction = exe_ALL_labels)

# W1B
e_g_domexe_ALL_reml_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "e_g_domexe_ALL_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", rep("Environment-environment interaction", 28), rep("Gene-environment interaction", 8), "Error")
)

# add interaction description
e_g_domexe_ALL_reml_W1B.output <- e_g_domexe_ALL_reml_W1B.output %>%
  mutate(interaction = exe_ALL_labels)

# W4
e_g_domexe_ALL_reml_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "e_g_domexe_ALL_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic", rep("Environment-environment interaction", 28), rep("Gene-environment interaction", 8), "Error")
)

# add interaction description
e_g_domexe_ALL_reml_W4.output <- e_g_domexe_ALL_reml_W4.output %>%
  mutate(interaction = exe_ALL_labels)

# combine
e_g_domexe_ALL_reml.output <- rbind(e_g_domexe_ALL_reml_W1A.output, e_g_domexe_ALL_reml_W1B.output, e_g_domexe_ALL_reml_W4.output)

# contribution
e_g_domexe_ALL_reml.output %>%
  mutate(contribution = variance_std/variance_std_tot*100)

# significant interaction
e_g_domexe_ALL_reml.output %>%
  filter((domain == "Environment-environment interaction" | domain == "Gene-environment interaction")
           & flag_neg_variance == "FALSE"
           & P_value < 0.05
           ) %>% 
  select(outcome, interaction, variance_std, P_value)
```

### B. Dropped/reduced exe, gxe - SUPPLEMENT

#### Delta output prep 

```{bash pick out LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_domexe_reml_W1A.out > e_g_domexe_reml_W1A.out2
       grep -vwE '(LKH|h2)' e_g_domexe_reml_W1B.out > e_g_domexe_reml_W1B.out2
       grep -vwE '(LKH|h2)' e_g_domexe_reml_W4.out > e_g_domexe_reml_W4.out2
done
```

```{r do file for e_g_domexe_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_domexe_reml_W1A.do")          # create file with this name
cat("e_g_domexe_reml_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(22, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22", "\n")               # line 3: compute prop. of variance due to frfam
cat("R 3 1 2 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22", "\n")               # line 4: compute prop. of variance due to poscog
cat("R 4 1 2 3 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22", "\n")               # line 4: compute prop. of variance due to school
cat("R 5 1 2 3 4 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22", "\n")               # line 4: compute prop. of variance due to ghealth
cat("R 6 1 2 3 4 5 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22", "\n")               # line 4: compute prop. of variance due to mhealth
cat("R 7 1 2 3 4 5 6 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22", "\n")               # line 4: compute prop. of variance due to phealth
cat("R 8 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16 17 18 19 20 21 22", "\n")               # line 4: compute prop. of variance due to frfam_poscog exe
cat("R 9 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 17 18 19 20 21 22", "\n")               # line 4: compute prop. of variance due to frfam_school exe
cat("R 10 1 2 3 4 5 6 7 8 9 11 12 13 14 15 16 17 18 19 20 21 22", "\n")              # line 4: compute prop. of variance due to frfam_ghealth exe
cat("R 11 1 2 3 4 5 6 7 8 9 10 12 13 14 15 16 17 18 19 20 21 22", "\n")              # line 4: compute prop. of variance due to frfam_mhealth exe
cat("R 12 1 2 3 4 5 6 7 8 9 10 11 13 14 15 16 17 18 19 20 21 22", "\n")              # line 4: compute prop. of variance due to frfam_phealth exe
cat("R 13 1 2 3 4 5 6 7 8 9 10 11 12 14 15 16 17 18 19 20 21 22", "\n")              # line 4: compute prop. of variance due to poscog_school exe
cat("R 14 1 2 3 4 5 6 7 8 9 10 11 12 13 15 16 17 18 19 20 21 22", "\n")              # line 4: compute prop. of variance due to ghealth_poscog exe
cat("R 15 1 2 3 4 5 6 7 8 9 10 11 12 13 14 16 17 18 19 20 21 22", "\n")              # line 4: compute prop. of variance due to mhealth_poscog exe
cat("R 16 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 17 18 19 20 21 22", "\n")              # line 4: compute prop. of variance due to phealth_poscog exe
cat("R 17 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 18 19 20 21 22", "\n")              # line 4: compute prop. of variance due to ghealth_school exe
cat("R 18 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 19 20 21 22", "\n")              # line 4: compute prop. of variance due to mhealth_school exe
cat("R 19 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 20 21 22", "\n")              # line 4: compute prop. of variance due to phealth_school exe
cat("R 20 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 21 22", "\n")              # line 4: compute prop. of variance due to mhealth_ghealth exe
cat("R 21 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 22", "\n")              # line 4: compute prop. of variance due to phealth_ghealth exe
cat("R 22 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21", "\n")              # line 4: compute prop. of variance due to mhealth_phealth exe
cat("R 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22", "\n")               # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_domexe_reml_W1B}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_domexe_reml_W1B.do")          # create file with this name
cat("e_g_domexe_reml_W1B.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(29, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")               # line 3: compute prop. of variance due to frfam
cat("R 3 1 2 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")               # line 4: compute prop. of variance due to poscog
cat("R 4 1 2 3 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")               # line 4: compute prop. of variance due to school
cat("R 5 1 2 3 4 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")               # line 4: compute prop. of variance due to ghealth
cat("R 6 1 2 3 4 5 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")               # line 4: compute prop. of variance due to mhealth
cat("R 7 1 2 3 4 5 6 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")               # line 4: compute prop. of variance due to phealth
cat("R 8 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")               # line 4: compute prop. of variance due to grm
cat("R 9 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")               # line 4: compute prop. of variance due to frfam_poscog exe
cat("R 10 1 2 3 4 5 6 7 8 9 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")               # line 4: compute prop. of variance due to frfam_school exe
cat("R 11 1 2 3 4 5 6 7 8 9 10 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to frfam_ghealth exe
cat("R 12 1 2 3 4 5 6 7 8 9 10 11 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to frfam_mhealth exe
cat("R 13 1 2 3 4 5 6 7 8 9 10 11 12 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to frfam_phealth exe
cat("R 14 1 2 3 4 5 6 7 8 9 10 11 12 13 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to frfam_grm gxe 
cat("R 15 1 2 3 4 5 6 7 8 9 10 11 12 13 14 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to poscog_school exe
cat("R 16 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to ghealth_poscog exe
cat("R 17 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 18 19 20 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to mhealth_poscog exe
cat("R 18 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 19 20 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to phealth_poscog exe
cat("R 19 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 20 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to poscog_grm gxe
cat("R 20 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to ghealth_school exe
cat("R 21 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to mhealth_school exe
cat("R 22 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to phealth_school exe
cat("R 23 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to school_grm gxe
cat("R 24 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to mhealth_ghealth exe
cat("R 25 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 26 27 28 29", "\n")              # line 4: compute prop. of variance due to phealth_ghealth exe
cat("R 26 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 28 29", "\n")              # line 4: compute prop. of variance due to ghealth_grm gxe
cat("R 27 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 28 29", "\n")              # line 4: compute prop. of variance due to mhealth_phealth exe
cat("R 28 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 29", "\n")              # line 4: compute prop. of variance due to mhealth_grm gxe
cat("R 29 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28", "\n")              # line 4: compute prop. of variance due to phealth_grm gxe
cat("R 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29", "\n")              # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_domexe_reml_W4}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_domexe_reml_W4.do")          # create file with this name
cat("e_g_domexe_reml_W4.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(16, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 12 13 14 15 16", "\n")               # line 3: compute prop. of variance due to dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 12 13 14 15 16", "\n")               # line 4: compute prop. of variance due to poscog
cat("R 4 1 2 3 5 6 7 8 9 10 11 12 13 14 15 16", "\n")               # line 4: compute prop. of variance due to sle
cat("R 5 1 2 3 4 6 7 8 9 10 11 12 13 14 15 16", "\n")               # line 4: compute prop. of variance due to mhealth
cat("R 6 1 2 3 4 5 7 8 9 10 11 12 13 14 15 16", "\n")               # line 4: compute prop. of variance due to phealth
cat("R 7 1 2 3 4 5 6 8 9 10 11 12 13 14 15 16", "\n")               # line 4: compute prop. of variance due to dem_poscog exe
cat("R 8 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16", "\n")               # line 4: compute prop. of variance due to dem_sle exe
cat("R 9 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16", "\n")               # line 4: compute prop. of variance due to dem_mhealth exe
cat("R 10 1 2 3 4 5 6 7 8 9 11 12 13 14 15 16", "\n")               # line 4: compute prop. of variance due to dem_phealth exe
cat("R 11 1 2 3 4 5 6 7 8 9 10 12 13 14 15 16", "\n")               # line 4: compute prop. of variance due to poscog_sle exe
cat("R 12 1 2 3 4 5 6 7 8 9 10 11 13 14 15 16", "\n")               # line 4: compute prop. of variance due to mhealth_poscog exe
cat("R 13 1 2 3 4 5 6 7 8 9 10 11 12 14 15 16", "\n")               # line 4: compute prop. of variance due to phealth_poscog exe
cat("R 14 1 2 3 4 5 6 7 8 9 10 11 12 13 15 16", "\n")               # line 4: compute prop. of variance due to mhealth_sle exe
cat("R 15 1 2 3 4 5 6 7 8 9 10 11 12 13 14 16", "\n")               # line 4: compute prop. of variance due to phealth_sle exe
cat("R 16 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15", "\n")               # line 4: compute prop. of variance due to mhealth_phealth exe
cat("R 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16", "\n")               # line 4: compute prop. of variance due to error
sink()
```

Output using Delta

```{bash output for e_g_gxe_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_domexe_reml_W1A.do > e_g_domexe_reml_W1A.out3
    ./../../../mtg2 -delta2 e_g_domexe_reml_W1B.do > e_g_domexe_reml_W1B.out3
    ./../../../mtg2 -delta2 e_g_domexe_reml_W4.do > e_g_domexe_reml_W4.out3
done
```

#### Output table 

```{r interaction labels}
exe_W1A_labels <- c("Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health",
                    "Friends and family x Positive cognition", "Friends and family x School", "Friends and family x General health", "Friends and family x Mental health", "Friends and family x Physical health", 
                    "Positive cognition x School", "General health x Positive cognition", "Mental health x Positive cognition", "Physical health x Positive cognition", 
                    "General health x School", "Mental health x School", "Physical health x School", 
                    "Mental health x General health", "Physical health x General health", 
                    "Mental health x Physical health",
                    "Error") 

exe_W1B_labels <- c("Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", "Genetic",
                    "Friends and family x Positive cognition", "Friends and family x School", "Friends and family x General health", "Friends and family x Mental health", "Friends and family x Physical health", "Friends and family x Genetic",
                    "Positive cognition x School", "General health x Positive cognition", "Mental health x Positive cognition", "Physical health x Positive cognition", "Positive cognition x Genetic",
                    "General health x School", "Mental health x School", "Physical health x School", "School x Genetic",
                    "Mental health x General health", "Physical health x General health", "General health x Genetic",
                    "Mental health x Physical health", "Mental health x Genetic",
                    "Physical health x Genetic",
                    "Error") 

exe_W4_labels <- c("Demographics", "Positive cognition", "Stressful life events", "Mental health", "Physical health",
                   "Demographics x Positive cognition", "Demographics x Stressful life events", "Demographics x Mental health", "Demographics x Physical health",
                   "Positive cognition x Stressful life events", "Mental health x Positive cognition", "Physical health x Positive cognition",
                   "Mental health x Stressful life events", "Physical health x Stressful life events", 
                   "Mental health x Physical health",
                   "Error") 
```

```{r output for dropped domains}
# W1A
e_g_domexe_reml_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "e_g_domexe_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", rep("Environment-environment interaction", 15), "Error")
)

# add interaction description
e_g_domexe_reml_W1A.output <- e_g_domexe_reml_W1A.output %>%
  mutate(interaction = exe_W1A_labels)

# W1B
e_g_domexe_reml_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "e_g_domexe_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Friends and family", "Positive cognition", "School", "General health", "Mental health", "Physical health", "Genetic", 
                    rep("Environment-environment interaction", 5), "Gene-environment interaction", 
                    rep("Environment-environment interaction", 4), "Gene-environment interaction", 
                    rep("Environment-environment interaction", 3), "Gene-environment interaction", 
                    rep("Environment-environment interaction", 2), "Gene-environment interaction", 
                    "Environment-environment interaction", "Gene-environment interaction", "Gene-environment interaction", 
                    "Error")
)

# add interaction description
e_g_domexe_reml_W1B.output <- e_g_domexe_reml_W1B.output %>%
  mutate(interaction = exe_W1B_labels)

# W4
e_g_domexe_reml_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "e_g_domexe_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Positive cognition", "Stressful life events", "Mental health", "Physical health", rep("Environment-environment interaction", 10), "Error")
)

# add interaction description
e_g_domexe_reml_W4.output <- e_g_domexe_reml_W4.output %>%
  mutate(interaction = exe_W4_labels)

# combine
e_g_domexe_reml.output <- rbind(e_g_domexe_reml_W1A.output, e_g_domexe_reml_W1B.output, e_g_domexe_reml_W4.output)

# contribution
e_g_domexe_reml.output %>%
  mutate(contribution = variance_std/variance_std_tot*100)

# significant interaction
e_g_domexe_reml.output %>%
  filter((domain == "Environment-environment interaction" | domain == "Gene-environment interaction")
           & flag_neg_variance == "FALSE"
           & P_value < 0.05
           ) %>% 
  select(outcome, interaction, variance_std, P_value)
```

#### LRT

```{r LRT reduced E covE}
LRT_dome_domexe_W1A <- LRT(path_full = "domains/e_g_domexe_reml_W1A.out", 
                           path_constrained = "domains/domain_reml_ge_health_dropped_W1A.out",
                           Outcome = "W1A",
                           Comparison = "Domains reduced E compared to Domains reduced ExE",
                           df = 15)

LRT_dome_domexe_W1B <- LRT(path_full = "domains/e_g_domexe_reml_W1B.out", 
                          path_constrained = "domains/domain_reml_ge_health_dropped_W1B.out",
                          Outcome = "W1B",
                          Comparison = "Domains reduced E compared to Domains reduced ExE",
                          df = 21)

LRT_dome_domexe_W4 <- LRT(path_full = "domains/e_g_domexe_reml_W4.out", 
                          path_constrained = "domains/domain_reml_ge_health_dropped_W4.out",
                          Outcome = "W4",
                          Comparison = "Domains reduced E compared to Domains reduced ExE",
                          df = 10)

LRT_dome_domexe <- rbind(LRT_dome_domexe_W1A, LRT_dome_domexe_W1B, LRT_dome_domexe_W4)
LRT_dome_domexe
```

```{r add up interaction contribution}
e_g_domexe_reml.output %>%
  filter(domain == "Environment-environment interaction" | domain == "Gene-environment interaction") %>%
  group_by(outcome) %>%
  summarise(total_variance = sum(variance_std, na.rm = TRUE))

e_g_domexe_reml.output %>%
  filter(domain == "Environment-environment interaction") %>%
  group_by(outcome) %>%
  summarise(total_variance = sum(variance_std, na.rm = TRUE))

e_g_domexe_reml.output %>%
  filter(domain == "Gene-environment interaction") %>%
  group_by(outcome) %>%
  summarise(total_variance = sum(variance_std, na.rm = TRUE))
```

### C. FINAL: all main effects with sig/OOB interations

We will take forward only the significant interaction effects that did not have a negative variance from the ALL model. 

##### Create lists of interactions

Significant interactions with no negative variances:
W1A dem x phealth
W1A dem x school
W1A	frfam x sle	 * giving neg var so have removed
W1A	phealth x school	
W1A	school x sle	
W1A	frfam x grm	
W1A	mhealth x grm	
W1A	phealth x grm	
W1A	ghealth x grm	
W1A	poscog x grm	

W1B	school x sle	
W1B	frfam x grm	
W1B	mhealth x grm	
W1B	phealth x grm	
W1B	poscog x grm	

W4	none - same as main effects model 

```{r create e_g txt file}
exe_dom_posvar_W1A <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat", 
              "dem_phealth.exe", "dem_school.exe", 
              "phealth_school.exe", 
              "school_sle.exe", 
              "frfam_grm.gxe", "mhealth_grm.gxe", "phealth_grm.gxe", "ghealth_grm.gxe", "poscog_grm.gxe") 

write.table(exe_dom_posvar_W1A, paste0(ancestry_sub_path, "domains/exe_dom_posvar_W1A.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

exe_dom_posvar_W1B <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat", 
                        "school_sle.exe", 
                        "frfam_grm.gxe", "mhealth_grm.gxe", "phealth_grm.gxe", "poscog_grm.gxe") 

write.table(exe_dom_posvar_W1B, paste0(ancestry_sub_path, "domains/exe_dom_posvar_W1B.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

##### REML

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_domexe_reml_dem_sub_posvar.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann:

```{bash domexe_reml_dem_sub_posvar}
# # Define directories
#  directories=(
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
#  )
#  
#  # Loop through each directory
#  for dir in "${directories[@]}"; do
#      echo "Processing directory: $dir"
#      
#      # Change to the target directory
#      cd "$dir" || { echo "Failed to enter directory: $dir"; continue; }
#      
#      # Compute E-REML
#      ./../../../mtg2 -p IDs_IID.fam -mg exe_dom_posvar_W1A.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_domexe_posvar_reml_W1A.out > e_g_domexe_posvar_reml_W1A.log
#      ./../../../mtg2 -p IDs_IID.fam -mg exe_dom_posvar_W1B.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_domexe_posvar_reml_W1B.out > e_g_domexe_posvar_reml_W1B.log
#      # there is no W4 taken forward here
#  done
```

##### Delta output prep 

```{bash pick out LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_domexe_posvar_reml_W1A.out > e_g_domexe_posvar_reml_W1A.out2
       grep -vwE '(LKH|h2)' e_g_domexe_posvar_reml_W1B.out > e_g_domexe_posvar_reml_W1B.out2
done
```
     
    
```{r do file for e_g_domexe_posvar_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_domexe_posvar_reml_W1A.do")          # create file with this name
cat("e_g_domexe_posvar_reml_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(19, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19", "\n")               # due to dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19", "\n")               # due to frfam
cat("R 4 1 2 3 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19", "\n")               # due to poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 12 13 14 15 16 17 18 19", "\n")               # due to school
cat("R 6 1 2 3 4 5 7 8 9 10 11 12 13 14 15 16 17 18 19", "\n")               # due to sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 12 13 14 15 16 17 18 19", "\n")               # due to ghealth
cat("R 8 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16 17 18 19", "\n")               # due to mhealth
cat("R 9 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 17 18 19", "\n")               # due to phealth 
cat("R 10 1 2 3 4 5 6 7 8 9 11 12 13 14 15 16 17 18 19", "\n")               # due to grm
cat("R 11 1 2 3 4 5 6 7 8 9 10 12 13 14 15 16 17 18 19", "\n")               # due to dem_phealth exe
cat("R 12 1 2 3 4 5 6 7 8 9 10 11 13 14 15 16 17 18 19", "\n")               # due to dem_school exe
cat("R 13 1 2 3 4 5 6 7 8 9 10 11 12 14 15 16 17 18 19", "\n")               # due to phealth_school exe
cat("R 14 1 2 3 4 5 6 7 8 9 10 11 12 13 15 16 17 18 19", "\n")               # due to school_sle exe
cat("R 15 1 2 3 4 5 6 7 8 9 10 11 12 13 14 16 17 18 19", "\n")               # due to frfam_grm gxe 
cat("R 16 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 17 18 19", "\n")               # due to mhealth_grm gxe
cat("R 17 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 18 19", "\n")               # due to phealth_grm gxe
cat("R 18 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 19", "\n")               # due to ghealth_grm gxe
cat("R 19 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18", "\n")               # due to poscog_grm gxe
cat("R 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19", "\n")               # due to error
sink()
```

```{r do file for e_g_domexe_posvar_reml_W1B}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_domexe_posvar_reml_W1B.do")          # create file with this name
cat("e_g_domexe_posvar_reml_W1B.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(15, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 12 13 14 15", "\n")               # due to dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 12 13 14 15", "\n")               # due to frfam
cat("R 4 1 2 3 5 6 7 8 9 10 11 12 13 14 15", "\n")               # due to poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 12 13 14 15", "\n")               # due to school
cat("R 6 1 2 3 4 5 7 8 9 10 11 12 13 14 15", "\n")               # due to sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 12 13 14 15", "\n")               # due to ghealth
cat("R 8 1 2 3 4 5 6 7 9 10 11 12 13 14 15", "\n")               # due to mhealth
cat("R 9 1 2 3 4 5 6 7 8 10 11 12 13 14 15", "\n")               # due to phealth
cat("R 10 1 2 3 4 5 6 7 8 9 11 12 13 14 15", "\n")               # due to grm
cat("R 11 1 2 3 4 5 6 7 8 9 10 12 13 14 15", "\n")               # due to school_sle.exe
cat("R 12 1 2 3 4 5 6 7 8 9 10 11 13 14 15", "\n")               # due to frfam_grm.gxe
cat("R 13 1 2 3 4 5 6 7 8 9 10 11 12 14 15", "\n")               # due to mhealth_grm.gxe
cat("R 14 1 2 3 4 5 6 7 8 9 10 11 12 13 15", "\n")               # due to phealth_grm.gxe
cat("R 15 1 2 3 4 5 6 7 8 9 10 11 12 13 14", "\n")               # due to poscog_grm.gxe
cat("R 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15", "\n")               # due to error
sink()
```

Output using Delta

```{bash output for e_g_gxe_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_domexe_posvar_reml_W1A.do > e_g_domexe_posvar_reml_W1A.out3
    ./../../../mtg2 -delta2 e_g_domexe_posvar_reml_W1B.do > e_g_domexe_posvar_reml_W1B.out3
done
```

#### LRT

DxD compared to Domains

```{r LRT reduced E covE}
LRT_OOB_domexe_W1A <- LRT(path_full = "domains/e_g_domexe_posvar_reml_W1A.out", 
                           path_constrained = "domains/domain_reml_ge_health_W1A.out",
                           Outcome = "W1A",
                           Comparison = "Domains compared to Domains OOB ExE",
                           df = 9)

LRT_OOB_domexe_W1B <- LRT(path_full = "domains/e_g_domexe_posvar_reml_W1B.out", 
                          path_constrained = "domains/domain_reml_ge_health_W1B.out",
                          Outcome = "W1B",
                          Comparison = "Domains compared to Domains OOB ExE",
                          df = 5)

LRT_OOB_domexe <- rbind(LRT_OOB_domexe_W1A, LRT_OOB_domexe_W1B)
LRT_OOB_domexe
```

#### Output table 

```{r interaction labels}
exe_posvar_W1A_labels <- c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic",
                    "Demographics x Physical health", "Demographics x School", 
                    "Physical health x School", 
                    "School x Stressful life events", 
                    "Friends and family x Genetic", "Mental health x Genetic", "Physical health x Genetic", "General health x Genetic", "Positive cognition x Genetic",
                    "Error") 

exe_posvar_W1B_labels <- c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic",
                           "School x Stressful life events",
                           "Friends and family x Genetic", "Mental health x Genetic", "Physical health x Genetic", "Positive cognition x Genetic",
                           "Error") 
```

```{r output for dropped domains}
# W1A
e_g_domexe_posvar_reml_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "e_g_domexe_posvar_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic",
                    rep("Environment-environment interaction", 4),
                    rep("Gene-environment interaction", 5), 
                    "Error")
)

# add interaction description
e_g_domexe_posvar_reml_W1A.output <- e_g_domexe_posvar_reml_W1A.output %>%
  mutate(interaction = exe_posvar_W1A_labels)

# W1B
e_g_domexe_posvar_reml_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "e_g_domexe_posvar_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic",
                    "Environment-environment interaction",
                    rep("Gene-environment interaction", 4), 
                    "Error")
)

# add interaction description
e_g_domexe_posvar_reml_W1B.output <- e_g_domexe_posvar_reml_W1B.output %>%
  mutate(interaction = exe_posvar_W1B_labels)

# W4 IS THE SAME AS THE NO INTERACTION
e_g_domexe_posvar_reml_W4.output <- domain_reml_ge_health_W4.output.est.df %>% 
  mutate(input = "Total environment and genetic, DxD",
         flag_neg_variance = NA,
         flag_corr_out_of_bounds = NA, 
         interaction = NA) 

# combine
e_g_domexe_posvar_reml.output <- rbind(e_g_domexe_posvar_reml_W1A.output, e_g_domexe_posvar_reml_W1B.output, e_g_domexe_posvar_reml_W4.output)

# contribution
e_g_domexe_posvar_reml.output %>%
  mutate(contribution = variance_std/variance_std_tot*100)

e_g_domexe_posvar_reml.output %>%
  select(outcome, domain, interaction, variance_std, P_value)
```

```{r add up interaction contribution}
e_g_domexe_posvar_reml.output %>%
  filter(domain == "Environment-environment interaction" | domain == "Gene-environment interaction") %>%
  filter(outcome == "W1A") %>%
  summarise(total_variance = sum(variance_std, na.rm = TRUE))

e_g_domexe_posvar_reml.output %>%
  filter(domain == "Environment-environment interaction" | domain == "Gene-environment interaction") %>%
  filter(outcome == "W1B") %>%
  summarise(total_variance = sum(variance_std, na.rm = TRUE))
```

#### Plot interactions

```{r include correlations in output}
color_blind_palette <- c("#56B4E9", "#E69F00")

e_g_exe_OOB_reml.cor <- e_g_domexe_posvar_reml.output %>% 
  mutate(
    significance = ifelse(P_value < 0.05, "Significant", "Not significant"),
    shape_group = case_when(flag_corr_out_of_bounds ~ "Out of bounds",
                            TRUE ~ as.character(significance))) %>%
  filter((domain == "Environment-environment interaction" | domain == "Gene-environment interaction"))
  
exe_OOB <- ggplot(e_g_exe_OOB_reml.cor, 
                      aes(x = variance_std, 
                          y = reorder(interaction, variance_std), 
                          color = outcome, 
                          shape = shape_group)) +   # <- use the new variable
  geom_point(size = 3, position = position_dodge(width = 1)) +
  geom_errorbarh(aes(xmin = CI_lower, xmax = CI_upper), height = 0.2, position = position_dodge(width = 1)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "gray50") +
  scale_shape_manual(values = c(
    "Out of bounds" = 1,      # hollow circle
    "Significant" = 16,            # solid circle
    "Not significant" = 17            # solid triangle, adjust to taste
  )) +
  scale_color_manual(values = color_blind_palette) +
  labs(
    x = "Variance explained by interaction",
    y = "",
    color = "Outcome",
    shape = "Flagged / Significant"
  ) +
  theme_classic(base_size = 16) +
  scale_x_continuous(limits = c(-0.1, 0.5)) +
  theme(
    legend.position = "bottom",
    legend.title = element_blank(),
    legend.text = element_text(size = 14),
    legend.key.width = unit(1.5, "cm"),
    axis.title.y = element_text(size = 16, margin = margin(r = 10)),
    axis.text.x = element_text(size = 12, angle = 30, hjust = 1),
    axis.text.y = element_text(size = 15),
    plot.title = element_text(size = 16, face = "bold"),
    plot.subtitle = element_text(size = 14),
    plot.margin = margin(10, 10, 10, 10)
  )

save_gg_cairo(plot = exe_OOB,  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/plot_exe_OOB.png",
              width = 800, height = 500)

save_gg_cairo(plot = exe_OOB,  "~/findmedep/figures/plot_exe_OOB.png",
              width = 800, height = 500)
```

### D. GxE and rGE

We now want to test whether the interaction effects reduce when including rGE 

Significant interactions with no negative variances:
W1A dem x phealth
W1A dem x school
W1A	frfam x sle	 * giving neg var so have removed
W1A	phealth x school	
W1A	school x sle	
W1A	frfam x grm	
W1A	mhealth x grm	
W1A	phealth x grm	
W1A	ghealth x grm	
W1A	poscog x grm	

W1B	school x sle	
W1B	frfam x grm	
W1B	mhealth x grm	
W1B	phealth x grm	
W1B	poscog x grm	

W4	none - same as main effects model 

Correlations to include to control for confounding rGE in the GxE

mhealth ~ grm
phealth ~ grm 
poscog ~ grm
frfam ~ grm

*No cors or interactions for W4*

```{r create gxe cov txt file}
exe_cov_dom_posvar_W1A <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat", 
              "dem_phealth.exe", "dem_school.exe", 
              "phealth_school.exe", 
              "school_sle.exe", 
              "frfam_grm.gxe", "mhealth_grm.gxe", "phealth_grm.gxe", "ghealth_grm.gxe", "poscog_grm.gxe",
              "health_mental.grm.chol.matmat2", "health_physical.grm.chol.matmat2", "poscog.grm.chol.matmat2", "frfam.grm.chol.matmat2") 

write.table(exe_cov_dom_posvar_W1A, paste0(ancestry_sub_path, "domains/exe_cov_dom_posvar_W1A.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

exe_cov_dom_posvar_W1B <- c("erm_dat_theory1_clean_dummy_dem", "erm_dat_theory1_clean_dummy_frfam", "erm_dat_theory1_clean_dummy_poscog", "erm_dat_theory1_clean_dummy_school", "erm_dat_theory1_clean_dummy_sle", "erm_dat_theory1_clean_dummy_health_general", "erm_dat_theory1_clean_dummy_health_mental", "erm_dat_theory1_clean_dummy_health_physical", "grm_full_out.grm_reformat", 
                        "school_sle.exe", 
                        "frfam_grm.gxe", "mhealth_grm.gxe", "phealth_grm.gxe", "poscog_grm.gxe",
                        "health_mental.grm.chol.matmat2", "health_physical.grm.chol.matmat2", "poscog.grm.chol.matmat2", "frfam.grm.chol.matmat2") 

write.table(exe_cov_dom_posvar_W1B, paste0(ancestry_sub_path, "domains/exe_cov_dom_posvar_W1B.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

##### REML

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_domexe_cov_reml_dem_sub_posvar.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann:

```{bash domexe_cov_reml_dem_sub_posvar}
#  # Define directories
#  directories=(
#      "/pdfs/rche/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
#      "/pdfs/rche/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
#      "/pdfs/rche/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
#      "/pdfs/rche/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
#      "/pdfs/rche/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
#  )
#  
#  # Loop through each directory
#  for dir in "${directories[@]}"; do
#      echo "Processing directory: $dir"
#      
#      # Change to the target directory
#      cd "$dir" || { echo "Failed to enter directory: $dir"; continue; }
#      
#      # Compute E-REML
#      ./../../../mtg2 -p IDs_IID.fam -mg exe_cov_dom_posvar_W1A.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_domexe_cov_posvar_reml_W1A.out > e_g_domexe_cov_posvar_reml_W1A.log
#      ./../../../mtg2 -p IDs_IID.fam -mg exe_cov_dom_posvar_W1B.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_domexe_cov_posvar_reml_W1B.out > e_g_domexe_cov_posvar_reml_W1B.log
#      # there is no W4 taken forward here
#  done
```

##### Delta output prep 

```{bash pick out LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_domexe_cov_posvar_reml_W1A.out > e_g_cov_domexe_posvar_reml_W1A.out2
       grep -vwE '(LKH|h2)' e_g_domexe_cov_posvar_reml_W1B.out > e_g_cov_domexe_posvar_reml_W1B.out2
done
```
     
```{r do file for e_g_domexe_cov_posvar_reml_W1A}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_cov_domexe_posvar_reml_W1A.do")          # create file with this name
cat("e_g_cov_domexe_posvar_reml_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(23, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to frfam
cat("R 4 1 2 3 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to school
cat("R 6 1 2 3 4 5 7 8 9 10 11 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to ghealth
cat("R 8 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to mhealth
cat("R 9 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to phealth 
cat("R 10 1 2 3 4 5 6 7 8 9 11 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to grm
cat("R 11 1 2 3 4 5 6 7 8 9 10 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to dem_phealth exe
cat("R 12 1 2 3 4 5 6 7 8 9 10 11 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to dem_school exe
cat("R 13 1 2 3 4 5 6 7 8 9 10 11 12 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to phealth_school exe
cat("R 14 1 2 3 4 5 6 7 8 9 10 11 12 13 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to school_sle exe
cat("R 15 1 2 3 4 5 6 7 8 9 10 11 12 13 14 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to frfam_grm gxe 
cat("R 16 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to mhealth_grm gxe
cat("R 17 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 18 19 20 20 21 21 22 22 23 23", "\n")               # due to phealth_grm gxe
cat("R 18 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 19 20 20 21 21 22 22 23 23", "\n")               # due to ghealth_grm gxe
cat("R 19 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 20 20 21 21 22 22 23 23", "\n")               # due to poscog_grm gxe
cat("R 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 20 21 21 22 22 23 23", "\n")               # due to error
cat("C 20 8 10", "\n")                                      # correlation between mhealth ~ grm
cat("C 21 9 10", "\n")                                      # correlation between phealth ~ grm  
cat("C 22 4 10", "\n")                                      # correlation between poscog ~ grm
cat("C 23 3 10", "\n")                                      # correlation between frfam ~ grm          
sink()
```

```{r do file for e_g_domexe_cov_posvar_reml_W1B}
setwd(paste0(ancestry_sub_path, "domains/"))
sink("e_g_cov_domexe_posvar_reml_W1B.do")          # create file with this name
cat("e_g_cov_domexe_posvar_reml_W1B.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(19, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4 5 6 7 8 9 10 11 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to dem
cat("R 3 1 2 4 5 6 7 8 9 10 11 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to frfam
cat("R 4 1 2 3 5 6 7 8 9 10 11 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to poscog
cat("R 5 1 2 3 4 6 7 8 9 10 11 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to school
cat("R 6 1 2 3 4 5 7 8 9 10 11 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to sle
cat("R 7 1 2 3 4 5 6 8 9 10 11 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to ghealth
cat("R 8 1 2 3 4 5 6 7 9 10 11 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to mhealth
cat("R 9 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to phealth
cat("R 10 1 2 3 4 5 6 7 8 9 11 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to grm
cat("R 11 1 2 3 4 5 6 7 8 9 10 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to school_sle.exe
cat("R 12 1 2 3 4 5 6 7 8 9 10 11 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to frfam_grm.gxe
cat("R 13 1 2 3 4 5 6 7 8 9 10 11 12 14 15 16 16 17 17 18 18 19 19", "\n")               # due to mhealth_grm.gxe
cat("R 14 1 2 3 4 5 6 7 8 9 10 11 12 13 15 16 16 17 17 18 18 19 19", "\n")               # due to phealth_grm.gxe
cat("R 15 1 2 3 4 5 6 7 8 9 10 11 12 13 14 16 16 17 17 18 18 19 19", "\n")               # due to poscog_grm.gxe
cat("R 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 16 17 17 18 18 19 19", "\n")               # due to error
cat("C 16 8 10", "\n")                                      # correlation between mhealth ~ grm
cat("C 17 9 10", "\n")                                      # correlation between phealth ~ grm  
cat("C 18 4 10", "\n")                                      # correlation between poscog ~ grm
cat("C 19 3 10", "\n")                                      # correlation between frfam ~ grm 
sink()
```

Output using Delta

```{bash output for e_g_gxe_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/domains/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/domains/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_cov_domexe_posvar_reml_W1A.do > e_g_cov_domexe_posvar_reml_W1A.out3
    ./../../../mtg2 -delta2 e_g_cov_domexe_posvar_reml_W1B.do > e_g_cov_domexe_posvar_reml_W1B.out3
done
```

#### LRT

```{r LRT reduced E covE}
LRT_OOB_domexe_cov_W1A <- LRT(path_full = "domains/e_g_domexe_cov_posvar_reml_W1A.out", 
                           path_constrained = "domains/e_g_domexe_posvar_reml_W1A.out",
                           Outcome = "W1A",
                           Comparison = "Domains compared to Domains OOB ExE",
                           df = 4)

LRT_OOB_domexe_cov_W1B <- LRT(path_full = "domains/e_g_domexe_cov_posvar_reml_W1B.out", 
                          path_constrained = "domains/e_g_domexe_posvar_reml_W1B.out",
                          Outcome = "W1B",
                          Comparison = "Domains compared to Domains OOB ExE",
                          df = 4)

LRT_OOB_domexe_cov <- rbind(LRT_OOB_domexe_cov_W1A, LRT_OOB_domexe_cov_W1B)
LRT_OOB_domexe_cov
```

#### Output table 

```{r interaction labels}
exe_cov_posvar_W1A_labels <- c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic",
                    "Demographics x Physical health", "Demographics x School", 
                    "Physical health x School", 
                    "School x Stressful life events", 
                    "Friends and family x Genetic", "Mental health x Genetic", "Physical health x Genetic", "General health x Genetic", "Positive cognition x Genetic",
                    "Error",
                    "Covariance", rep("Correlation", 4)) 

exe_cov_posvar_W1B_labels <- c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic",
                           "School x Stressful life events",
                           "Friends and family x Genetic", "Mental health x Genetic", "Physical health x Genetic", "Positive cognition x Genetic",
                           "Error",
                           "Covariance", rep("Correlation", 4))

r_list <- c("Mental health ~ Genetic", "Physical health ~ Genetic", "Positive cognition ~ Genetic", "Friends and family ~ Genetic")
```

```{r output for dropped domains}
# W1A
e_g_domexe_cov_posvar_reml_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "e_g_cov_domexe_posvar_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic",
                    rep("Environment-environment interaction", 4),
                    rep("Gene-environment interaction", 5), 
                    "Error"),
  manual_correlation_labels = r_list
)

# add interaction description
e_g_domexe_cov_posvar_reml_W1A.output <- e_g_domexe_cov_posvar_reml_W1A.output %>%
  mutate(interaction = exe_cov_posvar_W1A_labels)

# W1B
e_g_domexe_cov_posvar_reml_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "e_g_cov_domexe_posvar_reml",
  ancestry_sub_path = ancestry_sub_path,
  domain_model = TRUE,
  domain_labels = c("Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic",
                    "Environment-environment interaction",
                    rep("Gene-environment interaction", 4), 
                    "Error"),
  manual_correlation_labels = r_list
)

# add interaction description
e_g_domexe_cov_posvar_reml_W1B.output <- e_g_domexe_cov_posvar_reml_W1B.output %>%
  mutate(interaction = exe_cov_posvar_W1B_labels)

# W4 IS THE SAME AS THE NO INTERACTION
e_g_domexe_cov_posvar_reml_W4.output <- domain_reml_ge_health_W4.output.est.df %>% 
  mutate(input = "Total environment and genetic, DxD and rD",
         flag_neg_variance = NA,
         flag_corr_out_of_bounds = NA, 
         interaction = NA) 

# combine
e_g_domexe_cov_posvar_reml.output <- rbind(e_g_domexe_cov_posvar_reml_W1A.output, e_g_domexe_cov_posvar_reml_W1B.output, e_g_domexe_cov_posvar_reml_W4.output)
```

#### Plot interactions

```{r include correlations in output}
e_g_cov_gxe_reml.cor <- e_g_domexe_cov_posvar_reml.output %>% 
  mutate(
    significance = ifelse(P_value < 0.05, "Significant", "Not significant"),
    shape_group = case_when(flag_corr_out_of_bounds ~ "Out of bounds",
                            TRUE ~ as.character(significance))) %>%
  filter((domain == "Environment-environment interaction" | domain == "Gene-environment interaction"))
  
gxe_cor.plot <- ggplot(e_g_cov_gxe_reml.cor, 
                      aes(x = variance_std, 
                          y = reorder(interaction, variance_std), 
                          color = outcome, 
                          shape = shape_group)) +   # <- use the new variable
  geom_point(size = 3, position = position_dodge(width = 1)) +
  geom_errorbarh(aes(xmin = CI_lower, xmax = CI_upper), height = 0.2, position = position_dodge(width = 1)) +
  geom_vline(xintercept = 0, linetype = "dashed", color = "gray50") +
  scale_shape_manual(values = c(
    "Out of bounds" = 1,      # hollow circle
    "Significant" = 16,            # solid circle
    "Not significant" = 17            # solid triangle, adjust to taste
  )) +
  scale_color_manual(values = color_blind_palette) +
  labs(
    x = "Variance explained by interaction",
    y = "",
    color = "Outcome",
    shape = "Flagged / Significant"
  ) +
  theme_classic(base_size = 16) +
  scale_x_continuous(limits = c(-0.1, 0.5)) +
  theme(
    legend.position = "bottom",
    legend.title = element_blank(),
    legend.text = element_text(size = 14),
    legend.key.width = unit(1.5, "cm"),
    axis.title.y = element_text(size = 16, margin = margin(r = 10)),
    axis.text.x = element_text(size = 12, angle = 30, hjust = 1),
    axis.text.y = element_text(size = 15),
    plot.title = element_text(size = 16, face = "bold"),
    plot.subtitle = element_text(size = 14),
    plot.margin = margin(10, 10, 10, 10)
  )

save_gg_cairo(plot = gxe_cor.plot,  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/plot_interactions_cov.png",
              width = 800, height = 500)

save_gg_cairo(plot = gxe_cor.plot,  "~/findmedep/figures/plot_interactions_cov.png",
              width = 800, height = 500)
```

***

# Plot

e_g_reml.output.plot (B, D) - only include E and joint model here 
e_g_cov_gxe_reml.output (E)
e_g_gxe_reml.output (F)
domain_reml_ge_health.output (H1)
e_g_cov_all_OOB_reml.output (H3)
e_g_domexe_posvar_reml.output (G4)

```{r all domain models in one dataframe}
e_g_reml.output.plot <- e_g_reml.output %>% filter(input == "Total environment" | input == "Total environment and genetic")

combined_domains_cov_plot.data <- bind_rows(
e_g_reml.output.plot, 
e_g_gxe_reml.output,
domain_reml_ge_health.output,
e_g_cov_all_OOB_reml.output,
e_g_domexe_posvar_reml.output,
#e_g_domexe_cov_posvar_reml.output
)

combined_domains_cov_plot.data_filtered <- combined_domains_cov_plot.data %>% 
  filter(domain != "Error" & domain != "Correlation") %>%
  select(outcome, domain, input, variance_std, variance_std_tot, P_value)

# Determine all unique domain levels
all_domains <- c("Gene-environment interaction", "Environment-environment interaction", "Environment", "Covariance", "Demographics", "Friends and family", "Positive cognition", "School", "Stressful life events", "General health", "Mental health", "Physical health", "Genetic")   
all_input <- unique(combined_domains_cov_plot.data_filtered$input)

# Convert `domain` to a factor with consistent levels
combined_domains_cov_plot.data_filtered$domain <- factor(combined_domains_cov_plot.data_filtered$domain, levels = all_domains)
combined_domains_cov_plot.data_filtered$input <- factor(combined_domains_cov_plot.data_filtered$input, levels = all_input)

combined_domains_cov_plot.data_filtered <- combined_domains_cov_plot.data_filtered %>%
  mutate(significant = ifelse(P_value < 0.05, "Significant", "Not Significant"))

# Separate the significant and non-significant results
combined_domains_cov_plot.data_filtered_sig <- combined_domains_cov_plot.data_filtered %>%
  filter(P_value < 0.05)

combined_domains_cov_plot.data_filtered_nonsig <- combined_domains_cov_plot.data_filtered %>%
  filter(P_value >= 0.05)


N_female <- nrow(dat %>% filter(AID %in% ALL_matching_ids_nosibs) %>% filter(BIO_SEX_factor_clean == "Female"))
N_male <- nrow(dat %>% filter(AID %in% ALL_matching_ids_nosibs) %>% filter(BIO_SEX_factor_clean == "Male")) 
```

```{r plot colours}
domain_colors <- c(
    "#ADD8E6", # gxe
    "#E0BAA6", # domExE
    "#F0C8FF", # enviro
    "#fffee0", # Covariance
    "#C34A36", # dem
    "#FF7F00", # frfam
    "#F4C2C2", # pos cog
    "#CAB2D6", # school
    "#FFFF99", # sle
    "#2E8B57", # g health
    "#B2DF8A", # m health
    "#32CD32", # p health
    "#1F78B4"  # genetic
)
```

```{r joint domain plots}
# Create summarized data first
text_labels <- combined_domains_cov_plot.data_filtered %>%
  dplyr::group_by(outcome, input) %>%
  dplyr::summarize(
    variance_std_total = sum(variance_std, na.rm = TRUE), 
    .groups = "drop"
  )

plot_combined_domains_cov <- ggplot(combined_domains_cov_plot.data_filtered, aes(x = outcome, y = variance_std, fill = domain)) +
  geom_bar(stat = "identity", position = "stack", width = 0.7) +
  geom_text(
    data = text_labels,
    aes(x = outcome, y = variance_std_total + 0.08, label = scales::percent(variance_std_total, accuracy = 1)),
    inherit.aes = FALSE,
    size = 4,
    color = "black"
  ) +
  scale_y_continuous(
    limits = c(0, 1),
    labels = scales::percent_format(accuracy = 1)
  ) +
  scale_fill_manual(values = domain_colors) +
  facet_wrap(~input, scales = "free", ncol = 3) +
  labs(
    x = "",
    y = "Variance Explained (%)"
  ) +
  theme_classic(base_size = 14) +
  theme(
    legend.position = "bottom",
    legend.title = element_blank(),
    legend.text = element_text(size = 10),
    legend.key.width = unit(1.5, "cm"),
    axis.title.y = element_text(size = 16, margin = margin(r = 10)),
    axis.text.x = element_text(angle = 30, hjust = 1),
    plot.title = element_text(size = 16, face = "bold"),
    plot.subtitle = element_text(size = 14),
    plot.margin = margin(10, 10, 10, 10)
  )

# Add facet labels A, B, C...
plot_combined_domains_cov_tagged <- egg::tag_facet(
  plot_combined_domains_cov,
  open = "", close = "", tag_pool = LETTERS
)

# Save the updated plot
save_gg_cairo(plot = plot_combined_domains_cov_tagged,  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/plot_combined_domains_cov.png",
              width = 900, height = 700)

# Save the updated plot
save_gg_cairo(plot = plot_combined_domains_cov_tagged,  "~/findmedep/figures/plot_combined_domains_cov.png",
              width = 900, height = 700)
```

# Table - SUPPLEMENT

```{r table output of all models}
table.output <- bind_rows(
e_g_reml.output, 
e_g_cov_reml.output,
e_g_gxe_reml.output,
e_g_cov_gxe_reml.output,
domain_reml_ge_health.output,
domain_reml_ge_health_dropped.output,
# domain_reml_ge_health_seqdropped.output,
e_g_cov_all_reml.output,
e_g_cov_health_dropped_reml.output,
e_g_cov_all_OOB_reml.output,
e_g_domexe_ALL_reml.output,
e_g_domexe_reml.output,
e_g_domexe_posvar_reml.output,
e_g_domexe_cov_posvar_reml.output
)

table.output <- table.output %>% 
  mutate(
    construct = 
      case_when(
        .default = domain,
        domain == "Correlation" ~ parameter,
        domain == "Gene-environment interaction" | domain == "Environment-environment interaction" ~ interaction,
      ) 
  ) %>%
  mutate(construct = if_else(is.na(construct), "Genetic x Environment", construct)) 

table.output <- table.output %>%
  select(outcome, input, construct, variance_std, SE, CI_lower, CI_upper, P_value, variance_std_tot, correlation_r, flag_neg_variance, flag_corr_out_of_bounds) %>%
  rename(model = input) %>%
  rename(input = construct)
```

```{r table formatting}
# Step 1: Filter and format CI strings
table.formatted <- table.output %>%
  filter(outcome %in% c("W1A", "W1B", "W4")) %>%
  filter(!is.na(variance_std)) %>%
  mutate(
    ci = paste0(
      round(variance_std, 4), 
      " [", round(CI_lower, 4), ", ", round(CI_upper, 4), "]"
    )
  )

# Step 2: Pivot to wide format
table.formatted <- table.formatted %>%
  select(model, input, outcome, ci) %>%
  pivot_wider(
    names_from = outcome,
    values_from = ci
  ) %>%
  rename(Factor = input) %>%
  arrange(factor(model, levels = unique(table.formatted$model)))
```


***

# Sensitivity analyses 

## 1. Heritability of all factors - SUPPLEMENT

Create outcome files for every variable.

```{r create loop for file with each of the 80 predictors as outcomes}
dat.erm.theory1.dummy.herit <- dat.erm.theory1.dummy %>% select(-AID, -H1FP21_3_numeric_raw, - PA55_clean_impute) # remove the variable that is the same for everytone and also the continuous income variable
dat.erm.theory1.herit <- dat.erm.theory1 %>% select(-AID, -H1FP21_3_numeric_raw, - PA55_clean_impute)
dat.erm.theory1.dummy.heritID <- dat.erm.theory1.dummy %>% select(-H1FP21_3_numeric_raw, - PA55_clean_impute)
variable_names <- colnames(dat.erm.theory1.herit)
```

```{r count how many item levels to inform hertiability check}
# count unique non-NA levels per variable
likert_summary <- dat.erm.theory1.herit %>%
  summarise(across(everything(), ~ n_distinct(na.omit(.)))) %>%
  pivot_longer(cols = everything(),
               names_to = "item",
               values_to = "n_levels") %>%
  mutate(status = case_when(
    n_levels >= 5 & n_levels <= 7 ~ "OK (5–7 levels)",
    n_levels < 5 ~ "Too few levels (ordinal/binary)",
    n_levels > 7 ~ "Continuous"
  )) %>%
  arrange(desc(n_levels))

likert_summary
```

We will calculate the heritability of all the dummy variables individually as this is what has gone into the ERM. 

```{r calculate prevelence in sample}
trait_prevalence <- dat.erm.theory1.dummy.herit %>%
  summarise(across(everything(), ~ mean(.x, na.rm = TRUE))) %>%
  pivot_longer(cols = everything(), names_to = "trait", values_to = "prevalence")

write.table(trait_prevalence, paste0(ancestry_sub_path, "sensitivity/trait_prevalence.tsv"), quote = FALSE, row.names = FALSE, sep = "\t")
```

```{r create erm data with only matching ids}
# filter to only those that are matching for erm data 
dat.erm.theory1_sens <- as.data.frame(dat.erm.theory1.dummy.heritID) %>%
  filter(AID %in% ALL_matching_ids_nosibs) 

dim(dat.erm.theory1_sens)

# attach additional AID variable - mtg 
dat.erm.theory1_sens_mtg <- cbind(dat.erm.theory1_sens$AID, dat.erm.theory1_sens)
dat.erm.theory1_sens_mtg <- dat.erm.theory1_sens_mtg %>% dplyr::rename(AID2 = AID, AID = `dat.erm.theory1_sens$AID`) 

head(dat.erm.theory1_sens_mtg)

# reorder to match genetic file 
dat.erm.theory1_sens_reordered <- dat.erm.theory1_sens_mtg[order(match(dat.erm.theory1_sens_mtg$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

head(dat.erm.theory1_sens_reordered)

# check they are the same order
identical(as.numeric(matching_id_grm_erm_FID$AID), dat.erm.theory1_sens_reordered$AID) # these should be the same - so now need to run the ERM to match and remember to use this ID file as the input file when running the model 

# make all numeric
dat.erm.theory1_sens_reordered_numeric <- apply(dat.erm.theory1_sens_reordered, 2, as.numeric)
is.matrix(dat.erm.theory1_sens_reordered_numeric) && all(sapply(dat.erm.theory1_sens_reordered_numeric, is.numeric)) # check matrix is numeric

# have a look at data file
as.tibble(dat.erm.theory1_sens_reordered_numeric)
head(matching_id_grm_erm_AID)

# write pheno table
write.table(dat.erm.theory1_sens_reordered_numeric, paste0(ancestry_sub_path, "sensitivity/erm_dummies.pheno"), quote = FALSE, row.names = FALSE, col.names = TRUE)
```

I also have recalculated the grm in gcta here so that we can use one phenotype file rather than a big loop in mtg2

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/gcta_heritsens.txt

```{bash gcta_heritsens}
# # loop through each directory 
# directories=(
#   #  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/sensitivity/"
#     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/sensitivity/"
#   #  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/sensitivity/"
#   #  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/sensitivity/"
#   #  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/sensitivity/"
# )
# for dir in "${directories[@]}"; do
#     echo "Processing: $dir"
#     cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
#     
#     # GRM in gcta
#     /project/rche/data/datasets/addhealth/scratch/kthompson/findme/./gcta64 \
#         --bfile ../clean_genetic_dat_full \
#         --make-grm-gz \
#         --autosome \
#         --out grm_full_out # excludes sex chromosomes
#     
#     # copy covariate files
#     cp ../oneE/covariate_cat_IID.txt .
#     cp ../oneE/covariate_cont_IID.txt .
# 
#     # Load trait names and prevalence
#     cut -f1 trait_prevalence.tsv > trait_list.txt
#     cut -f2 trait_prevalence.tsv > prevalence_list.txt
#     
#     paste trait_list.txt prevalence_list.txt | while read trait prevalence; do
#       colnum=$(awk -v var="$trait" '{for (i=1; i<=NF; i++) if ($i == var) print i}' <(head -n1 erm_dummies.pheno))
#     
#       /project/rche/data/datasets/addhealth/scratch/kthompson/findme/./gcta64 \
#         --grm grm_full_out \
#         --pheno erm_dummies.pheno \
#         --mpheno $((colnum - 2)) \
#         --reml \
#         --prevalence $prevalence \
#         --reml-alg 1 \
#         --covar covariate_cat_IID.txt \
#         --qcovar covariate_cont_IID.txt \
#         --out h2_${trait}
#     done
# done 
```

```{r read all files and plot heritability estimates}
h2files <- list.files(path = paste0(ancestry_sub_path, "sensitivity"), pattern = "^h2_.*\\.hsq$", full.names = TRUE)

h2_data <- map_dfr(h2files, function(f) {
  dat <- read.table(f, header = TRUE, fill = TRUE)
  tibble(
    trait = stringr::str_remove(basename(f), "^h2_|\\.hsq$"),
    h2 = as.numeric(dat$Variance[dat$Source == "V(G)/Vp"]),
    se = as.numeric(dat$SE[dat$Source == "V(G)/Vp"])
  )
}) %>%
  mutate(
    CI_lower = h2 - 1.96 * se,
    CI_upper = h2 + 1.96 * se
  ) %>%
  arrange(desc(h2)) %>%
  mutate(sig = ifelse(CI_lower > 0, "Significant", "Not significant"))

# Plot
h2_all <- ggplot(h2_data, aes(y = h2, x = fct_reorder(trait, h2))) +
  geom_errorbar(aes(ymin = CI_lower, ymax = CI_upper), color = "grey80", width = 0.3) +
  geom_point(aes(color = sig), size = 1) +
  geom_hline(yintercept = 0, linetype = "dashed", color = "gray50") +
  scale_color_manual(values = c("Significant" = "#1b9e77", "Not significant" = "#d95f02")) +
  labs(
    title = "Heritability estimates of dummy-coded items",
    y = "Heritability (h²)", x = "Items",
    color = ""
  ) +
  scale_y_continuous(breaks = c(-0.4, -0.2, 0, 0.2, 0.4, 0.6, 0.8), limits = c(-0.4, 0.8)) +
  theme_classic(base_size = 14) +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    plot.title = element_text(face = "bold")
  )


# Save the updated plot
save_gg_cairo(plot = h2_all,  "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/plot_h2.png",
              width = 1000, height = 1100)

# Save the updated plot
save_gg_cairo(plot = h2_all,  "~/findmedep/figures/plot_h2.png",
              width = 800, height = 600)
```

Most heritable 10 factors:

H1PF35_numeric_raw_dummy_8.hsq : feel socially accepted - don’t know
H1GH28_numeric_raw_dummy_5.hsq : how do you think of yourself in weight - very overweight
H1NR3_numeric_raw_dummy_8.hsq  : given someone sex in exchange for drugs or money - dont know
H1GH48_numeric_raw_dummy_4.hsq : health or emotional problem caused day off school - every day
H1PF1_numeric_raw_dummy_3.hsq  : your mother is warm and loving to you - neither agree nor disagree
H1GH5_numeric_raw_dummy_2.hsq  : cold sweats - about once a week
H1HS3_numeric_raw_dummy_1.hsq  : In the past year, have you received psychological or emotional counseling? - yes
H1PF24_numeric_raw_dummy_3.hsq : You are satisfied with the way you and your father communicate with each other - neither agree nor disagree
H1PF33_numeric_raw_dummy_8.hsq : You like yourself just the way you are - dont know 
H1PF36_numeric_raw_dummy_8.hsq : You feel loved and wanted - dont know 

### Remove most heritable factors 

```{r remove heritable items}
# remove top 10 heritable items
dat.erm.theory1.dummy_reduced <- dat.erm.theory1.dummy %>%
  select(-H1PF35_numeric_raw_dummy_8, -H1GH28_numeric_raw_dummy_5, -H1NR3_numeric_raw_dummy_8, -H1GH48_numeric_raw_dummy_4, -H1PF1_numeric_raw_dummy_3, -H1GH5_numeric_raw_dummy_2, -H1HS3_numeric_raw_dummy_1, -H1PF24_numeric_raw_dummy_3, -H1PF33_numeric_raw_dummy_8, -H1PF36_numeric_raw_dummy_8)
```

PCA transformation

```{r PCA transformation for dat.erm.theory1.dummy}
# create data as matrix without ID column 
head(dat.erm.theory1.dummy_reduced)
dat.erm.theory1.dummy_reduced_noID <- as.matrix(dat.erm.theory1.dummy_reduced[,-1]) 
head(as.data.frame(dat.erm.theory1.dummy_reduced_noID))

# covariance matrix of all the variables 
dat.erm.theory1.dummy_reduced_covmat <- var(dat.erm.theory1.dummy_reduced_noID)  

# PCA transformation
dat.erm.theory1.dummy_reduced_eig <- eigen(dat.erm.theory1.dummy_reduced_covmat) # extract eigen values and vectors 
is.matrix(dat.erm.theory1.dummy_reduced_eig$vectors) # check this 

dat.erm.theory1.dummy_reduced_pca_vector <- dat.erm.theory1.dummy_reduced_noID%*%dat.erm.theory1.dummy_reduced_eig$vectors # multiply all exposures by the eigenvectors
dat.erm.theory1.dummy_reduced_pca_scaled <- scale(dat.erm.theory1.dummy_reduced_pca_vector) # normalise this matrix - minus the mean and divide by SD 

m <- dim(dat.erm.theory1.dummy_reduced_pca_scaled)[2] # m is the number of variables 

# divide each element by sqrt(m)
dat.erm.theory1.dummy_reduced_pca <- dat.erm.theory1.dummy_reduced_pca_scaled/sqrt(m)
head(as.data.frame(dat.erm.theory1.dummy_reduced_pca))

# attach AID variable
dat.erm.theory1.dummy_reduced_pca <- as.data.frame(cbind(dat.erm.theory1.dummy_reduced$AID, dat.erm.theory1.dummy_reduced_pca)) # add ID variable

# rename ID variable and make all vars numeric 
dat.erm.theory1.dummy_reduced_pca <- dat.erm.theory1.dummy_reduced_pca %>% dplyr::rename(AID = V1) %>% mutate(across(everything(), as.numeric))

# have a look
head(dat.erm.theory1.dummy_reduced_pca)
dim(dat.erm.theory1.dummy_reduced_pca)

# drop any empty variables
dat.erm.theory1.dummy_reduced_pca <- dat.erm.theory1.dummy_reduced_pca %>% select(where(~ any(!is.na(.))))
```

Reorder the dataset to match other files

```{r create erm data with only matching ids}
# filter to only those that are matching for erm data 
dat.erm.theory1_sens_reduced <- as.data.frame(dat.erm.theory1.dummy_reduced_pca) %>%
  filter(AID %in% ALL_matching_ids_nosibs) 

dim(dat.erm.theory1_sens)

# attach additional AID variable - mtg 
dat.erm.theory1_sens_reduced_mtg <- cbind(dat.erm.theory1_sens_reduced$AID, dat.erm.theory1_sens_reduced)
dat.erm.theory1_sens_reduced_mtg <- dat.erm.theory1_sens_reduced_mtg %>% dplyr::rename(AID2 = AID, AID = `dat.erm.theory1_sens_reduced$AID`) 

head(dat.erm.theory1_sens_reduced_mtg)

# reorder to match genetic file 
dat.erm.theory1_sens_reduced_reordered <- dat.erm.theory1_sens_reduced_mtg[order(match(dat.erm.theory1_sens_reduced_mtg$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

head(dat.erm.theory1_sens_reduced_reordered)

# check they are the same order
identical(as.numeric(matching_id_grm_erm_FID$AID), dat.erm.theory1_sens_reduced_reordered$AID) # these should be the same - so now need to run the ERM to match and remember to use this ID file as the input file when running the model 

# make all numeric
dat.erm.theory1_sens_reduced_reordered_numeric <- apply(dat.erm.theory1_sens_reduced_reordered, 2, as.numeric)
is.matrix(dat.erm.theory1_sens_reduced_reordered_numeric) && all(sapply(dat.erm.theory1_sens_reduced_reordered_numeric, is.numeric)) # check matrix is numeric

# have a look at data file
as.tibble(dat.erm.theory1_sens_reduced_reordered_numeric)
head(matching_id_grm_erm_AID)

# write pheno table
write.table(dat.erm.theory1_sens_reduced_reordered_numeric, paste0(ancestry_sub_path, "sensitivity/sens_h2_dummy_pca"), quote = FALSE, row.names = FALSE, col.names = FALSE)
```

### GE model 

Create a txt file with the erm and grm in a list together

```{r create e_g txt file}
# create txt file with erm and grm
erm.grm.h2sens.txt <- c("grm_full_out.grm_reformat", "sens_h2_dummy_pca_output.e") 
write.table(erm.grm.h2sens.txt, paste0(ancestry_sub_path, "sensitivity/erm.grm.h2sens.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_e_g_reml_sub_h2sens.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann: 

Compute REML for W1A and W1B with both g and e included in the model. 

```{bash compute ereml and greml model}
# directories=(
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/sensitivity/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/sensitivity/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/sensitivity/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/sensitivity/"
#      "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/sensitivity/"
# )
# 
# for dir in "${directories[@]}"; do
#      echo "Processing: $dir"
#      cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
#
#      # Create relationship matrix 
#      ./../../mtg2 -p IDs_IID.fam -pdmx sens_h2_dummy_pca -thread 100 -out sens_h2_dummy_pca_output.e
# 
#      # E-REML and G-REML
#      ./../../mtg2 -p IDs_IID.fam -mg erm.grm.h2sens.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_reml_h2sens_W1A.out > e_g_reml_h2sens_W1A.log
#      ./../../mtg2 -p IDs_IID.fam -mg erm.grm.h2sens.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_reml_h2sens_W1B.out > e_g_reml_h2sens_W1B.log
#      ./../../mtg2 -p IDs_IID.fam -mg erm.grm.h2sens.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_reml_h2sens_W4.out > e_g_reml_h2sens_W4.log
#      
#      echo "Finished processing $dir"
# done
```

#### Delta output prep

```{bash pick out LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/sensitivity/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_reml_h2sens_W1A.out > e_g_reml_h2sens_W1A.out2
       grep -vwE '(LKH|h2)' e_g_reml_h2sens_W1B.out > e_g_reml_h2sens_W1B.out2
       grep -vwE '(LKH|h2)' e_g_reml_h2sens_W4.out > e_g_reml_h2sens_W4.out2
done
```

Prepare a do file for MTG2 on to construct functions of model parameters in R: 
The numbers represent the rows in the out file - corresponding estimates of V. 

* ‘R’ is to get the ratio of variance
* R 2 1 3 4 (i.e. 2 / (1 + 2 + 3 + 4))
* R 3 1 2 4 (i.e. 3 / (1 + 2 + 3 + 4))
* R 4 1 2 3 (i.e. 4 / (1 + 2 + 3 + 4))
* The format here is to put the number of the estimate you want first and then all other numbers after

So R 2 1 3 4 4 will do 2 / 1 + 2 + 3 + 4 + 4
* Cannot compute the proportion for cov(ge) because it does not let you do 2*4 in the do file. But we can work this out ourselves doing 1-all variance components (without a SE estimate).

```{r do file for e_g_reml_h2sens_W1A}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_reml_h2sens_W1A.do")          # create file with this name
cat("e_g_reml_h2sens_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(3, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3", "\n")             # line 3: compute prop. of variance due to genetics
cat("R 3 1 2", "\n")             # line 4: compute prop. of variance due to environments
cat("R 1 2 3", "\n")             # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_reml_h2sens_W1B}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_reml_h2sens_W1B.do")
cat("e_g_reml_h2sens_W1B.out2", "\n")     # line 1: specify the file with parameter estimates
cat(3, "\n")                           # line 2: tot. # of variance & covariance components in the file
cat("R 2 1 3", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2", "\n")               # line 4: compute prop. of variance due to environments
cat("R 1 2 3", "\n")               # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_reml_h2sens_W4}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_reml_h2sens_W4.do")
cat("e_g_reml_h2sens_W4.out2", "\n")     # line 1: specify the file with parameter estimates
cat(3, "\n")                           # line 2: tot. # of variance & covariance components in the file
cat("R 2 1 3", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2", "\n")               # line 4: compute prop. of variance due to environments
cat("R 1 2 3", "\n")               # line 4: compute prop. of variance due to error
sink()
```

Output using Delta

```{bash output for e_g_cov_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/sensitivity/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_reml_h2sens_W1A.do > e_g_reml_W1A.out3  # NAME MATCHES ONE E - HAD TO CHANGE TO KEEP THE OUTPUT FUNCTION WORKING
    ./../../../mtg2 -delta2 e_g_reml_h2sens_W1B.do > e_g_reml_W1B.out3
    ./../../../mtg2 -delta2 e_g_reml_h2sens_W4.do > e_g_reml_W4.out3
done
```

#### Output table 

```{r output for e_reml_W1A}
e_g_reml_h2sens_W1A.output <- extract_delta_output(outcome = "W1A", 
                                         model_prefix = "e_g",
                                         ancestry_sub_path,
                                         sensitivity = TRUE,
                                         domain_labels = c("Genetic", "Environment", "Error"))

e_g_reml_h2sens_W1B.output <- extract_delta_output(outcome = "W1B", 
                                         model_prefix = "e_g",
                                         sensitivity = TRUE,
                                         ancestry_sub_path,
                                         domain_labels = c("Genetic", "Environment", "Error"))

e_g_reml_h2sens_W4.output <- extract_delta_output(outcome = "W4", 
                                         model_prefix = "e_g",
                                         sensitivity = TRUE,
                                         ancestry_sub_path,
                                         domain_labels = c("Genetic", "Environment", "Error"))

e_g_reml_h2sens_W1A.output
e_g_reml_h2sens_W1B.output
e_g_reml_h2sens_W4.output
```

### GxE

```{r filter and match nopca data }
# take out ID 
dat.erm.theory1_sens_reordered_noID <- as.matrix(dat.erm.theory1_sens_reordered[,-c(1,2)]) 

# have a look
head(as.data.frame(dat.erm.theory1_sens_reordered_noID))

# scale the data
dat.erm.theory1_sens_reordered_noID_scaled <- scale(dat.erm.theory1_sens_reordered_noID) # normalise this matrix - minus the mean and divide by SD 

# Need to remove any NA values
dat.erm.theory1_sens_reordered_noID_scaled <-  as.data.frame(dat.erm.theory1_sens_reordered_noID_scaled) %>% select(where(~ !all(is.na(.))))
dat.erm.theory1_sens_reordered_noID_scaled <- as.matrix(dat.erm.theory1_sens_reordered_noID_scaled)

# get number of variables
m <- dim(dat.erm.theory1_sens_reordered_noID_scaled)[2] # m is the number of variables - NOW 394

# divide by sqrt(m)
dat.erm.theory1_sens_reordered_noID_scaled <- dat.erm.theory1_sens_reordered_noID_scaled/sqrt(m)

head(as.data.frame(dat.erm.theory1_sens_reordered_noID_scaled))

# attach additional AID variable - mtg 
dat.erm.theory1_sens_reordered_mtg <- cbind(dat.erm.theory1_sens_reordered$AID, dat.erm.theory1_sens_reordered$AID, dat.erm.theory1_sens_reordered_noID_scaled)

dat.erm.theory1_sens_reordered_mtg <- as.data.frame(dat.erm.theory1_sens_reordered_mtg) %>% dplyr::rename(AID2 = V2, AID = V1) 

head(as.data.frame(dat.erm.theory1_sens_reordered_mtg))

# check they are the same order
identical(as.numeric(matching_id_grm_erm_FID$AID), dat.erm.theory1_sens_reordered_mtg$AID) # these should be the same - so now need to run the ERM to match and remember to use this ID file as the input file when running the model 

# make all numeric
dat.erm.theory1_sens_reordered_mtg_numeric <- apply(dat.erm.theory1_sens_reordered_mtg, 2, as.numeric)
is.matrix(dat.erm.theory1_sens_reordered_mtg_numeric) && all(sapply(dat.erm.theory1_sens_reordered_mtg_numeric, is.numeric)) # check matrix is numeric

# have a look at data file
as.tibble(dat.erm.theory1_sens_reordered_mtg_numeric)
head(matching_id_grm_erm_AID)

# write table 
write.table(dat.erm.theory1_sens_reordered_mtg_numeric, paste0(ancestry_sub_path, "sensitivity/sens_h2_dummy_nopca"), col.names = FALSE, row.names = FALSE, quote = FALSE)
```

```{r create e_g txt file}
# create txt file with erm and grm
erm.grm.gxe.txt <- c("grm_full_out.grm_reformat", "sens_h2_dummy_pca_output.e", "erm_grm.gxe") 
write.table(erm.grm.gxe.txt, paste0(ancestry_sub_path, "sensitivity/erm.grm.gxe.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

Create relatedness matrix without pca transformation and caluclate interaction term between ERM and GRM. 

This is included in h2sens txt file above! 

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_e_g_reml_sub_h2sens.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann:

```{bash e_g_reml_sub_h2sens}
# # GxE
#       # Create relationship matrix for non-transformed ERM
#        ./../../../mtg2 -p IDs_IID.fam -pdmx sens_h2_dummy_nopca -thread 100 -out sens_h2_dummy_nopca_output.e
#        
#       # create interaction term from erm x grm 
#        paste sens_h2_dummy_nopca_output.e grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > erm_grm.gxe
#        
#       # include in REML model 
#        ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.gxe.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_gxe_reml_W1A.out > e_g_gxe_reml_W1A.log
#        ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.gxe.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_gxe_reml_W1B.out > e_g_gxe_reml_W1B.log
#        ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.gxe.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_gxe_reml_W4.out > e_g_gxe_reml_W4.log
```

#### Output 

```{bash pick out LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/sensitivity/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_gxe_reml_W1A.out > e_g_gxe_reml_W1A.out2
       grep -vwE '(LKH|h2)' e_g_gxe_reml_W1B.out > e_g_gxe_reml_W1B.out2
       grep -vwE '(LKH|h2)' e_g_gxe_reml_W4.out > e_g_gxe_reml_W4.out2
done
```

```{r do file for e_g_gxe_reml_W1A}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_gxe_reml_W1A.do")          # create file with this name
cat("e_g_gxe_reml_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(4, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4", "\n")               # line 4: compute prop. of variance due to environments
cat("R 4 1 2 3", "\n")               # line 4: compute prop. of variance due to gxe
cat("R 1 2 3 4", "\n")               # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_gxe_reml_W1B}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_gxe_reml_W1B.do")
cat("e_g_gxe_reml_W1B.out2", "\n")     # line 1: specify the file with parameter estimates
cat(4, "\n")                           # line 2: tot. # of variance & gxeariance components in the file
cat("R 2 1 3 4", "\n")                 # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4", "\n")                 # line 4: compute prop. of variance due to environments
cat("R 4 1 2 3", "\n")                 # line 4: compute prop. of variance due to gxe
cat("R 1 2 3 4", "\n")                 # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_gxe_reml_W4}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_gxe_reml_W4.do")
cat("e_g_gxe_reml_W4.out2", "\n")      # line 1: specify the file with parameter estimates
cat(4, "\n")                           # line 2: tot. # of variance & gxeariance components in the file
cat("R 2 1 3 4", "\n")                 # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4", "\n")                 # line 4: compute prop. of variance due to environments
cat("R 4 1 2 3", "\n")                 # line 4: compute prop. of variance due to gxe
cat("R 1 2 3 4", "\n")                 # line 4: compute prop. of variance due to error
sink()
```

Output using Delta

```{bash output for e_g_gxe_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/sensitivity/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_gxe_reml_W1A.do > e_g_gxe_reml_W1A.out3
    ./../../../mtg2 -delta2 e_g_gxe_reml_W1B.do > e_g_gxe_reml_W1B.out3
    ./../../../mtg2 -delta2 e_g_gxe_reml_W4.do > e_g_gxe_reml_W4.out3
done
```


#### Output table

```{r gxe output}
e_g_gxe_reml_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "e_g_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error"),
  sensitivity = TRUE
)

e_g_gxe_reml_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "e_g_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error"),
  sensitivity = TRUE
)

e_g_gxe_reml_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "e_g_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error"),
  sensitivity = TRUE
)

e_g_gxe_reml.output <- rbind(e_g_gxe_reml_W1A.output, e_g_gxe_reml_W1B.output, e_g_gxe_reml_W4.output)
```

```{r h2sens output}
eg_h2sens <- rbind(e_g_reml_h2sens_W1A.output, e_g_reml_h2sens_W1B.output, e_g_reml_h2sens_W4.output)

h2sens.output <- rbind(eg_h2sens, e_g_gxe_reml.output)
```

## 2. Remove health domains - SUPPLEMENT

We conduct the same models but now we remove all health items. 

```{r health items}
health_items <- c(
  "H1GH1_numeric_raw",  # General health
  "H1PL1_numeric_raw",  # Permanent physical condition
  "H1GH2_numeric_raw",  # Headaches
  "H1GH3_numeric_raw",  # Feeling hot
  "H1GH4_numeric_raw",  # Stomach aches
  "H1GH5_numeric_raw",  # Cold sweats
  "H1GH6_numeric_raw",  # Feeling weak
  "H1GH7_numeric_raw",  # Sore throat
  "H1GH8_numeric_raw",  # Fatigue
  "H1GH9_numeric_raw",  # Painful urination
  "H1GH10_numeric_raw", # Feeling really sick
  "H1GH11_numeric_raw", # Waking up tired
  "H1GH12_numeric_raw", # Skin problems
  "H1GH13_numeric_raw", # Dizziness
  "H1GH14_numeric_raw", # Chest pains
  "H1GH15_numeric_raw", # Aches and pains
  "H1GH16_numeric_raw", # Period pains
  "H1GH17_numeric_raw", # Poor appetite
  "H1GH18_numeric_raw", # Sleep problems
  "H1GH19_numeric_raw", # Trouble relaxing
  "H1GH22_numeric_raw", # Fearfulness
  "H1GH26_numeric_raw", # Avoided medical care
  "H1GH28_numeric_raw", # Perceived weight
  "H1GH29_numeric_raw", # Trying to change weight
  "H1HS3_numeric_raw",  # Psychological counseling
  "H1GH48_numeric_raw", # Missed school due to health/emotion
  "H1GH49_numeric_raw"  # Missed recreation due to health/emotion
)

as.data.frame(health_items)

health_pattern <- paste(health_items, collapse = "|")

dat.erm.theory1.dummy_health_sens <- dat.erm.theory1.dummy %>%
  select(-matches(health_pattern))
```

PCA transformation

```{r PCA transformation for dat.erm.theory1.dummy}
# create data as matrix without ID column 
head(dat.erm.theory1.dummy_health_sens)
dat.erm.theory1.dummy_health_sens_noID <- as.matrix(dat.erm.theory1.dummy_health_sens[,-1]) 
head(as.data.frame(dat.erm.theory1.dummy_health_sens_noID))

# covariance matrix of all the variables 
dat.erm.theory1.dummy_health_sens_covmat <- var(dat.erm.theory1.dummy_health_sens_noID)  

# PCA transformation
dat.erm.theory1.dummy_health_sens_eig <- eigen(dat.erm.theory1.dummy_health_sens_covmat) # extract eigen values and vectors 
is.matrix(dat.erm.theory1.dummy_health_sens_eig$vectors) # check this 

dat.erm.theory1.dummy_health_sens_pca_vector <- dat.erm.theory1.dummy_health_sens_noID%*%dat.erm.theory1.dummy_health_sens_eig$vectors # multiply all exposures by the eigenvectors
dat.erm.theory1.dummy_health_sens_pca_scaled <- scale(dat.erm.theory1.dummy_health_sens_pca_vector) # normalise this matrix - minus the mean and divide by SD 

m <- dim(dat.erm.theory1.dummy_health_sens_pca_scaled)[2] # m is the number of variables 

# divide each element by sqrt(m)
dat.erm.theory1.dummy_health_sens_pca <- dat.erm.theory1.dummy_health_sens_pca_scaled/sqrt(m)
head(as.data.frame(dat.erm.theory1.dummy_health_sens_pca))

# attach AID variable
dat.erm.theory1.dummy_health_sens_pca <- as.data.frame(cbind(dat.erm.theory1.dummy_health_sens$AID, dat.erm.theory1.dummy_health_sens_pca)) # add ID variable

# rename ID variable and make all vars numeric 
dat.erm.theory1.dummy_health_sens_pca <- dat.erm.theory1.dummy_health_sens_pca %>% dplyr::rename(AID = V1) %>% mutate(across(everything(), as.numeric))

# have a look
head(dat.erm.theory1.dummy_health_sens_pca)
dim(dat.erm.theory1.dummy_health_sens_pca)

# drop any empty variables
dat.erm.theory1.dummy_health_sens_pca <- dat.erm.theory1.dummy_health_sens_pca %>% select(where(~ any(!is.na(.))))
```

Reorder the dataset to match other files

```{r create erm data with only matching ids}
# filter to only those that are matching for erm data 
dat.erm.theory1_sens_health <- as.data.frame(dat.erm.theory1.dummy_health_sens_pca) %>%
  filter(AID %in% ALL_matching_ids_nosibs) 

dim(dat.erm.theory1_sens_health)

# attach additional AID variable - mtg 
dat.erm.theory1_sens_health_mtg <- cbind(dat.erm.theory1_sens_health$AID, dat.erm.theory1_sens_health)
dat.erm.theory1_sens_health_mtg <- dat.erm.theory1_sens_health_mtg %>% dplyr::rename(AID2 = AID, AID = `dat.erm.theory1_sens_health$AID`) 

head(dat.erm.theory1_sens_health_mtg)

# reorder to match genetic file 
dat.erm.theory1_sens_health_reordered <- dat.erm.theory1_sens_health_mtg[order(match(dat.erm.theory1_sens_health_mtg$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

head(dat.erm.theory1_sens_health_reordered)

# check they are the same order
identical(as.numeric(matching_id_grm_erm_FID$AID), dat.erm.theory1_sens_health_reordered$AID) # these should be the same - so now need to run the ERM to match and remember to use this ID file as the input file when running the model 

# make all numeric
dat.erm.theory1_sens_health_reordered_numeric <- apply(dat.erm.theory1_sens_health_reordered, 2, as.numeric)
is.matrix(dat.erm.theory1_sens_health_reordered_numeric) && all(sapply(dat.erm.theory1_sens_health_reordered_numeric, is.numeric)) # check matrix is numeric

# have a look at data file
as.tibble(dat.erm.theory1_sens_health_reordered_numeric)
head(matching_id_grm_erm_AID)

# write pheno table
write.table(dat.erm.theory1_sens_health_reordered_numeric, paste0(ancestry_sub_path, "sensitivity/sens_health_dummy_pca"), quote = FALSE, row.names = FALSE, col.names = FALSE)
```

No pca file - reorder

```{r create erm data with only matching ids}
# filter to only those that are matching for erm data 
dat.erm.theory1_sens_health_nopca <- as.data.frame(dat.erm.theory1.dummy_health_sens) %>%
  filter(AID %in% ALL_matching_ids_nosibs) 

dim(dat.erm.theory1_sens_health_nopca)

# attach additional AID variable - mtg 
dat.erm.theory1_sens_health_nopca_mtg <- cbind(dat.erm.theory1_sens_health_nopca$AID, dat.erm.theory1_sens_health_nopca)
dat.erm.theory1_sens_health_nopca_mtg <- dat.erm.theory1_sens_health_nopca_mtg %>% dplyr::rename(AID2 = AID, AID = `dat.erm.theory1_sens_health_nopca$AID`) 

head(dat.erm.theory1_sens_health_nopca_mtg)

# reorder to match genetic file 
dat.erm.theory1_sens_health_nopca_reordered <- dat.erm.theory1_sens_health_nopca_mtg[order(match(dat.erm.theory1_sens_health_nopca_mtg$AID, as.numeric(matching_id_grm_erm_FID$AID))), ]

head(dat.erm.theory1_sens_health_nopca_reordered)
```

No pca file - scale

```{r filter and match nopca data }
# take out ID 
dat.erm.theory1_sens_health_reordered_noID <- as.matrix(dat.erm.theory1_sens_health_nopca_reordered[,-c(1,2)]) 

# have a look
head(as.data.frame(dat.erm.theory1_sens_health_reordered_noID))

# scale the data
dat.erm.theory1_sens_health_reordered_noID_scaled <- scale(dat.erm.theory1_sens_health_reordered_noID) # normalise this matrix - minus the mean and divide by SD 

# Need to remove any NA values
dat.erm.theory1_sens_health_reordered_noID_scaled <-  as.data.frame(dat.erm.theory1_sens_health_reordered_noID_scaled) %>% select(where(~ !all(is.na(.))))
dat.erm.theory1_sens_health_reordered_noID_scaled <- as.matrix(dat.erm.theory1_sens_health_reordered_noID_scaled)

# get number of variables
m <- dim(dat.erm.theory1_sens_health_reordered_noID_scaled)[2] # m is the number of variables - NOW 394

# divide by sqrt(m)
dat.erm.theory1_sens_health_reordered_noID_scaled <- dat.erm.theory1_sens_health_reordered_noID_scaled/sqrt(m)

head(as.data.frame(dat.erm.theory1_sens_health_reordered_noID_scaled))

# attach additional AID variable - mtg 
dat.erm.theory1_sens_health_reordered_mtg <- cbind(dat.erm.theory1_sens_health_nopca_reordered$AID, dat.erm.theory1_sens_health_nopca_reordered$AID, dat.erm.theory1_sens_health_reordered_noID_scaled)

dat.erm.theory1_sens_health_reordered_mtg <- as.data.frame(dat.erm.theory1_sens_health_reordered_mtg) %>% dplyr::rename(AID2 = V2, AID = V1) 

head(as.data.frame(dat.erm.theory1_sens_health_reordered_mtg))

# check they are the same order
identical(as.numeric(matching_id_grm_erm_FID$AID), dat.erm.theory1_sens_health_reordered_mtg$AID) # these should be the same - so now need to run the ERM to match and remember to use this ID file as the input file when running the model 

# make all numeric
dat.erm.theory1_sens_health_reordered_mtg_numeric <- apply(dat.erm.theory1_sens_health_reordered_mtg, 2, as.numeric)
is.matrix(dat.erm.theory1_sens_health_reordered_mtg_numeric) && all(sapply(dat.erm.theory1_sens_health_reordered_mtg_numeric, is.numeric)) # check matrix is numeric

# have a look at data file
as.tibble(dat.erm.theory1_sens_health_reordered_mtg_numeric)
head(matching_id_grm_erm_AID)

# write table 
write.table(dat.erm.theory1_sens_health_reordered_mtg_numeric, paste0(ancestry_sub_path, "sensitivity/sens_health_dummy_nopca"), col.names = FALSE, row.names = FALSE, quote = FALSE)
```

Create txt files

```{r create e_g txt file}
# create txt file with erm and grm
erm.grm.healthsens.txt <- c("grm_full_out.grm_reformat", "sens_health_dummy_pca_output.e") 
write.table(erm.grm.healthsens.txt, paste0(ancestry_sub_path, "sensitivity/erm.grm.healthsens.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")

# create txt file with gxe
erm.grm.healthsens.gxe.txt <- c("grm_full_out.grm_reformat", "sens_health_dummy_pca_output.e", "erm_grm_healthsens.gxe") 
write.table(erm.grm.healthsens.gxe.txt, paste0(ancestry_sub_path, "sensitivity/erm.grm.healthsens.gxe.txt"), quote=FALSE, row.names=FALSE, col.names=FALSE, sep="\t")
```

### GE model 

This will run both GE and GxE models

**Navigate to the terminal/shell on Rossmann, and execute:**
sbatch /project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/mtg2_e_g_reml_sub_healthsens.txt

You can then run the following to check it's status:
squeue -u thom1336

This will run the following as a job on Rossmann: 

```{bash e_g_reml_sub_healthsens}
# # Create relationship matrix 
#     #  ./../../../mtg2 -p IDs_IID.fam -pdmx sens_health_dummy_pca -thread 100 -out sens_health_dummy_pca_output.e
#  
#       # E-REML and G-REML
#     #  ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.healthsens.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_reml_healthsens_W1A.out > e_g_reml_healthsens_W1A.log
#     #  ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.healthsens.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_reml_healthsens_W1B.out > e_g_reml_healthsens_W1B.log
#     #  ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.healthsens.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_reml_healthsens_W4.out > e_g_reml_healthsens_W4.log
#       
#       # GxE
#       # Create relationship matrix for non-transformed ERM
#        ./../../../mtg2 -p IDs_IID.fam -pdmx sens_health_dummy_nopca -thread 100 -out sens_health_dummy_nopca_output.e
#        
#       # create interaction term from erm x grm 
#        paste sens_health_dummy_nopca_output.e grm_full_out.grm_reformat | awk '{ print $1, $2, $3 * $7 }' > erm_grm_healthsens.gxe
#        
#       # include in REML model 
#        ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.healthsens.gxe.txt -d pheno_W1A_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_gxe_reml_healthsens_W1A.out > e_g_gxe_reml_healthsens_W1A.log
#        ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.healthsens.gxe.txt -d pheno_W1B_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_gxe_reml_healthsens_W1B.out > e_g_gxe_reml_healthsens_W1B.log
#        ./../../../mtg2 -p IDs_IID.fam -mg erm.grm.healthsens.gxe.txt -d pheno_W4_IID.txt -cc covariate_cat_IID.txt –qc covariate_cont_IID.txt -mod 1 -thread 100 -out e_g_gxe_reml_healthsens_W4.out > e_g_gxe_reml_healthsens_W4.log
```

```{bash pick out LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/sensitivity/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_reml_healthsens_W1A.out > e_g_reml_healthsens_W1A.out2
       grep -vwE '(LKH|h2)' e_g_reml_healthsens_W1B.out > e_g_reml_healthsens_W1B.out2
       grep -vwE '(LKH|h2)' e_g_reml_healthsens_W4.out > e_g_reml_healthsens_W4.out2
done
```

Prepare a do file for MTG2 on to construct functions of model parameters in R: 
The numbers represent the rows in the out file - corresponding estimates of V. 

* ‘R’ is to get the ratio of variance
* R 2 1 3 4 (i.e. 2 / (1 + 2 + 3 + 4))
* R 3 1 2 4 (i.e. 3 / (1 + 2 + 3 + 4))
* R 4 1 2 3 (i.e. 4 / (1 + 2 + 3 + 4))
* The format here is to put the number of the estimate you want first and then all other numbers after

So R 2 1 3 4 4 will do 2 / 1 + 2 + 3 + 4 + 4
* Cannot compute the proportion for cov(ge) because it does not let you do 2*4 in the do file. But we can work this out ourselves doing 1-all variance components (without a SE estimate).

```{r do file for e_g_reml_healthsens_W1A}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_reml_healthsens_W1A.do")          # create file with this name
cat("e_g_reml_healthsens_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(3, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3", "\n")             # line 3: compute prop. of variance due to genetics
cat("R 3 1 2", "\n")             # line 4: compute prop. of variance due to environments
cat("R 1 2 3", "\n")             # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_reml_healthsens_W1B}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_reml_healthsens_W1B.do")
cat("e_g_reml_healthsens_W1B.out2", "\n")     # line 1: specify the file with parameter estimates
cat(3, "\n")                           # line 2: tot. # of variance & covariance components in the file
cat("R 2 1 3", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2", "\n")               # line 4: compute prop. of variance due to environments
cat("R 1 2 3", "\n")               # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_reml_healthsens_W4}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_reml_healthsens_W4.do")
cat("e_g_reml_healthsens_W4.out2", "\n")     # line 1: specify the file with parameter estimates
cat(3, "\n")                           # line 2: tot. # of variance & covariance components in the file
cat("R 2 1 3", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2", "\n")               # line 4: compute prop. of variance due to environments
cat("R 1 2 3", "\n")               # line 4: compute prop. of variance due to error
sink()
```

Output using Delta - IMPORTANT ALWAYS RUN THIS WITH 

```{bash output for e_g_cov_reml}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/sensitivity/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_reml_healthsens_W1A.do > e_g_reml_W1A.out3  # NAME MATCHES ONE E - HAD TO CHANGE TO KEEP THE OUTPUT FUNCTION WORKING
    ./../../../mtg2 -delta2 e_g_reml_healthsens_W1B.do > e_g_reml_W1B.out3
    ./../../../mtg2 -delta2 e_g_reml_healthsens_W4.do > e_g_reml_W4.out3
done
```

#### Output table 

**IMPORTANT ALWAYS RUN ABOVE CHUNK WITH THE OUTPUT TABLE CODE OTHERWISE MAY GET WRONG MODEL BEING CALLED**

```{r output for e_reml_W1A}
e_g_reml_healthsens_W1A.output <- extract_delta_output(outcome = "W1A", 
                                         model_prefix = "e_g",
                                         ancestry_sub_path,
                                         sensitivity = TRUE,
                                         domain_labels = c("Genetic", "Environment", "Error"))

e_g_reml_healthsens_W1B.output <- extract_delta_output(outcome = "W1B", 
                                         model_prefix = "e_g",
                                         sensitivity = TRUE,
                                         ancestry_sub_path,
                                         domain_labels = c("Genetic", "Environment", "Error"))

e_g_reml_healthsens_W4.output <- extract_delta_output(outcome = "W4", 
                                         model_prefix = "e_g",
                                         sensitivity = TRUE,
                                         ancestry_sub_path,
                                         domain_labels = c("Genetic", "Environment", "Error"))

e_g_reml_healthsens_W1A.output
e_g_reml_healthsens_W1B.output
e_g_reml_healthsens_W4.output
```

### GxE 

```{bash pick out LKH}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/sensitivity/"
)

for dir in "${directories[@]}"; do
cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
       grep -vwE '(LKH|h2)' e_g_gxe_reml_healthsens_W1A.out > e_g_gxe_reml_healthsens_W1A.out2
       grep -vwE '(LKH|h2)' e_g_gxe_reml_healthsens_W1B.out > e_g_gxe_reml_healthsens_W1B.out2
       grep -vwE '(LKH|h2)' e_g_gxe_reml_healthsens_W4.out > e_g_gxe_reml_healthsens_W4.out2
done
```

```{r do file for e_g_gxe_reml_healthsens_W1A}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_gxe_reml_healthsens_W1A.do")          # create file with this name
cat("e_g_gxe_reml_healthsens_W1A.out2", "\n")   # line 1: specify the file with parameter estimates (above)
cat(4, "\n")                         # line 2: Number of variance & covariance (Ve and Va) components in the file
cat("R 2 1 3 4", "\n")               # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4", "\n")               # line 4: compute prop. of variance due to environments
cat("R 4 1 2 3", "\n")               # line 4: compute prop. of variance due to gxe
cat("R 1 2 3 4", "\n")               # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_gxe_reml_healthsens_W1B}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_gxe_reml_healthsens_W1B.do")
cat("e_g_gxe_reml_healthsens_W1B.out2", "\n")     # line 1: specify the file with parameter estimates
cat(4, "\n")                           # line 2: tot. # of variance & gxeariance components in the file
cat("R 2 1 3 4", "\n")                 # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4", "\n")                 # line 4: compute prop. of variance due to environments
cat("R 4 1 2 3", "\n")                 # line 4: compute prop. of variance due to gxe
cat("R 1 2 3 4", "\n")                 # line 4: compute prop. of variance due to error
sink()
```

```{r do file for e_g_gxe_reml_healthsens_W4}
setwd(paste0(ancestry_sub_path, "sensitivity/"))
sink("e_g_gxe_reml_healthsens_W4.do")
cat("e_g_gxe_reml_healthsens_W4.out2", "\n")      # line 1: specify the file with parameter estimates
cat(4, "\n")                           # line 2: tot. # of variance & gxeariance components in the file
cat("R 2 1 3 4", "\n")                 # line 3: compute prop. of variance due to genetics
cat("R 3 1 2 4", "\n")                 # line 4: compute prop. of variance due to environments
cat("R 4 1 2 3", "\n")                 # line 4: compute prop. of variance due to gxe
cat("R 1 2 3 4", "\n")                 # line 4: compute prop. of variance due to error
sink()
```

Output using Delta

```{bash output for e_g_gxe_reml_healthsens}
directories=(
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/ALL/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EUR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AFR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/AMR/sensitivity/"
     "/project/rche/data/datasets/addhealth/scratch/kthompson/findme/sub_sample/EAS/sensitivity/"
)

for dir in "${directories[@]}"; do
    echo "Processing directory: $dir"
    cd "$dir" || { echo "Failed to cd into $dir"; exit 1; }
    
    ./../../../mtg2 -delta2 e_g_gxe_reml_healthsens_W1A.do > e_g_gxe_reml_W1A.out3
    ./../../../mtg2 -delta2 e_g_gxe_reml_healthsens_W1B.do > e_g_gxe_reml_W1B.out3
    ./../../../mtg2 -delta2 e_g_gxe_reml_healthsens_W4.do > e_g_gxe_reml_W4.out3
done
```

#### Output table 

**IMPORTANT ALWAYS RUN ABOVE CHUNK WITH THE OUTPUT TABLE CODE OTHERWISE MAY GET WRONG MODEL BEING CALLED**

```{r gxe output}
e_g_gxe_reml_healthsens_W1A.output <- extract_delta_output(
  outcome = "W1A",
  model_prefix = "e_g_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error"),
  sensitivity = TRUE
)

e_g_gxe_reml_healthsens_W1B.output <- extract_delta_output(
  outcome = "W1B",
  model_prefix = "e_g_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error"),
  sensitivity = TRUE
)

e_g_gxe_reml_healthsens_W4.output <- extract_delta_output(
  outcome = "W4",
  model_prefix = "e_g_gxe",
  ancestry_sub_path = ancestry_sub_path,
  domain_labels = c("Genetic", "Environment", "Gene-environment interaction", "Error"),
  sensitivity = TRUE
)

e_g_gxe_reml_healthsens.output <- rbind(e_g_gxe_reml_healthsens_W1A.output, e_g_gxe_reml_healthsens_W1B.output, e_g_gxe_reml_healthsens_W4.output)
```

```{r full health removed output table}
eg_healthsens <- rbind(e_g_reml_healthsens_W1A.output, e_g_reml_healthsens_W1B.output, e_g_reml_healthsens_W4.output)
healthsens_output <- rbind(eg_healthsens, e_g_gxe_reml_healthsens.output)
```


